// @cornerstonejs/tools/dist/esm/utilities/voi/windowlevel/extractWindowLevelRegionToolData.js
import { cache, utilities as csUtils, StackViewport, VolumeViewport, } from '@cornerstonejs/core';
function extractWindowLevelRegionToolData(viewport) {
    if (viewport instanceof VolumeViewport) {
        return extractImageDataVolume(viewport);
    }
    if (viewport instanceof StackViewport) {
        return extractImageDataStack(viewport);
    }
    if (csUtils.isGenericViewport(viewport)) {
        return extractImageDataGeneric(viewport);
    }
    throw new Error('Viewport not supported');
}
function extractImageDataVolume(viewport) {
    const { scalarData, width, height } = csUtils.getCurrentVolumeViewportSlice(viewport);
    const { min: minPixelValue, max: maxPixelValue } = csUtils.getMinMax(scalarData);
    return {
        scalarData,
        minPixelValue,
        maxPixelValue,
        width,
        height,
        rows: width,
        columns: height,
    };
}
function extractImageDataStack(viewport) {
    const imageData = viewport.getImageData();
    const { scalarData } = imageData;
    const { min: minPixelValue, max: maxPixelValue } = csUtils.getMinMax(scalarData);
    const width = imageData.dimensions[0];
    const height = imageData.dimensions[1];
    const { rows, columns, color } = viewport.getCornerstoneImage();
    return {
        scalarData,
        width,
        height,
        minPixelValue,
        maxPixelValue,
        rows,
        columns,
        color,
    };
}
function extractImageDataGeneric(viewport) {
    const imageId = viewport.getCurrentImageId();
    if (!imageId) {
        console.warn('WindowLevelRegionTool: the current native viewport slice has no backing image (reformatted/oblique volume plane); skipping region window level.');
        return undefined;
    }
    const image = cache.getImage(imageId);
    if (!image) {
        throw new Error('Viewport not supported');
    }
    const scalarData = image.getPixelData();
    const { rows, columns, color } = image;
    const { min: minPixelValue, max: maxPixelValue } = csUtils.getMinMax(scalarData);
    return {
        scalarData,
        width: columns,
        height: rows,
        minPixelValue,
        maxPixelValue,
        rows,
        columns,
        color,
    };
}
export { extractWindowLevelRegionToolData };
