// @cornerstonejs/tools/dist/esm/tools/displayTools/Labelmap/labelmapRenderPlan/legacyVolumePlan.js
import { addVolumesToViewports, Enums, cache, volumeLoader, utilities, } from '@cornerstonejs/core';
import { triggerSegmentationDataModified, triggerSegmentationModified, } from '../../../../stateManagement/segmentation/triggerSegmentationEvents';
import { getLabelmaps, getOrCreateLabelmapVolume, } from '../../../../stateManagement/segmentation/helpers/labelmapSegmentationState';
import { addVolumesAsIndependentComponents } from '../addVolumesAsIndependentComponents';
import { createLabelmapRepresentationUID } from '../labelmapRepresentationUID';
import { createLabelmapRenderPlan } from './createLabelmapRenderPlan';
import { addLabelmapToPlanarGenericViewport, isPlanarNextVolumeViewport, } from './planarGenericVolumeLabelmap';
const { uuidv4 } = utilities;
function createLegacyVolumeLabelmapPlan({ config, isVolumeImageMapper, renderMode, segmentation, segmentationId, useSliceRendering, viewport, }) {
    return createLabelmapRenderPlan({
        isVolumeImageMapper,
        kind: 'legacy-volume',
        renderMode,
        segmentationId,
        useSliceRendering,
        viewport,
        getExpectedRepresentationUIDs: () => getExpectedVolumeLabelmapRepresentationUIDs(segmentation, segmentationId),
        mount: ({ labelMapData }) => mountLegacyVolumeLabelmap({
            config,
            labelMapData,
            segmentation,
            segmentationId,
            viewport,
        }),
    });
}
function getExpectedVolumeLabelmapRepresentationUIDs(segmentation, segmentationId) {
    return getLabelmaps(segmentation)
        .filter(canResolveLayerAsVolume)
        .map((layer) => createLabelmapRepresentationUID({
        segmentationId,
        referencedId: layer.labelmapId,
    }));
}
async function mountLegacyVolumeLabelmap({ config, labelMapData, segmentation, segmentationId, viewport, }) {
    const { id: viewportId } = viewport;
    const visibility = true;
    const immediateRender = false;
    const suppressEvents = true;
    const volumeCompatibleViewport = viewport;
    const labelmapLayers = getVolumeBackedLabelmapLayers(segmentation);
    if (!labelmapLayers.length) {
        const volumeLabelMapData = labelMapData;
        const volumeId = ensureVolumeHasVolumeId(volumeLabelMapData, segmentation);
        if (!cache.getVolume(volumeId)) {
            await handleMissingVolume(labelMapData);
        }
        labelmapLayers.push({
            labelmapId: volumeId,
            storageKind: 'volume',
            volumeId,
            imageIds: cache.getVolume(volumeId)?.imageIds,
        });
    }
    let blendMode = config?.blendMode ?? Enums.BlendModes.MAXIMUM_INTENSITY_BLEND;
    let useIndependentComponents = blendMode === Enums.BlendModes.LABELMAP_EDGE_PROJECTION_BLEND;
    if (useIndependentComponents) {
        const referenceVolumeId = volumeCompatibleViewport.getVolumeId?.();
        const segLabelmapVolumeId = labelmapLayers[0]?.volumeId;
        const baseVolume = referenceVolumeId
            ? cache.getVolume(referenceVolumeId)
            : undefined;
        const segVolume = segLabelmapVolumeId
            ? cache.getVolume(segLabelmapVolumeId)
            : undefined;
        if (!baseVolume || !segVolume) {
            useIndependentComponents = false;
            blendMode = Enums.BlendModes.MAXIMUM_INTENSITY_BLEND;
            console.debug('Independent components unavailable (missing reference or segmentation volume) - falling back to regular volume addition');
        }
        else {
            const segDims = segVolume.dimensions;
            const refDims = baseVolume.dimensions;
            if (segDims[0] !== refDims[0] ||
                segDims[1] !== refDims[1] ||
                segDims[2] !== refDims[2]) {
                useIndependentComponents = false;
                blendMode = Enums.BlendModes.MAXIMUM_INTENSITY_BLEND;
                console.debug('Dimensions mismatch - falling back to regular volume addition');
            }
        }
    }
    const volumeInputs = labelmapLayers.map((layer) => ({
        volumeId: layer.volumeId,
        visibility,
        representationUID: createLabelmapRepresentationUID({
            segmentationId,
            referencedId: layer.labelmapId,
        }),
        useIndependentComponents,
        blendMode,
    }));
    if (isPlanarNextVolumeViewport(viewport)) {
        return addLabelmapToPlanarGenericViewport({
            blendMode,
            labelmapLayers,
            segmentationId,
            viewport,
            visibility,
        });
    }
    if (!volumeInputs[0].useIndependentComponents) {
        await addVolumesToViewports(viewport.getRenderingEngine(), volumeInputs, [viewportId], immediateRender, suppressEvents);
        triggerSegmentationDataModified(segmentationId);
        return;
    }
    return addVolumesAsIndependentComponents({
        viewport: volumeCompatibleViewport,
        volumeInputs,
        segmentationId,
    });
}
function canResolveLayerAsVolume(layer) {
    return Boolean(layer.volumeId || layer.geometryVolumeId || layer.imageIds?.length);
}
function getVolumeBackedLabelmapLayers(segmentation) {
    return getLabelmaps(segmentation)
        .map((layer) => {
        if (layer.volumeId) {
            return layer;
        }
        const volume = getOrCreateLabelmapVolume(layer);
        if (!volume?.volumeId) {
            return;
        }
        return {
            ...layer,
            volumeId: volume.volumeId,
        };
    })
        .filter((layer) => Boolean(layer));
}
function ensureVolumeHasVolumeId(labelMapData, segmentation) {
    let { volumeId } = labelMapData;
    if (!volumeId) {
        volumeId = uuidv4();
        segmentation.representationData.Labelmap = {
            ...segmentation.representationData.Labelmap,
            volumeId,
        };
        labelMapData.volumeId = volumeId;
        triggerSegmentationModified(segmentation.segmentationId);
    }
    return volumeId;
}
async function handleMissingVolume(labelMapData) {
    const stackData = labelMapData;
    const hasImageIds = stackData.imageIds.length > 0;
    if (!hasImageIds) {
        throw new Error('cannot create labelmap, no imageIds found for the volume labelmap');
    }
    const volume = await volumeLoader.createAndCacheVolumeFromImages(labelMapData.volumeId ?? uuidv4(), stackData.imageIds);
    return volume;
}
export { createLegacyVolumeLabelmapPlan };
