// @cornerstonejs/tools/dist/esm/tools/displayTools/Labelmap/labelmapRenderPlan/volumeSliceImageMapperPlan.js
import { triggerSegmentationDataModified } from '../../../../stateManagement/segmentation/triggerSegmentationEvents';
import { addVolumeLabelmapImageMapperActors, getVolumeLabelmapImageMapperRepresentationUIDs, updateVolumeLabelmapImageMapperActors, } from '../volumeLabelmapImageMapper';
import { createLabelmapRenderPlan } from './createLabelmapRenderPlan';
function createVolumeSliceImageMapperPlan({ isVolumeImageMapper, renderMode, segmentation, segmentationId, useSliceRendering, viewport, }) {
    return createLabelmapRenderPlan({
        isVolumeImageMapper,
        kind: 'volume-slice-image-mapper',
        renderMode,
        segmentationId,
        useSliceRendering,
        viewport,
        getExpectedRepresentationUIDs: () => getVolumeLabelmapImageMapperRepresentationUIDs(viewport, segmentationId, segmentation),
        mount: () => mountVolumeLabelmapImageMapper({
            viewport,
            segmentation,
            segmentationId,
        }),
        update: ({ actorEntries }) => updateVolumeLabelmapImageMapperActors({
            viewport,
            segmentation,
            segmentationId,
            actorEntries,
        }),
    });
}
async function mountVolumeLabelmapImageMapper({ viewport, segmentation, segmentationId, }) {
    await addVolumeLabelmapImageMapperActors({
        viewport,
        segmentation,
        segmentationId,
    });
    triggerSegmentationDataModified(segmentationId);
}
export { createVolumeSliceImageMapperPlan };
