// @cornerstonejs/tools/dist/esm/tools/segmentation/utils/LazyBrushEditController.js
import { eventTarget } from '@cornerstonejs/core';
import { Events } from '../../../enums';
import { appendLazyBrushPreviewCircle, appendLazyBrushStrokePoint, } from './lazyBrushPreview';
class LazyBrushEditController {
    constructor() {
        this.strokePointsWorld = [];
        this.previewPoints = [];
        this.pendingPreviewCleanup = null;
    }
    reset() {
        this.strokePointsWorld = [];
        this.previewPoints = [];
    }
    clearPendingCleanup() {
        if (!this.pendingPreviewCleanup) {
            return;
        }
        eventTarget.removeEventListener(Events.SEGMENTATION_RENDERED, this.pendingPreviewCleanup.listener);
        this.pendingPreviewCleanup = null;
    }
    appendStrokePoint(worldPoint) {
        this.strokePointsWorld = appendLazyBrushStrokePoint(this.strokePointsWorld, worldPoint);
    }
    getStrokePointsWorld() {
        return this.strokePointsWorld;
    }
    capturePreviewCircle(hoverData) {
        if (!hoverData) {
            return;
        }
        const circlePoints = hoverData.brushCursor?.data?.editPoints;
        this.previewPoints = appendLazyBrushPreviewCircle(this.previewPoints, circlePoints);
        hoverData.brushCursor.data.handles.points = this.previewPoints;
    }
    scheduleCleanup({ element, centerCanvas, viewportId, segmentationId, refreshCursor, }) {
        this.clearPendingCleanup();
        const listener = ((evt) => {
            const detail = evt.detail;
            if (detail.viewportId !== viewportId ||
                detail.segmentationId !== segmentationId) {
                return;
            }
            this.clearPendingCleanup();
            this.reset();
            refreshCursor(element, centerCanvas);
        });
        this.pendingPreviewCleanup = {
            viewportId,
            segmentationId,
            listener,
        };
        eventTarget.addEventListener(Events.SEGMENTATION_RENDERED, listener);
    }
}
export default LazyBrushEditController;
