// @cornerstonejs/tools/dist/esm/utilities/getCalibratedUnits.js
import { Enums } from '@cornerstonejs/core';
const { CalibrationTypes } = Enums;
const PIXEL_UNITS = 'px';
const VOXEL_UNITS = 'voxels';
const SUPPORTED_REGION_DATA_TYPES = [
    1,
    2,
    3,
    4,
];
const SUPPORTED_PROBE_VARIANT = [
    '4,3',
    '4,7',
    '4,-1',
];
const UNIT_MAPPING = {
    0: 'px',
    1: 'percent',
    2: 'dB',
    3: 'cm',
    4: 'seconds',
    5: 'hertz',
    6: 'dB/seconds',
    7: 'cm/sec',
    8: 'cm\xb2',
    9: 'cm\xb2/s',
    0xc: 'degrees',
    [-1]: 'mV',
};
const EPS = 1e-3;
const SQUARE = '\xb2';
const types = [
    CalibrationTypes.ERMF,
    CalibrationTypes.USER,
    CalibrationTypes.ERROR,
    CalibrationTypes.PROJECTION,
    CalibrationTypes.CALIBRATED,
    CalibrationTypes.UNKNOWN,
];
const getCalibratedLengthUnitsAndScale = (image, handles) => {
    const { calibration, hasPixelSpacing, spacing = [1, 1, 1] } = image;
    let unit = hasPixelSpacing ? 'mm' : PIXEL_UNITS;
    const volumeUnit = hasPixelSpacing ? 'mm\xb3' : VOXEL_UNITS;
    let areaUnit = unit + SQUARE;
    const baseScale = calibration?.scale || 1;
    let scale = baseScale / (calibration?.columnPixelSpacing || spacing[0]);
    let scaleY = baseScale / (calibration?.rowPixelSpacing || spacing[1]);
    let scaleZ = baseScale / spacing[2];
    let calibrationType = '';
    if (!calibration ||
        (!calibration.type && !calibration.sequenceOfUltrasoundRegions)) {
        return { unit, areaUnit, scale, scaleY, scaleZ, volumeUnit };
    }
    if (types.includes(calibration?.type)) {
        calibrationType = calibration.type;
    }
    if (calibration.type === CalibrationTypes.UNCALIBRATED) {
        return {
            unit: PIXEL_UNITS,
            areaUnit: PIXEL_UNITS + SQUARE,
            scale,
            scaleY,
            scaleZ,
            volumeUnit: VOXEL_UNITS,
        };
    }
    if (calibration.sequenceOfUltrasoundRegions) {
        const region = calibration.sequenceOfUltrasoundRegions.find((region) => handles.every((handle) => handle[0] >= region.regionLocationMinX0 &&
            handle[0] <= region.regionLocationMaxX1 &&
            handle[1] >= region.regionLocationMinY0 &&
            handle[1] <= region.regionLocationMaxY1) &&
            (SUPPORTED_REGION_DATA_TYPES.includes(region.regionDataType) ||
                SUPPORTED_PROBE_VARIANT.includes(`${region.physicalUnitsXDirection},${region.physicalUnitsYDirection}`)));
        if (region &&
            region.physicalUnitsXDirection === region.physicalUnitsYDirection) {
            const physicalDeltaX = Math.abs(region.physicalDeltaX);
            const physicalDeltaY = Math.abs(region.physicalDeltaY);
            scale = 1 / physicalDeltaX;
            scaleY = 1 / physicalDeltaY;
            calibrationType = 'US Region';
            unit = UNIT_MAPPING[region.physicalUnitsXDirection] || 'unknown';
            areaUnit = unit + SQUARE;
        }
        else if (region && region.physicalUnitsYDirection === -1) {
            const physicalDeltaX = Math.abs(region.physicalDeltaX);
            const physicalDeltaY = Math.abs(region.physicalDeltaY);
            scale = 1 / physicalDeltaX;
            scaleY = 1 / physicalDeltaY;
            calibrationType = 'ECG Region';
            unit =
                UNIT_MAPPING[region.physicalUnitsXDirection] ||
                    UNIT_MAPPING[region.physicalUnitsYDirection] ||
                    'unknown';
            areaUnit =
                (UNIT_MAPPING[region.physicalUnitsYDirection] || 'px') + SQUARE;
        }
    }
    else if (calibration.scale) {
        scale = calibration.scale;
    }
    return {
        unit: unit + (calibrationType ? ` ${calibrationType}` : ''),
        areaUnit: areaUnit + (calibrationType ? ` ${calibrationType}` : ''),
        volumeUnit: volumeUnit + (calibrationType ? ` ${calibrationType}` : ''),
        scale,
        scaleY,
        scaleZ,
    };
};
const getCalibratedProbeUnitsAndValue = (image, handles) => {
    const [imageIndex] = handles;
    const { calibration } = image;
    let units = ['raw'];
    let values = [null];
    let calibrationType = '';
    if (!calibration ||
        (!calibration.type && !calibration.sequenceOfUltrasoundRegions)) {
        return { units, values };
    }
    if (calibration.sequenceOfUltrasoundRegions) {
        const supportedRegionsMetadata = calibration.sequenceOfUltrasoundRegions.filter((region) => (SUPPORTED_REGION_DATA_TYPES.includes(region.regionDataType) ||
            SUPPORTED_PROBE_VARIANT.includes(`${region.physicalUnitsXDirection},${region.physicalUnitsYDirection}`)) &&
            SUPPORTED_PROBE_VARIANT.includes(`${region.physicalUnitsXDirection},${region.physicalUnitsYDirection}`));
        if (!supportedRegionsMetadata?.length) {
            return { units, values };
        }
        const region = supportedRegionsMetadata.find((region) => imageIndex[0] >= region.regionLocationMinX0 &&
            imageIndex[0] <= region.regionLocationMaxX1 &&
            imageIndex[1] >= region.regionLocationMinY0 &&
            imageIndex[1] <= region.regionLocationMaxY1);
        if (!region) {
            return { units, values };
        }
        const { referencePixelX0 = 0, referencePixelY0 = 0 } = region;
        const { physicalDeltaX, physicalDeltaY } = region;
        const yValue = (imageIndex[1] - region.regionLocationMinY0 - referencePixelY0) *
            physicalDeltaY;
        const xValue = (imageIndex[0] - region.regionLocationMinX0 - referencePixelX0) *
            physicalDeltaX;
        calibrationType =
            region.physicalUnitsYDirection === -1 ? 'ECG Region' : 'US Region';
        values = [xValue, yValue];
        units = [
            UNIT_MAPPING[region.physicalUnitsXDirection] ?? 'unknown',
            UNIT_MAPPING[region.physicalUnitsYDirection] ?? 'unknown',
        ];
    }
    return {
        units,
        values,
        calibrationType,
    };
};
const getCalibratedAspect = (image) => image.calibration?.aspect || 1;
export { getCalibratedLengthUnitsAndScale, getCalibratedAspect, getCalibratedProbeUnitsAndValue, };
