// @cornerstonejs/metadata/dist/esm/displayset/createDisplaySetFromGroup.js
import { BaseDisplaySet } from './BaseDisplaySet';
import { ImageStackDisplaySet } from './ImageStackDisplaySet';
import { isEcgInstance } from './isEcgInstance';
import { isVideoInstance } from './isVideoInstance';
import { isWsiInstance } from './isWsiInstance';
import { getPreferredViewportType, getViewportTypesForGroup, } from './viewportTypes';
const RESERVED_ATTRIBUTE_KEYS = new Set([
    'imageIds',
    'underlyingImageIds',
    'instances',
    'displaySetId',
]);
function isAssignable(target, key) {
    let obj = target;
    while (obj) {
        const descriptor = Object.getOwnPropertyDescriptor(obj, key);
        if (descriptor) {
            if (descriptor.get || descriptor.set) {
                return typeof descriptor.set === 'function';
            }
            return descriptor.writable !== false;
        }
        obj = Object.getPrototypeOf(obj);
    }
    return true;
}
function applyCustomAttributes(displaySet, group, viewportTypes, options) {
    const { instances, matchedRule } = group;
    const first = instances[0];
    if (!matchedRule.customAttributes || !first) {
        return;
    }
    const sopClassUids = [
        ...new Set(instances.map((i) => i.SOPClassUID).filter(Boolean)),
    ];
    const isMultiFrame = Number(first.NumberOfFrames) > 1;
    const attributes = matchedRule.customAttributes({ instance: first, isMultiFrame, sopClassUids, viewportTypes }, {
        instances,
        splitNumber: options.splitNumber,
        descriptionName: options.descriptionName,
    });
    if (!attributes) {
        return;
    }
    for (const [key, value] of Object.entries(attributes)) {
        if (RESERVED_ATTRIBUTE_KEYS.has(key)) {
            continue;
        }
        if (isAssignable(displaySet, key)) {
            displaySet[key] = value;
        }
    }
    displaySet.preferredViewportType = getPreferredViewportType(displaySet.viewportTypes);
}
export function createDisplaySetFromGroup(group, options = {}) {
    const viewportTypes = getViewportTypesForGroup(group);
    const { instances } = group;
    const baseDisplaySetId = instances[0]?.SeriesInstanceUID ??
        `display-set-${instances[0]?.imageId ?? 'unknown'}`;
    const displaySetId = options.displaySetId ??
        (options.splitNumber
            ? `${baseDisplaySetId}:${options.splitNumber}`
            : baseDisplaySetId);
    const first = instances[0];
    let displaySet;
    if (first &&
        (isVideoInstance(first) || isEcgInstance(first) || isWsiInstance(first))) {
        const imageIds = instances.map((i) => i.imageId).filter(Boolean);
        displaySet = new BaseDisplaySet({
            displaySetId,
            viewportTypes,
            instances,
            imageIds: options.imageIds ?? imageIds,
            underlyingImageIds: imageIds,
        });
    }
    else {
        displaySet = ImageStackDisplaySet.fromInstances(instances, {
            displaySetId,
            viewportTypes,
            imageIds: options.imageIds,
        });
    }
    applyCustomAttributes(displaySet, group, viewportTypes, options);
    return displaySet;
}
