// @cornerstonejs/metadata/dist/esm/displayset/defaultDisplaySetSplitRules.js
import { isEcgInstance } from './isEcgInstance';
import { isImageInstance } from './isImageInstance';
import { isVideoInstance } from './isVideoInstance';
import { isWsiInstance } from './isWsiInstance';
const VOLUME_MODALITIES = new Set(['CT', 'MR', 'PT', 'NM']);
export const defaultDisplaySetSplitRules = [
    {
        id: 'video',
        viewportTypes: ['video'],
        matches: (instance) => isVideoInstance(instance),
        groupBy: ['SOPInstanceUID'],
    },
    {
        id: 'ecg',
        viewportTypes: ['ecg'],
        matches: (instance) => isEcgInstance(instance),
        groupBy: ['SOPInstanceUID'],
    },
    {
        id: 'wholeslide',
        viewportTypes: ['wholeslide'],
        matches: (instance) => isWsiInstance(instance),
        groupBy: ['SeriesInstanceUID'],
    },
    {
        id: 'singleImageModality',
        viewportTypes: ['stack'],
        matches: (instance) => ['CR', 'DX', 'MG'].includes(instance.Modality ?? '') &&
            isImageInstance(instance) &&
            !!instance.Rows,
        groupBy: [
            'SeriesInstanceUID',
            (instance) => `rows=${Math.round(Number(instance.Rows) / 64)}&cols=${Math.round(Number(instance.Columns) / 64)}`,
        ],
    },
    {
        id: 'multiFrame',
        viewportTypes: ['stack'],
        series: ({ instances }) => {
            const first = instances[0];
            return {
                isMultiFrame: Number(first?.NumberOfFrames) > 1 &&
                    first?.SliceLocation !== undefined,
            };
        },
        matches: (instance, { series }) => !!series.isMultiFrame && isImageInstance(instance) && !!instance.Rows,
        groupBy: ['SeriesInstanceUID', 'InstanceNumber'],
        customAttributes: ({ isMultiFrame }, options) => {
            const numberOfFrames = options.instances[0]?.NumberOfFrames;
            return {
                isClip: true,
                numImageFrames: numberOfFrames === undefined ? undefined : Number(numberOfFrames),
                splitNumber: options.splitNumber,
                isMultiFrame,
            };
        },
    },
    {
        id: 'mixedDimensionalityBValue',
        viewportTypes: ['volume', 'volume3d', 'stack'],
        series: ({ instances }) => {
            const [instance] = instances;
            if (!instance || instance.Modality !== 'MR') {
                return { mixedBValue: false };
            }
            const hasBValue = instances.some((i) => i.DiffusionBValue !== undefined);
            const missingBValue = instances.some((i) => i.DiffusionBValue === undefined);
            return { mixedBValue: hasBValue && missingBValue };
        },
        matches: (instance, { series }) => !!series.mixedBValue && isImageInstance(instance) && !!instance.Rows,
        groupBy: [
            'SeriesInstanceUID',
            (instance) => instance.DiffusionBValue === undefined,
        ],
    },
    {
        id: 'volume3d',
        viewportTypes: ['volume', 'volume3d', 'stack'],
        series: ({ instances }) => {
            const modality = instances[0]?.Modality;
            return {
                supportsVolume3d: !!modality && VOLUME_MODALITIES.has(modality) && instances.length > 1,
            };
        },
        matches: (instance, { series }) => !!series.supportsVolume3d && isImageInstance(instance) && !!instance.Rows,
        groupBy: ['SeriesInstanceUID'],
    },
    {
        id: 'defaultImageRule',
        viewportTypes: ['stack', 'volume', 'volume3d'],
        matches: (instance) => isImageInstance(instance) && !!instance.Rows,
    },
];
