// @cornerstonejs/tools/dist/esm/tools/base/BaseTool.js
import { utilities as csUtils, viewportHasPan, viewportHasZoom, } from '@cornerstonejs/core';
import ToolModes from '../../enums/ToolModes';
const { DefaultHistoryMemo } = csUtils.HistoryMemo;
class BaseTool {
    get configurationTyped() {
        return this.configuration;
    }
    constructor(toolProps, defaultToolProps) {
        this.isPrimary = false;
        const mergedDefaults = BaseTool.mergeDefaultProps(BaseTool.defaults, defaultToolProps);
        const initialProps = csUtils.deepMerge(mergedDefaults, toolProps);
        const { configuration = {}, supportedInteractionTypes, toolGroupId, } = initialProps;
        this.toolGroupId = toolGroupId;
        this.supportedInteractionTypes = supportedInteractionTypes || [];
        this.configuration = Object.assign({}, configuration);
        this.mode = ToolModes.Disabled;
    }
    static mergeDefaultProps(defaultProps = {}, additionalProps) {
        if (!additionalProps) {
            return defaultProps;
        }
        return csUtils.deepMerge(defaultProps, additionalProps);
    }
    static isSpecifiedTargetId(desiredVolumeId) {
        return (_viewport, { targetId }) => {
            return targetId.includes(desiredVolumeId);
        };
    }
    get toolName() {
        return this.getToolName();
    }
    getToolName() {
        return this.constructor.toolName;
    }
    applyActiveStrategy(enabledElement, operationData) {
        const { strategies, activeStrategy } = this.configuration;
        return strategies[activeStrategy]?.call(this, enabledElement, operationData);
    }
    applyActiveStrategyCallback(enabledElement, operationData, callbackType, ...extraArgs) {
        const { strategies, activeStrategy } = this.configuration;
        if (!strategies[activeStrategy]) {
            throw new Error(`applyActiveStrategyCallback: active strategy ${activeStrategy} not found, check tool configuration or spellings`);
        }
        return strategies[activeStrategy][callbackType]?.call(this, enabledElement, operationData, ...extraArgs);
    }
    setConfiguration(newConfiguration) {
        this.configuration = csUtils.deepMerge(this.configuration, newConfiguration);
    }
    setActiveStrategy(strategyName) {
        this.setConfiguration({ activeStrategy: strategyName });
    }
    getTargetImageData(targetId) {
        if (targetId.startsWith('imageId:')) {
            const imageId = targetId.split('imageId:')[1];
            const imageURI = csUtils.imageIdToURI(imageId);
            let viewports = csUtils.getViewportsWithImageURI(imageURI);
            if (!viewports || !viewports.length) {
                return;
            }
            viewports = viewports.filter((viewport) => {
                return viewport.getCurrentImageId() === imageId;
            });
            if (!viewports || !viewports.length) {
                return;
            }
            return viewports[0].getImageData();
        }
        else if (targetId.startsWith('volumeId:')) {
            const volumeId = csUtils.getVolumeId(targetId);
            const viewports = csUtils.getViewportsWithVolumeId(volumeId);
            if (!viewports || !viewports.length) {
                return;
            }
            return viewports[0].getImageData();
        }
        else if (targetId.startsWith('videoId:')) {
            const imageURI = csUtils.imageIdToURI(targetId);
            const viewports = csUtils.getViewportsWithImageURI(imageURI);
            if (!viewports || !viewports.length) {
                return;
            }
            return viewports[0].getImageData();
        }
        else {
            throw new Error('getTargetIdImage: targetId must start with "imageId:" or "volumeId:"');
        }
    }
    getTargetId(viewport, data) {
        const { isPreferredTargetId } = this.configurationTyped;
        if (isPreferredTargetId && data?.cachedStats) {
            for (const [targetId, cachedStat] of Object.entries(data.cachedStats)) {
                if (isPreferredTargetId(viewport, { targetId, cachedStat })) {
                    return targetId;
                }
            }
        }
        const defaultTargetId = viewport.getViewReferenceId?.();
        if (defaultTargetId) {
            return defaultTargetId;
        }
        throw new Error('getTargetId: viewport must have a getViewReferenceId method');
    }
    undo() {
        this.doneEditMemo();
        DefaultHistoryMemo.undo();
    }
    redo() {
        DefaultHistoryMemo.redo();
    }
    static createZoomPanMemo(viewport) {
        if (!viewportHasPan(viewport) || !viewportHasZoom(viewport)) {
            return;
        }
        const state = {
            pan: viewport.getPan(),
            zoom: viewport.getZoom(),
        };
        const zoomPanMemo = {
            restoreMemo: () => {
                const currentPan = viewport.getPan();
                const currentZoom = viewport.getZoom();
                const renderableViewport = viewport;
                viewport.setZoom(state.zoom);
                viewport.setPan(state.pan);
                if (typeof renderableViewport.render === 'function') {
                    renderableViewport.render();
                }
                state.pan = currentPan;
                state.zoom = currentZoom;
            },
        };
        DefaultHistoryMemo.push(zoomPanMemo);
        return zoomPanMemo;
    }
    doneEditMemo() {
        if (this.memo?.commitMemo?.()) {
            DefaultHistoryMemo.push(this.memo);
        }
        this.memo = null;
    }
    static startGroupRecording() {
        DefaultHistoryMemo.startGroupRecording();
    }
    static endGroupRecording() {
        DefaultHistoryMemo.endGroupRecording();
    }
    static calculateLengthInIndex(calibrate, indexPoints, closed = false) {
        const scale = calibrate?.scale || 1;
        const scaleY = calibrate?.scaleY || scale;
        const scaleZ = calibrate?.scaleZ || scale;
        let length = 0;
        const count = indexPoints.length;
        const start = closed ? 0 : 1;
        let lastPoint = closed ? indexPoints[count - 1] : indexPoints[0];
        for (let i = start; i < count; i++) {
            const point = indexPoints[i];
            const dx = (point[0] - lastPoint[0]) / scale;
            const dy = (point[1] - lastPoint[1]) / scaleY;
            const dz = (point[2] - lastPoint[2]) / scaleZ;
            length += Math.sqrt(dx * dx + dy * dy + dz * dz);
            lastPoint = point;
        }
        return length;
    }
    static isInsideVolume(dimensions, indexPoints) {
        const { length: count } = indexPoints;
        for (let i = 0; i < count; i++) {
            if (!csUtils.indexWithinDimensions(indexPoints[i], dimensions)) {
                return false;
            }
        }
        return true;
    }
}
BaseTool.defaults = {
    configuration: {
        strategies: {},
        defaultStrategy: undefined,
        activeStrategy: undefined,
        strategyOptions: {},
    },
};
BaseTool.toolName = 'BaseTool';
export default BaseTool;
