// @cornerstonejs/tools/dist/esm/utilities/genericViewportToolHelpers.js
import { CONSTANTS, utilities as csUtils, } from '@cornerstonejs/core';
import { getViewportPresentation } from './viewportPresentation';
const { RENDERING_DEFAULTS } = CONSTANTS;
export function getSlabThicknessOrDefault(viewport) {
    if (csUtils.isGenericViewport(viewport)) {
        return RENDERING_DEFAULTS.MINIMUM_SLAB_THICKNESS;
    }
    return viewport.getSlabThickness();
}
export function jumpToFocalPoint(viewport, cameraFocalPoint) {
    if (csUtils.isGenericViewport(viewport)) {
        viewport.setViewReference({ cameraFocalPoint });
    }
}
export function getNativeSourceProperties(viewport) {
    if (!csUtils.viewportSupportsDisplaySetPresentation(viewport)) {
        return { properties: {} };
    }
    const sourceDataId = viewport.getSourceDataId();
    const properties = {
        ...((sourceDataId
            ? viewport.getDisplaySetPresentation(sourceDataId)
            : {}) ?? {}),
    };
    const voiLUTFunction = properties.VOILUTFunction ?? properties.voiLUTFunction;
    if (voiLUTFunction !== undefined) {
        properties.VOILUTFunction = voiLUTFunction;
    }
    const presentation = (getViewportPresentation(viewport) ?? {});
    return {
        properties,
        rotation: presentation.rotation,
        flipHorizontal: presentation.flipHorizontal,
        flipVertical: presentation.flipVertical,
        currentImageId: viewport.getCurrentImageId(),
    };
}
