// @cornerstonejs/tools/dist/esm/tools/annotation/PlanarFreehandContourSegmentationTool.js
var _a;
import { utilities } from '@cornerstonejs/core';
import { triggerSegmentationDataModified } from '../../stateManagement/segmentation/triggerSegmentationEvents';
import PlanarFreehandROITool from './PlanarFreehandROITool';
import AnnotationToPointData from '../../utilities/contours/AnnotationToPointData';
export class PlanarFreehandContourSegmentationTool extends PlanarFreehandROITool {
    constructor(toolProps) {
        const initialProps = utilities.deepMerge({
            configuration: {
                calculateStats: false,
                allowOpenContours: false,
            },
        }, toolProps);
        super(initialProps);
    }
    isContourSegmentationTool() {
        return true;
    }
    renderAnnotationInstance(renderContext) {
        const annotation = renderContext.annotation;
        const { invalidated } = annotation;
        const renderResult = super.renderAnnotationInstance(renderContext);
        if (invalidated) {
            const { segmentationId } = annotation.data.segmentation;
            triggerSegmentationDataModified(segmentationId);
        }
        return renderResult;
    }
}
_a = PlanarFreehandContourSegmentationTool;
PlanarFreehandContourSegmentationTool.toolName = 'PlanarFreehandContourSegmentationTool';
(() => {
    AnnotationToPointData.register(_a);
})();
export default PlanarFreehandContourSegmentationTool;
