// @cornerstonejs/tools/dist/esm/tools/annotation/planarFreehandROITool/findOpenUShapedContourVectorToPeak.js
import { vec2 } from 'gl-matrix';
export default function findOpenUShapedContourVectorToPeak(canvasPoints, viewport) {
    const first = canvasPoints[0];
    const last = canvasPoints[canvasPoints.length - 1];
    const firstToLastUnitVector = vec2.create();
    vec2.set(firstToLastUnitVector, last[0] - first[0], last[1] - first[1]);
    vec2.normalize(firstToLastUnitVector, firstToLastUnitVector);
    const normalVector1 = vec2.create();
    const normalVector2 = vec2.create();
    vec2.set(normalVector1, -firstToLastUnitVector[1], firstToLastUnitVector[0]);
    vec2.set(normalVector2, firstToLastUnitVector[1], -firstToLastUnitVector[0]);
    const centerOfFirstToLast = [
        (first[0] + last[0]) / 2,
        (first[1] + last[1]) / 2,
    ];
    const furthest = {
        dist: 0,
        index: null,
    };
    for (let i = 0; i < canvasPoints.length; i++) {
        const canvasPoint = canvasPoints[i];
        const distance = vec2.dist(canvasPoint, centerOfFirstToLast);
        if (distance > furthest.dist) {
            furthest.dist = distance;
            furthest.index = i;
        }
    }
    const toFurthest = [
        canvasPoints[furthest.index],
        centerOfFirstToLast,
    ];
    const toFurthestWorld = toFurthest.map(viewport.canvasToWorld);
    return toFurthestWorld;
}
export function resolveVectorToPeak(canvasPoints, viewport, variant) {
    if (variant === 'orthogonalT') {
        return findOpenUShapedContourVectorToPeakOrthogonal(canvasPoints, viewport);
    }
    if (variant === 'lineSegment') {
        return null;
    }
    if (variant) {
        return findOpenUShapedContourVectorToPeak(canvasPoints, viewport);
    }
    return null;
}
export function resolveVectorToPeakOnRender(enabledElement, annotation) {
    const { viewport } = enabledElement;
    const canvasPoints = annotation.data.contour.polyline.map(viewport.worldToCanvas);
    return resolveVectorToPeak(canvasPoints, viewport, annotation.data.isOpenUShapeContour);
}
function findOpenUShapedContourVectorToPeakOrthogonal(canvasPoints, viewport) {
    const first = canvasPoints[0];
    const last = canvasPoints[canvasPoints.length - 1];
    const firstToLastUnitVector = vec2.sub(vec2.create(), last, first);
    vec2.normalize(firstToLastUnitVector, firstToLastUnitVector);
    const chordDir = [
        firstToLastUnitVector[0],
        firstToLastUnitVector[1],
    ];
    const centerOfFirstToLast = [
        (first[0] + last[0]) / 2,
        (first[1] + last[1]) / 2,
    ];
    const delta = vec2.create();
    let prevDp = null;
    let prevPoint = null;
    let orthogonalPoint = null;
    for (const p of canvasPoints) {
        vec2.sub(delta, p, centerOfFirstToLast);
        const dp = vec2.dot(chordDir, delta);
        if (prevDp !== null && prevDp * dp < 0) {
            const t = Math.abs(prevDp) / (Math.abs(prevDp) + Math.abs(dp));
            orthogonalPoint = [
                prevPoint[0] + t * (p[0] - prevPoint[0]),
                prevPoint[1] + t * (p[1] - prevPoint[1]),
            ];
            break;
        }
        if (Math.abs(dp) < 1e-10) {
            orthogonalPoint = p;
            break;
        }
        prevDp = dp;
        prevPoint = p;
    }
    if (!orthogonalPoint) {
        console.warn('No orthogonal intersection found for open U-shaped contour');
        return null;
    }
    const toOrthogonal = [
        orthogonalPoint,
        centerOfFirstToLast,
    ];
    const toOrthogonalWorld = toOrthogonal.map(viewport.canvasToWorld);
    return toOrthogonalWorld;
}
