// @cornerstonejs/tools/dist/esm/tools/SculptorTool/CircleSculptCursor.js
import { vec2, vec3 } from 'gl-matrix';
import { getEnabledElement } from '@cornerstonejs/core';
import { distancePointToContour } from '../distancePointToContour';
import { drawCircle as drawCircleSvg } from '../../drawingSvg';
import { point } from '../../utilities/math';
class CircleSculptCursor {
    constructor() {
        this.toolInfo = {
            toolSize: null,
            radius: null,
            maxToolSize: null,
        };
    }
    renderShape(svgDrawingHelper, canvasLocation, options) {
        const circleUID = '0';
        drawCircleSvg(svgDrawingHelper, 'SculptorTool', circleUID, canvasLocation, this.toolInfo.toolSize, options);
    }
    configureToolSize(evt) {
        const toolInfo = this.toolInfo;
        if (toolInfo.toolSize && toolInfo.maxToolSize) {
            return;
        }
        const eventData = evt.detail;
        const element = eventData.element;
        const minDim = Math.min(element.clientWidth, element.clientHeight);
        const maxRadius = minDim / 24;
        toolInfo.toolSize = maxRadius;
        toolInfo.radius = null;
        toolInfo.maxToolSize = maxRadius;
    }
    updateToolSize(canvasCoords, viewport, activeAnnotation) {
        const toolInfo = this.toolInfo;
        const radius = distancePointToContour(viewport, activeAnnotation, canvasCoords);
        if (radius > 0) {
            toolInfo.toolSize = Math.min(toolInfo.maxToolSize, radius);
            this.computeWorldRadius(viewport, true);
        }
    }
    getMaxSpacing(minSpacing) {
        return Math.max(this.toolInfo.toolSize / 4, minSpacing);
    }
    computeWorldRadius(viewport, clearExisting = false) {
        if (!this.toolInfo.radius || clearExisting) {
            const p0 = viewport.canvasToWorld([0, 0]);
            const p1 = viewport.canvasToWorld([this.toolInfo.toolSize, 0]);
            this.toolInfo.radius = vec3.length(vec3.sub(vec3.create(), p0, p1));
        }
        return this.toolInfo.radius;
    }
    getEdge(viewport, p1, p2, mouseCanvas) {
        const midPoint = vec3.add(vec3.create(), p1, p2 || p1);
        vec3.scale(midPoint, midPoint, 0.5);
        const canvasMidPoint = viewport.worldToCanvas(midPoint);
        const canvasDelta = vec2.sub(vec2.create(), canvasMidPoint, mouseCanvas);
        const angle = Math.atan2(canvasDelta[1], canvasDelta[0]);
        const point = this.interpolatePoint(viewport, angle, mouseCanvas);
        const canvasPoint = viewport.worldToCanvas(point);
        return {
            point,
            angle,
            canvasPoint,
        };
    }
    interpolatePoint(viewport, angle, center) {
        const [cx, cy] = center;
        const r = this.toolInfo.toolSize;
        const dx = Math.cos(angle) * r;
        const dy = Math.sin(angle) * r;
        const newPoint2 = [cx + dx, cy + dy];
        return viewport.canvasToWorld(newPoint2);
    }
    isInCursor(point, mousePoint) {
        return vec3.distance(point, mousePoint) < this.toolInfo.radius;
    }
}
CircleSculptCursor.shapeName = 'Circle';
export default CircleSculptCursor;
