// @cornerstonejs/tools/dist/esm/tools/CrosshairsTool.js
import { vec2, vec3 } from 'gl-matrix';
import vtkMath from '@kitware/vtk.js/Common/Core/Math';
import vtkMatrixBuilder from '@kitware/vtk.js/Common/Core/MatrixBuilder';
import { AnnotationTool } from './base';
import { getRenderingEngine } from '@cornerstonejs/core';
import { getEnabledElementByIds, getEnabledElement, utilities as csUtils, Enums, CONSTANTS, triggerEvent, eventTarget, } from '@cornerstonejs/core';
import { getToolGroup, getToolGroupForViewport, } from '../store/ToolGroupManager';
import { addAnnotation, getAnnotations, removeAnnotation, } from '../stateManagement/annotation/annotationState';
import { drawCircle as drawCircleSvg, drawHandles as drawHandlesSvg, drawLine as drawLineSvg, } from '../drawingSvg';
import { state } from '../store/state';
import { Events } from '../enums';
import { getViewportIdsWithToolToRender } from '../utilities/viewportFilters';
import getViewportICamera from '../utilities/getViewportICamera';
import { resetElementCursor, hideElementCursor, } from '../cursors/elementCursor';
import liangBarksyClip from '../utilities/math/vec2/liangBarksyClip';
import { getSlabThicknessOrDefault, jumpToFocalPoint, } from '../utilities/genericViewportToolHelpers';
import * as lineSegment from '../utilities/math/line';
import { isAnnotationLocked } from '../stateManagement/annotation/annotationLocking';
import triggerAnnotationRenderForViewportIds from '../utilities/triggerAnnotationRenderForViewportIds';
const { RENDERING_DEFAULTS } = CONSTANTS;
function getDisplayedCanvasSize(viewport) {
    if (csUtils.isGenericViewport(viewport)) {
        const { element } = viewport;
        return {
            clientWidth: element.clientWidth,
            clientHeight: element.clientHeight,
        };
    }
    const { clientWidth, clientHeight } = viewport.canvas;
    return { clientWidth, clientHeight };
}
function defaultReferenceLineColor() {
    return 'rgb(0, 200, 0)';
}
function defaultReferenceLineControllable() {
    return true;
}
function defaultReferenceLineDraggableRotatable() {
    return true;
}
function defaultReferenceLineSlabThicknessControlsOn() {
    return true;
}
const OPERATION = {
    DRAG: 1,
    ROTATE: 2,
    SLAB: 3,
};
class CrosshairsTool extends AnnotationTool {
    constructor(toolProps = {}, defaultToolProps = {
        supportedInteractionTypes: ['Mouse'],
        configuration: {
            shadow: true,
            viewportIndicators: false,
            viewportIndicatorsConfig: {
                radius: 5,
                x: null,
                y: null,
            },
            autoPan: {
                enabled: false,
                panSize: 10,
            },
            handleRadius: 3,
            enableHDPIHandles: false,
            referenceLinesCenterGapRadius: 20,
            referenceLinesCenterGapRatio: null,
            minimal: {
                enabled: false,
                lineLengthInPx: 40,
            },
            filterActorUIDsToSetSlabThickness: [],
            slabThicknessBlendMode: Enums.BlendModes.MAXIMUM_INTENSITY_BLEND,
            centerPoint: {
                enabled: false,
                color: 'rgba(255, 255, 0, 0.5)',
                size: 2,
            },
            mobile: {
                enabled: false,
                opacity: 0.8,
                handleRadius: 9,
                referenceLinesCenterGapRatio: 0.05,
            },
        },
    }) {
        super(toolProps, defaultToolProps);
        this.toolCenter = [0, 0, 0];
        this._volumeViewportNewVolumeListeners = new Map();
        this._toolGroupViewportAddedListener = null;
        this._toolGroupViewportRemovedListener = null;
        this._ignoreFiredEvents = false;
        this.initializeViewport = ({ renderingEngineId, viewportId, }) => {
            const enabledElement = getEnabledElementByIds(viewportId, renderingEngineId);
            if (!enabledElement) {
                return;
            }
            const { FrameOfReferenceUID, viewport } = enabledElement;
            const { element } = viewport;
            const { position, focalPoint, viewPlaneNormal } = getViewportICamera(viewport);
            let annotations = this._getAnnotations(enabledElement);
            annotations = this.filterInteractableAnnotationsForElement(element, annotations);
            if (annotations?.length) {
                removeAnnotation(annotations[0].annotationUID);
            }
            const annotation = {
                highlighted: false,
                metadata: {
                    cameraPosition: [...position],
                    cameraFocalPoint: [...focalPoint],
                    FrameOfReferenceUID,
                    toolName: this.getToolName(),
                },
                data: {
                    handles: {
                        rotationPoints: [],
                        slabThicknessPoints: [],
                        toolCenter: this.toolCenter,
                    },
                    activeOperation: null,
                    activeViewportIds: [],
                    viewportId,
                },
            };
            addAnnotation(annotation, element);
            const { clientWidth, clientHeight } = getDisplayedCanvasSize(viewport);
            return {
                normal: viewPlaneNormal,
                point: viewport.canvasToWorld([clientWidth / 2, clientHeight / 2]),
            };
        };
        this._getViewportsInfo = () => {
            const viewports = getToolGroup(this.toolGroupId).viewportsInfo;
            return viewports;
        };
        this._reinitializeListenersAndCenter = () => {
            this._unbindToolGroupViewportListeners();
            this._clearAllVolumeListenersAndViewportState();
            this._bindToolGroupViewportListeners();
            this._syncVolumeListenersWithToolGroup();
            this._computeToolCenter(this._getViewportsInfo());
        };
        this.resetCrosshairs = () => {
            const viewportsInfo = this._getViewportsInfo();
            for (const viewportInfo of viewportsInfo) {
                const { viewportId, renderingEngineId } = viewportInfo;
                const enabledElement = getEnabledElementByIds(viewportId, renderingEngineId);
                if (!enabledElement) {
                    continue;
                }
                const viewport = enabledElement.viewport;
                const resetPan = true;
                const resetZoom = true;
                const resetToCenter = true;
                const resetRotation = true;
                const suppressEvents = true;
                if (csUtils.isGenericViewport(viewport)) {
                    viewport.resetViewState({
                        resetPan,
                        resetZoom,
                        resetOrientation: resetRotation,
                        resetFlip: true,
                    });
                }
                else {
                    viewport.resetCamera({
                        resetPan,
                        resetZoom,
                        resetToCenter,
                        resetRotation,
                        suppressEvents,
                    });
                    viewport.resetSlabThickness();
                }
                const { element } = viewport;
                let annotations = this._getAnnotations(enabledElement);
                annotations = this.filterInteractableAnnotationsForElement(element, annotations);
                if (annotations.length) {
                    removeAnnotation(annotations[0].annotationUID);
                }
                viewport.render();
            }
            this._computeToolCenter(viewportsInfo);
        };
        this.computeToolCenter = () => {
            const viewportsInfo = this._getViewportsInfo();
            this._computeToolCenter(viewportsInfo);
        };
        this._computeToolCenter = (viewportsInfo) => {
            if (!viewportsInfo.length || viewportsInfo.length === 1) {
                console.warn('For crosshairs to operate, at least two viewports must be given.');
                return;
            }
            viewportsInfo.forEach((viewportInfo) => {
                this.initializeViewport(viewportInfo);
            });
            this._recomputeToolCenterFromAbsoluteCameras({
                emitEvent: true,
                updateViewportCameras: true,
            });
        };
        this.addNewAnnotation = (evt) => {
            const eventDetail = evt.detail;
            const { element } = eventDetail;
            const { currentPoints } = eventDetail;
            const jumpWorld = currentPoints.world;
            this._syncVolumeListenersWithToolGroup();
            this._recomputeToolCenterFromAbsoluteCameras({
                emitEvent: false,
                updateViewportCameras: false,
            });
            const enabledElement = getEnabledElement(element);
            const { viewport } = enabledElement;
            this._jump(enabledElement, jumpWorld);
            const annotations = this._getAnnotations(enabledElement);
            const filteredAnnotations = this.filterInteractableAnnotationsForElement(viewport.element, annotations);
            const { data } = filteredAnnotations[0];
            const { rotationPoints } = data.handles;
            const viewportIdArray = [];
            for (let i = 0; i < rotationPoints.length - 1; ++i) {
                const otherViewport = rotationPoints[i][1];
                const viewportControllable = this._getReferenceLineControllable(otherViewport.id);
                const viewportDraggableRotatable = this._getReferenceLineDraggableRotatable(otherViewport.id);
                if (!viewportControllable || !viewportDraggableRotatable) {
                    continue;
                }
                viewportIdArray.push(otherViewport.id);
                i++;
            }
            data.activeViewportIds = [...viewportIdArray];
            data.handles.activeOperation = OPERATION.DRAG;
            evt.preventDefault();
            hideElementCursor(element);
            this._activateModify(element);
            return filteredAnnotations[0];
        };
        this.cancel = () => {
            console.log('Not implemented yet');
        };
        this.handleSelectedCallback = (evt, annotation) => {
            const eventDetail = evt.detail;
            const { element } = eventDetail;
            annotation.highlighted = true;
            this._activateModify(element);
            hideElementCursor(element);
            evt.preventDefault();
        };
        this.isPointNearTool = (element, annotation, canvasCoords, proximity) => {
            if (this._pointNearTool(element, annotation, canvasCoords, 6)) {
                return true;
            }
            return false;
        };
        this.toolSelectedCallback = (evt, annotation, interactionType) => {
            const eventDetail = evt.detail;
            const { element } = eventDetail;
            annotation.highlighted = true;
            this._activateModify(element);
            hideElementCursor(element);
            evt.preventDefault();
        };
        this.onCameraModified = (evt) => {
            const eventDetail = evt.detail;
            const { element } = eventDetail;
            const enabledElement = getEnabledElement(element);
            if (!enabledElement) {
                return;
            }
            const { renderingEngine } = enabledElement;
            const viewport = enabledElement.viewport;
            this._syncVolumeListenersWithToolGroup();
            if (this._ignoreFiredEvents) {
                return;
            }
            const isSourceInToolGroup = this._getViewportsInfo().some(({ viewportId, renderingEngineId }) => viewportId === viewport.id && renderingEngineId === renderingEngine.id);
            if (!isSourceInToolGroup) {
                return;
            }
            const annotations = this._getAnnotations(enabledElement);
            const filteredToolAnnotations = this.filterInteractableAnnotationsForElement(element, annotations);
            const viewportAnnotation = filteredToolAnnotations[0];
            const currentCamera = getViewportICamera(viewport);
            if (viewportAnnotation) {
                viewportAnnotation.metadata.cameraPosition = [...currentCamera.position];
                viewportAnnotation.metadata.cameraFocalPoint = [
                    ...currentCamera.focalPoint,
                ];
            }
            this._recomputeToolCenterFromAbsoluteCameras({
                emitEvent: true,
                updateViewportCameras: false,
            });
            if (this.configuration.autoPan?.enabled) {
                const toolGroup = getToolGroupForViewport(viewport.id, renderingEngine.id);
                const otherViewportIds = toolGroup
                    .getViewportIds()
                    .filter((id) => id !== viewport.id);
                otherViewportIds.forEach((viewportId) => {
                    this._autoPanViewportIfNecessary(viewportId, renderingEngine);
                });
            }
            const requireSameOrientation = false;
            const viewportIdsToRender = getViewportIdsWithToolToRender(element, this.getToolName(), requireSameOrientation);
            triggerAnnotationRenderForViewportIds(viewportIdsToRender);
        };
        this.onResetCamera = (evt) => {
            this.resetCrosshairs();
        };
        this.mouseMoveCallback = (evt, filteredToolAnnotations) => {
            const { element, currentPoints } = evt.detail;
            const canvasCoords = currentPoints.canvas;
            let imageNeedsUpdate = false;
            for (let i = 0; i < filteredToolAnnotations.length; i++) {
                const annotation = filteredToolAnnotations[i];
                if (isAnnotationLocked(annotation.annotationUID)) {
                    continue;
                }
                const { data, highlighted } = annotation;
                if (!data.handles) {
                    continue;
                }
                const previousActiveOperation = data.handles.activeOperation;
                const previousActiveViewportIds = data.activeViewportIds && data.activeViewportIds.length > 0
                    ? [...data.activeViewportIds]
                    : [];
                data.activeViewportIds = [];
                data.handles.activeOperation = null;
                const handleNearImagePoint = this.getHandleNearImagePoint(element, annotation, canvasCoords, 6);
                let near = false;
                if (handleNearImagePoint) {
                    near = true;
                }
                else {
                    near = this._pointNearTool(element, annotation, canvasCoords, 6);
                }
                const nearToolAndNotMarkedActive = near && !highlighted;
                const notNearToolAndMarkedActive = !near && highlighted;
                if (nearToolAndNotMarkedActive || notNearToolAndMarkedActive) {
                    annotation.highlighted = !highlighted;
                    imageNeedsUpdate = true;
                }
                else if (data.handles.activeOperation !== previousActiveOperation ||
                    !this._areViewportIdArraysEqual(data.activeViewportIds, previousActiveViewportIds)) {
                    imageNeedsUpdate = true;
                }
            }
            return imageNeedsUpdate;
        };
        this.filterInteractableAnnotationsForElement = (element, annotations) => {
            if (!annotations || !annotations.length) {
                return [];
            }
            const enabledElement = getEnabledElement(element);
            const { viewportId } = enabledElement;
            const viewportUIDSpecificCrosshairs = annotations.filter((annotation) => annotation.data.viewportId === viewportId);
            return viewportUIDSpecificCrosshairs;
        };
        this.renderAnnotation = (enabledElement, svgDrawingHelper) => {
            let renderStatus = false;
            const { viewport, renderingEngine } = enabledElement;
            const { element } = viewport;
            const annotations = this._getAnnotations(enabledElement);
            const camera = getViewportICamera(viewport);
            const filteredToolAnnotations = this.filterInteractableAnnotationsForElement(element, annotations);
            const viewportAnnotation = filteredToolAnnotations[0];
            if (!annotations?.length || !viewportAnnotation?.data) {
                return renderStatus;
            }
            const annotationUID = viewportAnnotation.annotationUID;
            const { clientWidth, clientHeight } = getDisplayedCanvasSize(viewport);
            const canvasDiagonalLength = Math.sqrt(clientWidth * clientWidth + clientHeight * clientHeight);
            const canvasMinDimensionLength = Math.min(clientWidth, clientHeight);
            const data = viewportAnnotation.data;
            const crosshairCenterCanvas = viewport.worldToCanvas(this.toolCenter);
            const otherViewportAnnotations = this._filterAnnotationsByUniqueViewportOrientations(enabledElement, annotations);
            const referenceLines = [];
            const canvasBox = [0, 0, clientWidth, clientHeight];
            otherViewportAnnotations.forEach((annotation) => {
                const { data } = annotation;
                data.handles.toolCenter = this.toolCenter;
                const otherViewport = renderingEngine.getViewport(data.viewportId);
                const otherCamera = getViewportICamera(otherViewport);
                const otherViewportControllable = this._getReferenceLineControllable(otherViewport.id);
                const otherViewportDraggableRotatable = this._getReferenceLineDraggableRotatable(otherViewport.id);
                const otherViewportSlabThicknessControlsOn = this._getReferenceLineSlabThicknessControlsOn(otherViewport.id);
                const { clientWidth, clientHeight } = getDisplayedCanvasSize(otherViewport);
                const otherCanvasDiagonalLength = Math.sqrt(clientWidth * clientWidth + clientHeight * clientHeight);
                const otherCanvasCenter = [
                    clientWidth * 0.5,
                    clientHeight * 0.5,
                ];
                const otherViewportCenterWorld = otherViewport.canvasToWorld(otherCanvasCenter);
                const direction = [0, 0, 0];
                vtkMath.cross(camera.viewPlaneNormal, otherCamera.viewPlaneNormal, direction);
                vtkMath.normalize(direction);
                vtkMath.multiplyScalar(direction, otherCanvasDiagonalLength);
                const pointWorld0 = [0, 0, 0];
                vtkMath.add(otherViewportCenterWorld, direction, pointWorld0);
                const pointWorld1 = [0, 0, 0];
                vtkMath.subtract(otherViewportCenterWorld, direction, pointWorld1);
                const pointCanvas0 = viewport.worldToCanvas(pointWorld0);
                const otherViewportCenterCanvas = viewport.worldToCanvas(otherViewportCenterWorld);
                const canvasUnitVectorFromCenter = vec2.create();
                vec2.subtract(canvasUnitVectorFromCenter, pointCanvas0, otherViewportCenterCanvas);
                vec2.normalize(canvasUnitVectorFromCenter, canvasUnitVectorFromCenter);
                const canvasVectorFromCenterLong = vec2.create();
                vec2.scale(canvasVectorFromCenterLong, canvasUnitVectorFromCenter, canvasDiagonalLength * 100);
                const canvasVectorFromCenterShort = vec2.create();
                vec2.scale(canvasVectorFromCenterShort, canvasUnitVectorFromCenter, canvasMinDimensionLength * 0.2);
                const canvasVectorFromCenterStart = vec2.create();
                const mobileConfig = this.configuration.mobile;
                const activeConfiguration = mobileConfig?.enabled
                    ? mobileConfig
                    : this.configuration;
                const { referenceLinesCenterGapRatio } = activeConfiguration;
                const minimalCrosshairConfig = getMinimalCrosshairConfig(this.configuration);
                const centerGap = referenceLinesCenterGapRatio > 0
                    ? canvasMinDimensionLength * referenceLinesCenterGapRatio
                    : this.configuration.referenceLinesCenterGapRadius;
                vec2.scale(canvasVectorFromCenterStart, canvasUnitVectorFromCenter, otherViewportAnnotations.length === 2 ? centerGap : 0);
                const refLinePointOne = vec2.create();
                const refLinePointTwo = vec2.create();
                const refLinePointThree = vec2.create();
                const refLinePointFour = vec2.create();
                let refLinesCenter = vec2.clone(crosshairCenterCanvas);
                if (!otherViewportDraggableRotatable || !otherViewportControllable) {
                    refLinesCenter = vec2.clone(otherViewportCenterCanvas);
                }
                vec2.add(refLinePointOne, refLinesCenter, canvasVectorFromCenterStart);
                vec2.add(refLinePointTwo, refLinesCenter, canvasVectorFromCenterLong);
                vec2.subtract(refLinePointThree, refLinesCenter, canvasVectorFromCenterStart);
                vec2.subtract(refLinePointFour, refLinesCenter, canvasVectorFromCenterLong);
                liangBarksyClip(refLinePointOne, refLinePointTwo, canvasBox);
                liangBarksyClip(refLinePointThree, refLinePointFour, canvasBox);
                if (minimalCrosshairConfig.enabled) {
                    const minimalCanvasVectorFromCenterLong = vec2.create();
                    vec2.scale(minimalCanvasVectorFromCenterLong, canvasUnitVectorFromCenter, minimalCrosshairConfig.lineLengthInPx);
                    vec2.add(refLinePointOne, refLinesCenter, canvasVectorFromCenterStart);
                    vec2.add(refLinePointTwo, refLinePointOne, minimalCanvasVectorFromCenterLong);
                    vec2.subtract(refLinePointThree, refLinesCenter, canvasVectorFromCenterStart);
                    vec2.subtract(refLinePointFour, refLinePointThree, minimalCanvasVectorFromCenterLong);
                    liangBarksyClip(refLinePointOne, refLinePointTwo, canvasBox);
                    liangBarksyClip(refLinePointThree, refLinePointFour, canvasBox);
                }
                const rotHandleOne = getSegmentMidpoint(refLinePointOne, refLinePointTwo);
                const rotHandleTwo = getSegmentMidpoint(refLinePointThree, refLinePointFour);
                let stHandlesCenterCanvas = vec2.clone(crosshairCenterCanvas);
                if (!otherViewportDraggableRotatable &&
                    otherViewportSlabThicknessControlsOn) {
                    stHandlesCenterCanvas = vec2.clone(otherViewportCenterCanvas);
                }
                let stHandlesCenterWorld = [...this.toolCenter];
                if (!otherViewportDraggableRotatable &&
                    otherViewportSlabThicknessControlsOn) {
                    stHandlesCenterWorld = [...otherViewportCenterWorld];
                }
                const worldUnitVectorFromCenter = [0, 0, 0];
                vtkMath.subtract(pointWorld0, pointWorld1, worldUnitVectorFromCenter);
                vtkMath.normalize(worldUnitVectorFromCenter);
                const { viewPlaneNormal } = camera;
                const { matrix } = vtkMatrixBuilder
                    .buildFromDegree()
                    .rotate(90, viewPlaneNormal);
                const worldUnitOrthoVectorFromCenter = [0, 0, 0];
                vec3.transformMat4(worldUnitOrthoVectorFromCenter, worldUnitVectorFromCenter, matrix);
                const slabThicknessValue = getSlabThicknessOrDefault(otherViewport);
                const worldOrthoVectorFromCenter = [
                    ...worldUnitOrthoVectorFromCenter,
                ];
                vtkMath.multiplyScalar(worldOrthoVectorFromCenter, slabThicknessValue);
                const worldVerticalRefPoint = [0, 0, 0];
                vtkMath.add(stHandlesCenterWorld, worldOrthoVectorFromCenter, worldVerticalRefPoint);
                const canvasVerticalRefPoint = viewport.worldToCanvas(worldVerticalRefPoint);
                const canvasOrthoVectorFromCenter = vec2.create();
                vec2.subtract(canvasOrthoVectorFromCenter, stHandlesCenterCanvas, canvasVerticalRefPoint);
                const stLinePointOne = vec2.create();
                vec2.subtract(stLinePointOne, stHandlesCenterCanvas, canvasVectorFromCenterLong);
                vec2.add(stLinePointOne, stLinePointOne, canvasOrthoVectorFromCenter);
                const stLinePointTwo = vec2.create();
                vec2.add(stLinePointTwo, stHandlesCenterCanvas, canvasVectorFromCenterLong);
                vec2.add(stLinePointTwo, stLinePointTwo, canvasOrthoVectorFromCenter);
                liangBarksyClip(stLinePointOne, stLinePointTwo, canvasBox);
                const stLinePointThree = vec2.create();
                vec2.add(stLinePointThree, stHandlesCenterCanvas, canvasVectorFromCenterLong);
                vec2.subtract(stLinePointThree, stLinePointThree, canvasOrthoVectorFromCenter);
                const stLinePointFour = vec2.create();
                vec2.subtract(stLinePointFour, stHandlesCenterCanvas, canvasVectorFromCenterLong);
                vec2.subtract(stLinePointFour, stLinePointFour, canvasOrthoVectorFromCenter);
                liangBarksyClip(stLinePointThree, stLinePointFour, canvasBox);
                const stHandleOne = vec2.create();
                const stHandleTwo = vec2.create();
                const stHandleThree = vec2.create();
                const stHandleFour = vec2.create();
                vec2.subtract(stHandleOne, stHandlesCenterCanvas, canvasVectorFromCenterShort);
                vec2.add(stHandleOne, stHandleOne, canvasOrthoVectorFromCenter);
                vec2.add(stHandleTwo, stHandlesCenterCanvas, canvasVectorFromCenterShort);
                vec2.add(stHandleTwo, stHandleTwo, canvasOrthoVectorFromCenter);
                vec2.subtract(stHandleThree, stHandlesCenterCanvas, canvasVectorFromCenterShort);
                vec2.subtract(stHandleThree, stHandleThree, canvasOrthoVectorFromCenter);
                vec2.add(stHandleFour, stHandlesCenterCanvas, canvasVectorFromCenterShort);
                vec2.subtract(stHandleFour, stHandleFour, canvasOrthoVectorFromCenter);
                referenceLines.push([
                    otherViewport,
                    refLinePointOne,
                    refLinePointTwo,
                    refLinePointThree,
                    refLinePointFour,
                    stLinePointOne,
                    stLinePointTwo,
                    stLinePointThree,
                    stLinePointFour,
                    rotHandleOne,
                    rotHandleTwo,
                    stHandleOne,
                    stHandleTwo,
                    stHandleThree,
                    stHandleFour,
                ]);
            });
            const newRtpoints = [];
            const newStpoints = [];
            const viewportColor = this._getReferenceLineColor(viewport.id);
            const color = viewportColor !== undefined ? viewportColor : 'rgb(200, 200, 200)';
            referenceLines.forEach((line, lineIndex) => {
                const otherViewport = line[0];
                const minimalCrosshairConfig = getMinimalCrosshairConfig(this.configuration);
                const viewportColor = this._getReferenceLineColor(otherViewport.id);
                const viewportControllable = this._getReferenceLineControllable(otherViewport.id);
                const viewportDraggableRotatable = !minimalCrosshairConfig.enabled &&
                    (this._getReferenceLineDraggableRotatable(otherViewport.id) ||
                        this.configuration.mobile?.enabled);
                const viewportSlabThicknessControlsOn = !minimalCrosshairConfig.enabled &&
                    (this._getReferenceLineSlabThicknessControlsOn(otherViewport.id) ||
                        this.configuration.mobile?.enabled);
                const selectedViewportId = data.activeViewportIds.find((id) => id === otherViewport.id);
                let color = viewportColor !== undefined ? viewportColor : 'rgb(200, 200, 200)';
                let lineWidth = 1;
                const lineActive = data.handles.activeOperation !== null &&
                    data.handles.activeOperation === OPERATION.DRAG &&
                    selectedViewportId;
                if (lineActive) {
                    lineWidth = 2.5;
                }
                let lineUID = `${lineIndex}`;
                if (viewportControllable &&
                    (viewportDraggableRotatable || minimalCrosshairConfig.enabled)) {
                    lineUID = `${lineIndex}One`;
                    drawLineSvg(svgDrawingHelper, annotationUID, lineUID, line[1], line[2], {
                        color,
                        lineWidth,
                    });
                    lineUID = `${lineIndex}Two`;
                    drawLineSvg(svgDrawingHelper, annotationUID, lineUID, line[3], line[4], {
                        color,
                        lineWidth,
                    });
                }
                else {
                    drawLineSvg(svgDrawingHelper, annotationUID, lineUID, line[2], line[4], {
                        color,
                        lineWidth,
                    });
                }
                if (viewportControllable) {
                    color =
                        viewportColor !== undefined ? viewportColor : 'rgb(200, 200, 200)';
                    const rotHandlesActive = data.handles.activeOperation === OPERATION.ROTATE;
                    const rotationHandles = [line[9], line[10]];
                    const rotHandleWorldOne = [
                        viewport.canvasToWorld(line[9]),
                        otherViewport,
                        line[1],
                        line[2],
                    ];
                    const rotHandleWorldTwo = [
                        viewport.canvasToWorld(line[10]),
                        otherViewport,
                        line[3],
                        line[4],
                    ];
                    newRtpoints.push(rotHandleWorldOne, rotHandleWorldTwo);
                    const slabThicknessHandlesActive = data.handles.activeOperation === OPERATION.SLAB;
                    const slabThicknessHandles = [line[11], line[12], line[13], line[14]];
                    if (!minimalCrosshairConfig.enabled) {
                        const slabThicknessHandleWorldOne = [
                            viewport.canvasToWorld(line[11]),
                            otherViewport,
                            line[5],
                            line[6],
                        ];
                        const slabThicknessHandleWorldTwo = [
                            viewport.canvasToWorld(line[12]),
                            otherViewport,
                            line[5],
                            line[6],
                        ];
                        const slabThicknessHandleWorldThree = [
                            viewport.canvasToWorld(line[13]),
                            otherViewport,
                            line[7],
                            line[8],
                        ];
                        const slabThicknessHandleWorldFour = [
                            viewport.canvasToWorld(line[14]),
                            otherViewport,
                            line[7],
                            line[8],
                        ];
                        newStpoints.push(slabThicknessHandleWorldOne, slabThicknessHandleWorldTwo, slabThicknessHandleWorldThree, slabThicknessHandleWorldFour);
                    }
                    let handleRadius = this.configuration.handleRadius *
                        (this.configuration.enableHDPIHandles ? window.devicePixelRatio : 1);
                    let opacity = 1;
                    if (this.configuration.mobile?.enabled) {
                        handleRadius = this.configuration.mobile.handleRadius;
                        opacity = this.configuration.mobile.opacity;
                    }
                    if ((lineActive || this.configuration.mobile?.enabled) &&
                        !rotHandlesActive &&
                        !slabThicknessHandlesActive &&
                        viewportDraggableRotatable &&
                        viewportSlabThicknessControlsOn) {
                        let handleUID = `${lineIndex}One`;
                        drawHandlesSvg(svgDrawingHelper, annotationUID, handleUID, rotationHandles, {
                            color,
                            handleRadius,
                            opacity,
                            type: 'circle',
                        });
                        handleUID = `${lineIndex}Two`;
                        drawHandlesSvg(svgDrawingHelper, annotationUID, handleUID, slabThicknessHandles, {
                            color,
                            handleRadius,
                            opacity,
                            type: 'rect',
                        });
                    }
                    else if (lineActive &&
                        !rotHandlesActive &&
                        !slabThicknessHandlesActive &&
                        viewportDraggableRotatable) {
                        const handleUID = `${lineIndex}`;
                        drawHandlesSvg(svgDrawingHelper, annotationUID, handleUID, rotationHandles, {
                            color,
                            handleRadius,
                            opacity,
                            type: 'circle',
                        });
                    }
                    else if (selectedViewportId &&
                        !rotHandlesActive &&
                        !slabThicknessHandlesActive &&
                        viewportSlabThicknessControlsOn) {
                        const handleUID = `${lineIndex}`;
                        drawHandlesSvg(svgDrawingHelper, annotationUID, handleUID, slabThicknessHandles, {
                            color,
                            handleRadius,
                            opacity,
                            type: 'rect',
                        });
                    }
                    else if (rotHandlesActive && viewportDraggableRotatable) {
                        const handleUID = `${lineIndex}`;
                        const handleRadius = this.configuration.handleRadius *
                            (this.configuration.enableHDPIHandles
                                ? window.devicePixelRatio
                                : 1);
                        drawHandlesSvg(svgDrawingHelper, annotationUID, handleUID, rotationHandles, {
                            color,
                            handleRadius,
                            fill: color,
                            type: 'circle',
                        });
                    }
                    else if (slabThicknessHandlesActive &&
                        selectedViewportId &&
                        viewportSlabThicknessControlsOn) {
                        const handleRadius = this.configuration.handleRadius *
                            (this.configuration.enableHDPIHandles
                                ? window.devicePixelRatio
                                : 1);
                        drawHandlesSvg(svgDrawingHelper, annotationUID, lineUID, slabThicknessHandles, {
                            color,
                            handleRadius,
                            fill: color,
                            type: 'rect',
                        });
                    }
                    const slabThicknessValue = getSlabThicknessOrDefault(otherViewport);
                    if (slabThicknessValue > 0.5 && viewportSlabThicknessControlsOn) {
                        lineUID = `${lineIndex}STOne`;
                        drawLineSvg(svgDrawingHelper, annotationUID, lineUID, line[5], line[6], {
                            color,
                            width: 1,
                            lineDash: [2, 3],
                        });
                        lineUID = `${lineIndex}STTwo`;
                        drawLineSvg(svgDrawingHelper, annotationUID, lineUID, line[7], line[8], {
                            color,
                            width: line,
                            lineDash: [2, 3],
                        });
                    }
                }
            });
            renderStatus = true;
            data.handles.rotationPoints = newRtpoints;
            data.handles.slabThicknessPoints = newStpoints;
            if (this.configuration.viewportIndicators) {
                const { viewportIndicatorsConfig } = this.configuration;
                const xOffset = viewportIndicatorsConfig?.xOffset || 0.95;
                const yOffset = viewportIndicatorsConfig?.yOffset || 0.05;
                const referenceColorCoordinates = [
                    clientWidth * xOffset,
                    clientHeight * yOffset,
                ];
                const circleRadius = viewportIndicatorsConfig?.circleRadius || canvasDiagonalLength * 0.01;
                const circleUID = '0';
                drawCircleSvg(svgDrawingHelper, annotationUID, circleUID, referenceColorCoordinates, circleRadius, { color, fill: color });
            }
            if (this.configuration.centerPoint?.enabled) {
                const defaultColor = 'rgba(255, 255, 0, 0.5)';
                const defaultSize = 2;
                const maxAllowedSize = 5;
                const centerPointColor = this.configuration.centerPoint.color || defaultColor;
                const centerPointSize = Math.min(this.configuration.centerPoint.size || defaultSize, maxAllowedSize);
                drawCircleSvg(svgDrawingHelper, annotationUID, 'centerPoint', crosshairCenterCanvas, centerPointSize, {
                    color: centerPointColor,
                    fill: centerPointColor,
                });
            }
            return renderStatus;
        };
        this._getAnnotations = (enabledElement) => {
            const { viewport } = enabledElement;
            const annotations = getAnnotations(this.getToolName(), viewport.element) || [];
            const viewportIds = this._getViewportsInfo().map(({ viewportId }) => viewportId);
            const toolGroupAnnotations = annotations.filter((annotation) => {
                const { data } = annotation;
                return viewportIds.includes(data.viewportId);
            });
            return toolGroupAnnotations;
        };
        this._onNewVolume = (_evt) => {
            this._syncVolumeListenersWithToolGroup();
            this._recomputeToolCenterFromAbsoluteCameras({
                emitEvent: true,
                updateViewportCameras: false,
            });
        };
        this._areViewportIdArraysEqual = (viewportIdArrayOne, viewportIdArrayTwo) => {
            if (viewportIdArrayOne.length !== viewportIdArrayTwo.length) {
                return false;
            }
            for (let index = 0; index < viewportIdArrayOne.length; index++) {
                const id = viewportIdArrayOne[index];
                let itemFound = false;
                for (let i = 0; i < viewportIdArrayTwo.length; ++i) {
                    if (id === viewportIdArrayTwo[i]) {
                        itemFound = true;
                        break;
                    }
                }
                if (itemFound === false) {
                    return false;
                }
            }
            return true;
        };
        this._getAnnotationsForViewportsWithDifferentCameras = (enabledElement, annotations) => {
            const { viewportId, renderingEngine, viewport } = enabledElement;
            const otherViewportAnnotations = annotations.filter((annotation) => annotation.data.viewportId !== viewportId);
            if (!otherViewportAnnotations || !otherViewportAnnotations.length) {
                return [];
            }
            const camera = getViewportICamera(viewport);
            const { viewPlaneNormal, position } = camera;
            const viewportsWithDifferentCameras = otherViewportAnnotations.filter((annotation) => {
                const { viewportId } = annotation.data;
                const targetViewport = renderingEngine.getViewport(viewportId);
                const cameraOfTarget = getViewportICamera(targetViewport);
                return !(csUtils.isEqual(cameraOfTarget.viewPlaneNormal, viewPlaneNormal, 1e-2) && csUtils.isEqual(cameraOfTarget.position, position, 1));
            });
            return viewportsWithDifferentCameras;
        };
        this._filterViewportWithSameOrientation = (enabledElement, referenceAnnotation, annotations) => {
            const { renderingEngine } = enabledElement;
            const { data } = referenceAnnotation;
            const viewport = renderingEngine.getViewport(data.viewportId);
            const linkedViewportAnnotations = annotations.filter((annotation) => {
                const { data } = annotation;
                const otherViewport = renderingEngine.getViewport(data.viewportId);
                const otherViewportControllable = this._getReferenceLineControllable(otherViewport.id);
                return otherViewportControllable === true;
            });
            if (!linkedViewportAnnotations || !linkedViewportAnnotations.length) {
                return [];
            }
            const camera = getViewportICamera(viewport);
            const viewPlaneNormal = camera.viewPlaneNormal;
            vtkMath.normalize(viewPlaneNormal);
            const otherViewportsAnnotationsWithSameCameraDirection = linkedViewportAnnotations.filter((annotation) => {
                const { viewportId } = annotation.data;
                const otherViewport = renderingEngine.getViewport(viewportId);
                const otherCamera = getViewportICamera(otherViewport);
                const otherViewPlaneNormal = otherCamera.viewPlaneNormal;
                vtkMath.normalize(otherViewPlaneNormal);
                return (csUtils.isEqual(viewPlaneNormal, otherViewPlaneNormal, 1e-2) &&
                    csUtils.isEqual(camera.viewUp, otherCamera.viewUp, 1e-2));
            });
            return otherViewportsAnnotationsWithSameCameraDirection;
        };
        this._filterAnnotationsByUniqueViewportOrientations = (enabledElement, annotations) => {
            const { renderingEngine, viewport } = enabledElement;
            const camera = getViewportICamera(viewport);
            const viewPlaneNormal = camera.viewPlaneNormal;
            vtkMath.normalize(viewPlaneNormal);
            const otherLinkedViewportAnnotationsFromSameScene = annotations.filter((annotation) => {
                const { data } = annotation;
                const otherViewport = renderingEngine.getViewport(data.viewportId);
                const otherViewportControllable = this._getReferenceLineControllable(otherViewport.id);
                return (viewport !== otherViewport &&
                    otherViewportControllable === true);
            });
            const otherViewportsAnnotationsWithUniqueCameras = [];
            for (let i = 0; i < otherLinkedViewportAnnotationsFromSameScene.length; ++i) {
                const annotation = otherLinkedViewportAnnotationsFromSameScene[i];
                const { viewportId } = annotation.data;
                const otherViewport = renderingEngine.getViewport(viewportId);
                const otherCamera = getViewportICamera(otherViewport);
                const otherViewPlaneNormal = otherCamera.viewPlaneNormal;
                vtkMath.normalize(otherViewPlaneNormal);
                if (csUtils.isEqual(viewPlaneNormal, otherViewPlaneNormal, 1e-2) ||
                    csUtils.isOpposite(viewPlaneNormal, otherViewPlaneNormal, 1e-2)) {
                    continue;
                }
                let cameraFound = false;
                for (let jj = 0; jj < otherViewportsAnnotationsWithUniqueCameras.length; ++jj) {
                    const annotation = otherViewportsAnnotationsWithUniqueCameras[jj];
                    const { viewportId } = annotation.data;
                    const stockedViewport = renderingEngine.getViewport(viewportId);
                    const cameraOfStocked = getViewportICamera(stockedViewport);
                    if (csUtils.isEqual(cameraOfStocked.viewPlaneNormal, otherCamera.viewPlaneNormal, 1e-2) &&
                        csUtils.isEqual(cameraOfStocked.position, otherCamera.position, 1)) {
                        cameraFound = true;
                    }
                }
                if (!cameraFound) {
                    otherViewportsAnnotationsWithUniqueCameras.push(annotation);
                }
            }
            const otherNonLinkedViewportAnnotationsFromSameScene = annotations.filter((annotation) => {
                const { data } = annotation;
                const otherViewport = renderingEngine.getViewport(data.viewportId);
                const otherViewportControllable = this._getReferenceLineControllable(otherViewport.id);
                return (viewport !== otherViewport &&
                    otherViewportControllable !== true);
            });
            for (let i = 0; i < otherNonLinkedViewportAnnotationsFromSameScene.length; ++i) {
                const annotation = otherNonLinkedViewportAnnotationsFromSameScene[i];
                const { viewportId } = annotation.data;
                const otherViewport = renderingEngine.getViewport(viewportId);
                const otherCamera = getViewportICamera(otherViewport);
                const otherViewPlaneNormal = otherCamera.viewPlaneNormal;
                vtkMath.normalize(otherViewPlaneNormal);
                if (csUtils.isEqual(viewPlaneNormal, otherViewPlaneNormal, 1e-2) ||
                    csUtils.isOpposite(viewPlaneNormal, otherViewPlaneNormal, 1e-2)) {
                    continue;
                }
                let cameraFound = false;
                for (let jj = 0; jj < otherViewportsAnnotationsWithUniqueCameras.length; ++jj) {
                    const annotation = otherViewportsAnnotationsWithUniqueCameras[jj];
                    const { viewportId } = annotation.data;
                    const stockedViewport = renderingEngine.getViewport(viewportId);
                    const cameraOfStocked = getViewportICamera(stockedViewport);
                    if (csUtils.isEqual(cameraOfStocked.viewPlaneNormal, otherCamera.viewPlaneNormal, 1e-2) &&
                        csUtils.isEqual(cameraOfStocked.position, otherCamera.position, 1)) {
                        cameraFound = true;
                    }
                }
                if (!cameraFound) {
                    otherViewportsAnnotationsWithUniqueCameras.push(annotation);
                }
            }
            const otherViewportAnnotations = this._getAnnotationsForViewportsWithDifferentCameras(enabledElement, annotations);
            for (let i = 0; i < otherViewportAnnotations.length; ++i) {
                const annotation = otherViewportAnnotations[i];
                if (otherViewportsAnnotationsWithUniqueCameras.some((element) => element === annotation)) {
                    continue;
                }
                const { viewportId } = annotation.data;
                const otherViewport = renderingEngine.getViewport(viewportId);
                const otherCamera = getViewportICamera(otherViewport);
                const otherViewPlaneNormal = otherCamera.viewPlaneNormal;
                vtkMath.normalize(otherViewPlaneNormal);
                if (csUtils.isEqual(viewPlaneNormal, otherViewPlaneNormal, 1e-2) ||
                    csUtils.isOpposite(viewPlaneNormal, otherViewPlaneNormal, 1e-2)) {
                    continue;
                }
                let cameraFound = false;
                for (let jj = 0; jj < otherViewportsAnnotationsWithUniqueCameras.length; ++jj) {
                    const annotation = otherViewportsAnnotationsWithUniqueCameras[jj];
                    const { viewportId } = annotation.data;
                    const stockedViewport = renderingEngine.getViewport(viewportId);
                    const cameraOfStocked = getViewportICamera(stockedViewport);
                    if (csUtils.isEqual(cameraOfStocked.viewPlaneNormal, otherCamera.viewPlaneNormal, 1e-2) &&
                        csUtils.isEqual(cameraOfStocked.position, otherCamera.position, 1)) {
                        cameraFound = true;
                    }
                }
                if (!cameraFound) {
                    otherViewportsAnnotationsWithUniqueCameras.push(annotation);
                }
            }
            return otherViewportsAnnotationsWithUniqueCameras;
        };
        this._checkIfViewportsRenderingSameScene = (viewport, otherViewport) => {
            if (csUtils.isGenericViewport(viewport) ||
                csUtils.isGenericViewport(otherViewport)) {
                const ref = viewport.getViewReference?.();
                const otherRef = otherViewport.getViewReference?.();
                const a = ref?.dataId ?? ref?.FrameOfReferenceUID;
                const b = otherRef?.dataId ?? otherRef?.FrameOfReferenceUID;
                return !!a && a === b;
            }
            const volumeIds = viewport.getAllVolumeIds();
            const otherVolumeIds = otherViewport.getAllVolumeIds();
            return (volumeIds.length === otherVolumeIds.length &&
                volumeIds.every((id) => otherVolumeIds.includes(id)));
        };
        this._jump = (enabledElement, jumpWorld) => {
            state.isInteractingWithTool = true;
            const { viewport, renderingEngine } = enabledElement;
            const annotations = this._getAnnotations(enabledElement);
            const delta = [0, 0, 0];
            vtkMath.subtract(jumpWorld, this.toolCenter, delta);
            const otherViewportAnnotations = this._getAnnotationsForViewportsWithDifferentCameras(enabledElement, annotations);
            const viewportsAnnotationsToUpdate = otherViewportAnnotations.filter((annotation) => {
                const { data } = annotation;
                const otherViewport = renderingEngine.getViewport(data.viewportId);
                const sameScene = this._checkIfViewportsRenderingSameScene(viewport, otherViewport);
                return (this._getReferenceLineControllable(otherViewport.id) &&
                    this._getReferenceLineDraggableRotatable(otherViewport.id) &&
                    sameScene);
            });
            if (viewportsAnnotationsToUpdate.length === 0) {
                state.isInteractingWithTool = false;
                return false;
            }
            this._applyDeltaShiftToSelectedViewportCameras(renderingEngine, viewportsAnnotationsToUpdate, delta);
            this._recomputeToolCenterFromAbsoluteCameras({
                emitEvent: true,
                updateViewportCameras: false,
            });
            viewport.render();
            state.isInteractingWithTool = false;
            return true;
        };
        this._activateModify = (element) => {
            this._syncVolumeListenersWithToolGroup();
            this._recomputeToolCenterFromAbsoluteCameras({
                emitEvent: false,
                updateViewportCameras: false,
            });
            state.isInteractingWithTool = !this.configuration.mobile?.enabled;
            element.addEventListener(Events.MOUSE_UP, this._endCallback);
            element.addEventListener(Events.MOUSE_DRAG, this._dragCallback);
            element.addEventListener(Events.MOUSE_CLICK, this._endCallback);
            element.addEventListener(Events.TOUCH_END, this._endCallback);
            element.addEventListener(Events.TOUCH_DRAG, this._dragCallback);
            element.addEventListener(Events.TOUCH_TAP, this._endCallback);
        };
        this._deactivateModify = (element) => {
            state.isInteractingWithTool = false;
            element.removeEventListener(Events.MOUSE_UP, this._endCallback);
            element.removeEventListener(Events.MOUSE_DRAG, this._dragCallback);
            element.removeEventListener(Events.MOUSE_CLICK, this._endCallback);
            element.removeEventListener(Events.TOUCH_END, this._endCallback);
            element.removeEventListener(Events.TOUCH_DRAG, this._dragCallback);
            element.removeEventListener(Events.TOUCH_TAP, this._endCallback);
        };
        this._endCallback = (evt) => {
            const eventDetail = evt.detail;
            const { element } = eventDetail;
            if (this.editData?.annotation?.data) {
                this.editData.annotation.data.handles.activeOperation = null;
                this.editData.annotation.data.activeViewportIds = [];
            }
            this._deactivateModify(element);
            this._recomputeToolCenterFromAbsoluteCameras({
                emitEvent: true,
                updateViewportCameras: false,
            });
            resetElementCursor(element);
            this.editData = null;
            const requireSameOrientation = false;
            const viewportIdsToRender = getViewportIdsWithToolToRender(element, this.getToolName(), requireSameOrientation);
            triggerAnnotationRenderForViewportIds(viewportIdsToRender);
        };
        this._dragCallback = (evt) => {
            const eventDetail = evt.detail;
            const delta = eventDetail.deltaPoints.world;
            if (Math.abs(delta[0]) < 1e-3 &&
                Math.abs(delta[1]) < 1e-3 &&
                Math.abs(delta[2]) < 1e-3) {
                return;
            }
            const { element } = eventDetail;
            const enabledElement = getEnabledElement(element);
            const { renderingEngine, viewport } = enabledElement;
            const annotations = this._getAnnotations(enabledElement);
            const filteredToolAnnotations = this.filterInteractableAnnotationsForElement(element, annotations);
            const viewportAnnotation = filteredToolAnnotations[0];
            if (!viewportAnnotation) {
                return;
            }
            const { handles } = viewportAnnotation.data;
            const { currentPoints } = evt.detail;
            const canvasCoords = currentPoints.canvas;
            if (handles.activeOperation === OPERATION.DRAG) {
                const otherViewportAnnotations = this._getAnnotationsForViewportsWithDifferentCameras(enabledElement, annotations);
                const viewportsAnnotationsToUpdate = otherViewportAnnotations.filter((annotation) => {
                    const { data } = annotation;
                    const otherViewport = renderingEngine.getViewport(data.viewportId);
                    const otherViewportControllable = this._getReferenceLineControllable(otherViewport.id);
                    const otherViewportDraggableRotatable = this._getReferenceLineDraggableRotatable(otherViewport.id);
                    return (otherViewportControllable === true &&
                        otherViewportDraggableRotatable === true &&
                        viewportAnnotation.data.activeViewportIds.find((id) => id === otherViewport.id));
                });
                this._applyDeltaShiftToSelectedViewportCameras(renderingEngine, viewportsAnnotationsToUpdate, delta);
                this._recomputeToolCenterFromAbsoluteCameras({
                    emitEvent: true,
                    updateViewportCameras: false,
                });
            }
            else if (handles.activeOperation === OPERATION.ROTATE) {
                if (csUtils.isGenericViewport(enabledElement.viewport)) {
                    return;
                }
                const otherViewportAnnotations = this._getAnnotationsForViewportsWithDifferentCameras(enabledElement, annotations);
                const viewportsAnnotationsToUpdate = otherViewportAnnotations.filter((annotation) => {
                    const { data } = annotation;
                    const otherViewport = renderingEngine.getViewport(data.viewportId);
                    const otherViewportControllable = this._getReferenceLineControllable(otherViewport.id);
                    const otherViewportDraggableRotatable = this._getReferenceLineDraggableRotatable(otherViewport.id);
                    return (otherViewportControllable === true &&
                        otherViewportDraggableRotatable === true);
                });
                const dir1 = vec2.create();
                const dir2 = vec2.create();
                const center = [
                    this.toolCenter[0],
                    this.toolCenter[1],
                    this.toolCenter[2],
                ];
                const centerCanvas = viewport.worldToCanvas(center);
                const finalPointCanvas = eventDetail.currentPoints.canvas;
                const originalPointCanvas = vec2.create();
                vec2.sub(originalPointCanvas, finalPointCanvas, eventDetail.deltaPoints.canvas);
                vec2.sub(dir1, originalPointCanvas, centerCanvas);
                vec2.sub(dir2, finalPointCanvas, centerCanvas);
                let angle = vec2.angle(dir1, dir2);
                if (this._isClockWise(centerCanvas, originalPointCanvas, finalPointCanvas)) {
                    angle *= -1;
                }
                angle = Math.round(angle * 100) / 100;
                const rotationAxis = getViewportICamera(viewport).viewPlaneNormal;
                const { matrix } = vtkMatrixBuilder
                    .buildFromRadian()
                    .translate(center[0], center[1], center[2])
                    .rotate(angle, rotationAxis)
                    .translate(-center[0], -center[1], -center[2]);
                const otherViewportsIds = [];
                const previousIgnoreFiredEvents = this._ignoreFiredEvents;
                this._ignoreFiredEvents = true;
                try {
                    viewportsAnnotationsToUpdate.forEach((annotation) => {
                        const { data } = annotation;
                        data.handles.toolCenter = center;
                        const otherViewport = renderingEngine.getViewport(data.viewportId);
                        const camera = getViewportICamera(otherViewport);
                        const { viewUp, position, focalPoint } = camera;
                        viewUp[0] += position[0];
                        viewUp[1] += position[1];
                        viewUp[2] += position[2];
                        vec3.transformMat4(focalPoint, focalPoint, matrix);
                        vec3.transformMat4(position, position, matrix);
                        vec3.transformMat4(viewUp, viewUp, matrix);
                        viewUp[0] -= position[0];
                        viewUp[1] -= position[1];
                        viewUp[2] -= position[2];
                        otherViewport.setCamera({
                            position,
                            viewUp,
                            focalPoint,
                        });
                        otherViewportsIds.push(otherViewport.id);
                    });
                }
                finally {
                    this._ignoreFiredEvents = previousIgnoreFiredEvents;
                }
                renderingEngine.renderViewports(otherViewportsIds);
                this._recomputeToolCenterFromAbsoluteCameras({
                    emitEvent: true,
                    updateViewportCameras: false,
                });
            }
            else if (handles.activeOperation === OPERATION.SLAB) {
                if (csUtils.isGenericViewport(enabledElement.viewport)) {
                    return;
                }
                const otherViewportAnnotations = this._getAnnotationsForViewportsWithDifferentCameras(enabledElement, annotations);
                const referenceAnnotations = otherViewportAnnotations.filter((annotation) => {
                    const { data } = annotation;
                    const otherViewport = renderingEngine.getViewport(data.viewportId);
                    const otherViewportControllable = this._getReferenceLineControllable(otherViewport.id);
                    const otherViewportSlabThicknessControlsOn = this._getReferenceLineSlabThicknessControlsOn(otherViewport.id);
                    return (otherViewportControllable === true &&
                        otherViewportSlabThicknessControlsOn === true &&
                        viewportAnnotation.data.activeViewportIds.find((id) => id === otherViewport.id));
                });
                if (referenceAnnotations.length === 0) {
                    return;
                }
                const viewportsAnnotationsToUpdate = this._filterViewportWithSameOrientation(enabledElement, referenceAnnotations[0], annotations);
                const viewportsIds = [];
                viewportsIds.push(viewport.id);
                viewportsAnnotationsToUpdate.forEach((annotation) => {
                    const { data } = annotation;
                    const otherViewport = renderingEngine.getViewport(data.viewportId);
                    const camera = getViewportICamera(otherViewport);
                    const normal = camera.viewPlaneNormal;
                    const dotProd = vtkMath.dot(delta, normal);
                    const projectedDelta = [...normal];
                    vtkMath.multiplyScalar(projectedDelta, dotProd);
                    if (Math.abs(projectedDelta[0]) > 1e-3 ||
                        Math.abs(projectedDelta[1]) > 1e-3 ||
                        Math.abs(projectedDelta[2]) > 1e-3) {
                        const mod = Math.sqrt(projectedDelta[0] * projectedDelta[0] +
                            projectedDelta[1] * projectedDelta[1] +
                            projectedDelta[2] * projectedDelta[2]);
                        const currentPoint = eventDetail.lastPoints.world;
                        const direction = [0, 0, 0];
                        const currentCenter = [
                            this.toolCenter[0],
                            this.toolCenter[1],
                            this.toolCenter[2],
                        ];
                        const viewportDraggableRotatable = this._getReferenceLineDraggableRotatable(otherViewport.id);
                        if (!viewportDraggableRotatable) {
                            const { rotationPoints } = (this.editData.annotation.data).handles;
                            const otherViewportRotationPoints = rotationPoints.filter((point) => point[1].uid === otherViewport.id);
                            if (otherViewportRotationPoints.length === 2) {
                                const point1 = viewport.canvasToWorld(otherViewportRotationPoints[0][3]);
                                const point2 = viewport.canvasToWorld(otherViewportRotationPoints[1][3]);
                                vtkMath.add(point1, point2, currentCenter);
                                vtkMath.multiplyScalar(currentCenter, 0.5);
                            }
                        }
                        vtkMath.subtract(currentPoint, currentCenter, direction);
                        const dotProdDirection = vtkMath.dot(direction, normal);
                        const projectedDirection = [...normal];
                        vtkMath.multiplyScalar(projectedDirection, dotProdDirection);
                        const normalizedProjectedDirection = [
                            projectedDirection[0],
                            projectedDirection[1],
                            projectedDirection[2],
                        ];
                        vec3.normalize(normalizedProjectedDirection, normalizedProjectedDirection);
                        const normalizedProjectedDelta = [
                            projectedDelta[0],
                            projectedDelta[1],
                            projectedDelta[2],
                        ];
                        vec3.normalize(normalizedProjectedDelta, normalizedProjectedDelta);
                        let slabThicknessValue = getSlabThicknessOrDefault(otherViewport);
                        if (csUtils.isOpposite(normalizedProjectedDirection, normalizedProjectedDelta, 1e-3)) {
                            slabThicknessValue -= mod;
                        }
                        else {
                            slabThicknessValue += mod;
                        }
                        slabThicknessValue = Math.abs(slabThicknessValue);
                        slabThicknessValue = Math.max(RENDERING_DEFAULTS.MINIMUM_SLAB_THICKNESS, slabThicknessValue);
                        const near = this._pointNearReferenceLine(viewportAnnotation, canvasCoords, 6, otherViewport);
                        if (near) {
                            slabThicknessValue = RENDERING_DEFAULTS.MINIMUM_SLAB_THICKNESS;
                        }
                        const toolGroup = getToolGroupForViewport(otherViewport.id, renderingEngine.id);
                        const crosshairsInstance = toolGroup.getToolInstance(this.getToolName());
                        crosshairsInstance.setSlabThickness(otherViewport, slabThicknessValue);
                        viewportsIds.push(otherViewport.id);
                    }
                });
                renderingEngine.renderViewports(viewportsIds);
                this._recomputeToolCenterFromAbsoluteCameras({
                    emitEvent: true,
                    updateViewportCameras: false,
                });
            }
            const requireSameOrientation = false;
            const viewportIdsToRender = getViewportIdsWithToolToRender(element, this.getToolName(), requireSameOrientation);
            triggerAnnotationRenderForViewportIds(viewportIdsToRender);
        };
        this._pointNearReferenceLine = (annotation, canvasCoords, proximity, lineViewport) => {
            const { data } = annotation;
            const { rotationPoints } = data.handles;
            for (let i = 0; i < rotationPoints.length - 1; ++i) {
                const otherViewport = rotationPoints[i][1];
                if (otherViewport.id !== lineViewport.id) {
                    continue;
                }
                const viewportControllable = this._getReferenceLineControllable(otherViewport.id);
                if (!viewportControllable) {
                    continue;
                }
                const lineSegment1 = {
                    start: {
                        x: rotationPoints[i][2][0],
                        y: rotationPoints[i][2][1],
                    },
                    end: {
                        x: rotationPoints[i][3][0],
                        y: rotationPoints[i][3][1],
                    },
                };
                const distanceToPoint1 = lineSegment.distanceToPoint([lineSegment1.start.x, lineSegment1.start.y], [lineSegment1.end.x, lineSegment1.end.y], [canvasCoords[0], canvasCoords[1]]);
                const lineSegment2 = {
                    start: {
                        x: rotationPoints[i + 1][2][0],
                        y: rotationPoints[i + 1][2][1],
                    },
                    end: {
                        x: rotationPoints[i + 1][3][0],
                        y: rotationPoints[i + 1][3][1],
                    },
                };
                const distanceToPoint2 = lineSegment.distanceToPoint([lineSegment2.start.x, lineSegment2.start.y], [lineSegment2.end.x, lineSegment2.end.y], [canvasCoords[0], canvasCoords[1]]);
                if (distanceToPoint1 <= proximity || distanceToPoint2 <= proximity) {
                    return true;
                }
                i++;
            }
            return false;
        };
        this._toViewportKey = (renderingEngineId, viewportId) => {
            return `${renderingEngineId}::${viewportId}`;
        };
        this._isFinitePoint3 = (point) => {
            if (!point || point.length !== 3) {
                return false;
            }
            return (Number.isFinite(point[0]) &&
                Number.isFinite(point[1]) &&
                Number.isFinite(point[2]));
        };
        this._bindToolGroupViewportListeners = () => {
            if (!this._toolGroupViewportAddedListener) {
                this._toolGroupViewportAddedListener = ((evt) => {
                    if (evt.detail?.toolGroupId !== this.toolGroupId) {
                        return;
                    }
                    this._syncVolumeListenersWithToolGroup();
                    this._computeToolCenter(this._getViewportsInfo());
                });
                eventTarget.addEventListener(Events.TOOLGROUP_VIEWPORT_ADDED, this._toolGroupViewportAddedListener);
            }
            if (!this._toolGroupViewportRemovedListener) {
                this._toolGroupViewportRemovedListener = ((evt) => {
                    if (evt.detail?.toolGroupId !== this.toolGroupId) {
                        return;
                    }
                    this._syncVolumeListenersWithToolGroup();
                    this._recomputeToolCenterFromAbsoluteCameras({
                        emitEvent: true,
                        updateViewportCameras: false,
                    });
                });
                eventTarget.addEventListener(Events.TOOLGROUP_VIEWPORT_REMOVED, this._toolGroupViewportRemovedListener);
            }
        };
        this._unbindToolGroupViewportListeners = () => {
            if (this._toolGroupViewportAddedListener) {
                eventTarget.removeEventListener(Events.TOOLGROUP_VIEWPORT_ADDED, this._toolGroupViewportAddedListener);
                this._toolGroupViewportAddedListener = null;
            }
            if (this._toolGroupViewportRemovedListener) {
                eventTarget.removeEventListener(Events.TOOLGROUP_VIEWPORT_REMOVED, this._toolGroupViewportRemovedListener);
                this._toolGroupViewportRemovedListener = null;
            }
        };
        this._syncVolumeListenersWithToolGroup = () => {
            const viewportsInfo = this._getViewportsInfo();
            const activeViewportKeys = new Set();
            viewportsInfo.forEach((viewportInfo) => {
                const { viewportId, renderingEngineId } = viewportInfo;
                const viewportKey = this._toViewportKey(renderingEngineId, viewportId);
                activeViewportKeys.add(viewportKey);
                const enabledElement = getEnabledElementByIds(viewportId, renderingEngineId);
                const existingListenerInfo = this._volumeViewportNewVolumeListeners.get(viewportKey);
                if (!enabledElement) {
                    if (existingListenerInfo) {
                        existingListenerInfo.element.removeEventListener(Enums.Events.VOLUME_VIEWPORT_NEW_VOLUME, existingListenerInfo.handler);
                        this._volumeViewportNewVolumeListeners.delete(viewportKey);
                    }
                    return;
                }
                const { viewport } = enabledElement;
                const { element } = viewport;
                if (existingListenerInfo && existingListenerInfo.element !== element) {
                    existingListenerInfo.element.removeEventListener(Enums.Events.VOLUME_VIEWPORT_NEW_VOLUME, existingListenerInfo.handler);
                    this._volumeViewportNewVolumeListeners.delete(viewportKey);
                }
                if (this._volumeViewportNewVolumeListeners.has(viewportKey)) {
                    return;
                }
                const handler = ((evt) => this._onNewVolume(evt));
                element.addEventListener(Enums.Events.VOLUME_VIEWPORT_NEW_VOLUME, handler);
                this._volumeViewportNewVolumeListeners.set(viewportKey, {
                    element,
                    handler,
                });
            });
            Array.from(this._volumeViewportNewVolumeListeners.entries()).forEach(([viewportKey, listenerInfo]) => {
                if (activeViewportKeys.has(viewportKey)) {
                    return;
                }
                listenerInfo.element.removeEventListener(Enums.Events.VOLUME_VIEWPORT_NEW_VOLUME, listenerInfo.handler);
                this._volumeViewportNewVolumeListeners.delete(viewportKey);
            });
        };
        this._clearAllVolumeListenersAndViewportState = () => {
            this._volumeViewportNewVolumeListeners.forEach((listenerInfo) => {
                listenerInfo.element.removeEventListener(Enums.Events.VOLUME_VIEWPORT_NEW_VOLUME, listenerInfo.handler);
            });
            this._volumeViewportNewVolumeListeners.clear();
        };
        this._calculateToolCenterFromAbsoluteCameras = () => {
            const viewportsInfo = this._getViewportsInfo();
            const uniquePlanes = [];
            viewportsInfo.forEach((viewportInfo) => {
                const enabledElement = getEnabledElementByIds(viewportInfo.viewportId, viewportInfo.renderingEngineId);
                if (!enabledElement) {
                    return;
                }
                const camera = getViewportICamera(enabledElement.viewport);
                const normal = [...camera.viewPlaneNormal];
                const point = [...camera.focalPoint];
                if (!this._isFinitePoint3(normal) || !this._isFinitePoint3(point)) {
                    return;
                }
                vec3.normalize(normal, normal);
                const alreadyTracked = uniquePlanes.some((plane) => csUtils.isEqual(plane.normal, normal, 1e-3) ||
                    csUtils.isOpposite(plane.normal, normal, 1e-3));
                if (!alreadyTracked) {
                    uniquePlanes.push({ normal, point });
                }
            });
            if (uniquePlanes.length < 2) {
                return null;
            }
            const firstPlane = csUtils.planar.planeEquation(uniquePlanes[0].normal, uniquePlanes[0].point);
            const secondPlane = csUtils.planar.planeEquation(uniquePlanes[1].normal, uniquePlanes[1].point);
            let thirdPlane;
            if (uniquePlanes.length >= 3) {
                thirdPlane = csUtils.planar.planeEquation(uniquePlanes[2].normal, uniquePlanes[2].point);
            }
            else {
                const thirdNormal = vec3.create();
                vec3.cross(thirdNormal, uniquePlanes[0].normal, uniquePlanes[1].normal);
                if (vec3.length(thirdNormal) < 1e-6) {
                    return null;
                }
                vec3.normalize(thirdNormal, thirdNormal);
                const thirdPoint = this._isFinitePoint3(this.toolCenter)
                    ? [...this.toolCenter]
                    : [
                        (uniquePlanes[0].point[0] + uniquePlanes[1].point[0]) * 0.5,
                        (uniquePlanes[0].point[1] + uniquePlanes[1].point[1]) * 0.5,
                        (uniquePlanes[0].point[2] + uniquePlanes[1].point[2]) * 0.5,
                    ];
                thirdPlane = csUtils.planar.planeEquation(thirdNormal, thirdPoint);
            }
            const center = csUtils.planar.threePlaneIntersection(firstPlane, secondPlane, thirdPlane);
            return this._isFinitePoint3(center) ? center : null;
        };
        this._recomputeToolCenterFromAbsoluteCameras = ({ emitEvent = true, updateViewportCameras = false, } = {}) => {
            const toolCenter = this._calculateToolCenterFromAbsoluteCameras();
            if (!toolCenter) {
                return null;
            }
            const hasChanged = !csUtils.isEqual(this.toolCenter, toolCenter, 1e-3);
            if (!hasChanged) {
                return toolCenter;
            }
            if (updateViewportCameras) {
                this.setToolCenter(toolCenter, !emitEvent);
            }
            else {
                this.toolCenter = toolCenter;
                if (emitEvent) {
                    triggerEvent(eventTarget, Events.CROSSHAIR_TOOL_CENTER_CHANGED, {
                        toolGroupId: this.toolGroupId,
                        toolCenter: this.toolCenter,
                    });
                }
            }
            return toolCenter;
        };
        this._getReferenceLineColor =
            toolProps.configuration?.getReferenceLineColor ||
                defaultReferenceLineColor;
        this._getReferenceLineControllable =
            toolProps.configuration?.getReferenceLineControllable ||
                defaultReferenceLineControllable;
        this._getReferenceLineDraggableRotatable =
            toolProps.configuration?.getReferenceLineDraggableRotatable ||
                defaultReferenceLineDraggableRotatable;
        this._getReferenceLineSlabThicknessControlsOn =
            toolProps.configuration?.getReferenceLineSlabThicknessControlsOn ||
                defaultReferenceLineSlabThicknessControlsOn;
    }
    onSetToolActive() {
        this._reinitializeListenersAndCenter();
    }
    onSetToolPassive() {
        this._reinitializeListenersAndCenter();
    }
    onSetToolEnabled() {
        this._reinitializeListenersAndCenter();
    }
    onSetToolDisabled() {
        const viewportsInfo = this._getViewportsInfo();
        this._unbindToolGroupViewportListeners();
        this._clearAllVolumeListenersAndViewportState();
        this._ignoreFiredEvents = false;
        this.editData = null;
        state.isInteractingWithTool = false;
        viewportsInfo.forEach(({ renderingEngineId, viewportId }) => {
            const enabledElement = getEnabledElementByIds(viewportId, renderingEngineId);
            if (!enabledElement) {
                return;
            }
            const annotations = this._getAnnotations(enabledElement);
            if (annotations?.length) {
                annotations.forEach((annotation) => {
                    removeAnnotation(annotation.annotationUID);
                });
            }
        });
    }
    setToolCenter(toolCenter, suppressEvents = false) {
        const viewportsInfo = this._getViewportsInfo();
        const previousIgnoreFiredEvents = this._ignoreFiredEvents;
        this._ignoreFiredEvents = true;
        try {
            viewportsInfo.forEach(({ renderingEngineId, viewportId }) => {
                const renderingEngine = getRenderingEngine(renderingEngineId);
                if (!renderingEngine) {
                    return;
                }
                const viewport = renderingEngine.getViewport(viewportId);
                if (!viewport) {
                    return;
                }
                const camera = getViewportICamera(viewport);
                const { focalPoint, position, viewPlaneNormal } = camera;
                const delta = [
                    toolCenter[0] - focalPoint[0],
                    toolCenter[1] - focalPoint[1],
                    toolCenter[2] - focalPoint[2],
                ];
                const scroll = delta[0] * viewPlaneNormal[0] +
                    delta[1] * viewPlaneNormal[1] +
                    delta[2] * viewPlaneNormal[2];
                const scrollDelta = [
                    scroll * viewPlaneNormal[0],
                    scroll * viewPlaneNormal[1],
                    scroll * viewPlaneNormal[2],
                ];
                const newFocalPoint = [
                    focalPoint[0] + scrollDelta[0],
                    focalPoint[1] + scrollDelta[1],
                    focalPoint[2] + scrollDelta[2],
                ];
                const newPosition = [
                    position[0] + scrollDelta[0],
                    position[1] + scrollDelta[1],
                    position[2] + scrollDelta[2],
                ];
                if (csUtils.isGenericViewport(viewport)) {
                    jumpToFocalPoint(viewport, newFocalPoint);
                }
                else {
                    viewport.setCamera({
                        focalPoint: newFocalPoint,
                        position: newPosition,
                    });
                }
                viewport.render();
            });
        }
        finally {
            this._ignoreFiredEvents = previousIgnoreFiredEvents;
        }
        this.toolCenter = toolCenter;
        if (!suppressEvents) {
            triggerEvent(eventTarget, Events.CROSSHAIR_TOOL_CENTER_CHANGED, {
                toolGroupId: this.toolGroupId,
                toolCenter: this.toolCenter,
            });
        }
    }
    getHandleNearImagePoint(element, annotation, canvasCoords, proximity) {
        const enabledElement = getEnabledElement(element);
        const { viewport } = enabledElement;
        let point = this._getRotationHandleNearImagePoint(viewport, annotation, canvasCoords, proximity);
        if (point !== null) {
            return point;
        }
        point = this._getSlabThicknessHandleNearImagePoint(viewport, annotation, canvasCoords, proximity);
        if (point !== null) {
            return point;
        }
    }
    _unsubscribeToViewportNewVolumeSet(_viewportsInfo) {
        this._syncVolumeListenersWithToolGroup();
    }
    _subscribeToViewportNewVolumeSet(_viewports) {
        this._syncVolumeListenersWithToolGroup();
    }
    _autoPanViewportIfNecessary(viewportId, renderingEngine) {
        const viewport = renderingEngine.getViewport(viewportId);
        if (csUtils.isGenericViewport(viewport)) {
            return;
        }
        const { clientWidth, clientHeight } = getDisplayedCanvasSize(viewport);
        const toolCenterCanvas = viewport.worldToCanvas(this.toolCenter);
        const pan = this.configuration.autoPan.panSize;
        const visiblePointCanvas = [
            toolCenterCanvas[0],
            toolCenterCanvas[1],
        ];
        if (toolCenterCanvas[0] < 0) {
            visiblePointCanvas[0] = pan;
        }
        else if (toolCenterCanvas[0] > clientWidth) {
            visiblePointCanvas[0] = clientWidth - pan;
        }
        if (toolCenterCanvas[1] < 0) {
            visiblePointCanvas[1] = pan;
        }
        else if (toolCenterCanvas[1] > clientHeight) {
            visiblePointCanvas[1] = clientHeight - pan;
        }
        if (visiblePointCanvas[0] === toolCenterCanvas[0] &&
            visiblePointCanvas[1] === toolCenterCanvas[1]) {
            return;
        }
        const visiblePointWorld = viewport.canvasToWorld(visiblePointCanvas);
        const deltaPointsWorld = [
            visiblePointWorld[0] - this.toolCenter[0],
            visiblePointWorld[1] - this.toolCenter[1],
            visiblePointWorld[2] - this.toolCenter[2],
        ];
        const camera = getViewportICamera(viewport);
        const { focalPoint, position } = camera;
        const updatedPosition = [
            position[0] - deltaPointsWorld[0],
            position[1] - deltaPointsWorld[1],
            position[2] - deltaPointsWorld[2],
        ];
        const updatedFocalPoint = [
            focalPoint[0] - deltaPointsWorld[0],
            focalPoint[1] - deltaPointsWorld[1],
            focalPoint[2] - deltaPointsWorld[2],
        ];
        const previousIgnoreFiredEvents = this._ignoreFiredEvents;
        this._ignoreFiredEvents = true;
        try {
            viewport.setCamera({
                focalPoint: updatedFocalPoint,
                position: updatedPosition,
            });
        }
        finally {
            this._ignoreFiredEvents = previousIgnoreFiredEvents;
        }
        viewport.render();
    }
    setSlabThickness(viewport, slabThickness) {
        if (csUtils.isGenericViewport(viewport)) {
            return;
        }
        let actorUIDs;
        const { filterActorUIDsToSetSlabThickness } = this.configuration;
        if (filterActorUIDsToSetSlabThickness &&
            filterActorUIDsToSetSlabThickness.length > 0) {
            actorUIDs = filterActorUIDsToSetSlabThickness;
        }
        let blendModeToUse = this.configuration.slabThicknessBlendMode;
        if (slabThickness === RENDERING_DEFAULTS.MINIMUM_SLAB_THICKNESS) {
            blendModeToUse = Enums.BlendModes.COMPOSITE;
        }
        const immediate = false;
        viewport.setBlendMode(blendModeToUse, actorUIDs, immediate);
        viewport.setSlabThickness(slabThickness, actorUIDs);
    }
    _isClockWise(a, b, c) {
        return (b[0] - a[0]) * (c[1] - a[1]) - (b[1] - a[1]) * (c[0] - a[0]) > 0;
    }
    _applyDeltaShiftToSelectedViewportCameras(renderingEngine, viewportsAnnotationsToUpdate, delta) {
        viewportsAnnotationsToUpdate.forEach((annotation) => {
            this._applyDeltaShiftToViewportCamera(renderingEngine, annotation, delta);
        });
    }
    _applyDeltaShiftToViewportCamera(renderingEngine, annotation, delta) {
        const { data } = annotation;
        const viewport = renderingEngine.getViewport(data.viewportId);
        const camera = getViewportICamera(viewport);
        const normal = camera.viewPlaneNormal;
        const dotProd = vtkMath.dot(delta, normal);
        const projectedDelta = [...normal];
        vtkMath.multiplyScalar(projectedDelta, dotProd);
        if (Math.abs(projectedDelta[0]) > 1e-3 ||
            Math.abs(projectedDelta[1]) > 1e-3 ||
            Math.abs(projectedDelta[2]) > 1e-3) {
            const newFocalPoint = [0, 0, 0];
            const newPosition = [0, 0, 0];
            vtkMath.add(camera.focalPoint, projectedDelta, newFocalPoint);
            vtkMath.add(camera.position, projectedDelta, newPosition);
            const previousIgnoreFiredEvents = this._ignoreFiredEvents;
            this._ignoreFiredEvents = true;
            try {
                if (csUtils.isGenericViewport(viewport)) {
                    jumpToFocalPoint(viewport, newFocalPoint);
                }
                else {
                    viewport.setCamera({
                        focalPoint: newFocalPoint,
                        position: newPosition,
                    });
                }
            }
            finally {
                this._ignoreFiredEvents = previousIgnoreFiredEvents;
            }
            viewport.render();
        }
    }
    _getRotationHandleNearImagePoint(viewport, annotation, canvasCoords, proximity) {
        const minimalCrosshairConfig = getMinimalCrosshairConfig(this.configuration);
        const { data } = annotation;
        const { rotationPoints } = data.handles;
        for (let i = 0; i < rotationPoints.length; i++) {
            const point = rotationPoints[i][0];
            const otherViewport = rotationPoints[i][1];
            const viewportControllable = this._getReferenceLineControllable(otherViewport.id);
            if (!viewportControllable) {
                continue;
            }
            const viewportDraggableRotatable = !minimalCrosshairConfig.enabled &&
                this._getReferenceLineDraggableRotatable(otherViewport.id);
            if (!viewportDraggableRotatable) {
                continue;
            }
            const annotationCanvasCoordinate = viewport.worldToCanvas(point);
            if (vec2.distance(canvasCoords, annotationCanvasCoordinate) < proximity) {
                data.handles.activeOperation = OPERATION.ROTATE;
                this.editData = {
                    annotation,
                };
                return point;
            }
        }
        return null;
    }
    _getSlabThicknessHandleNearImagePoint(viewport, annotation, canvasCoords, proximity) {
        const minimalCrosshairConfig = getMinimalCrosshairConfig(this.configuration);
        const { data } = annotation;
        const { slabThicknessPoints } = data.handles;
        for (let i = 0; i < slabThicknessPoints.length; i++) {
            const point = slabThicknessPoints[i][0];
            const otherViewport = slabThicknessPoints[i][1];
            const viewportControllable = this._getReferenceLineControllable(otherViewport.id);
            if (!viewportControllable) {
                continue;
            }
            const viewportSlabThicknessControlsOn = !minimalCrosshairConfig.enabled &&
                this._getReferenceLineSlabThicknessControlsOn(otherViewport.id);
            if (!viewportSlabThicknessControlsOn) {
                continue;
            }
            const annotationCanvasCoordinate = viewport.worldToCanvas(point);
            if (vec2.distance(canvasCoords, annotationCanvasCoordinate) < proximity) {
                data.handles.activeOperation = OPERATION.SLAB;
                data.activeViewportIds = [otherViewport.id];
                this.editData = {
                    annotation,
                };
                return point;
            }
        }
        return null;
    }
    _pointNearTool(element, annotation, canvasCoords, proximity) {
        const minimalCrosshairConfig = getMinimalCrosshairConfig(this.configuration);
        const enabledElement = getEnabledElement(element);
        const { viewport } = enabledElement;
        const { clientWidth, clientHeight } = getDisplayedCanvasSize(viewport);
        const canvasDiagonalLength = Math.sqrt(clientWidth * clientWidth + clientHeight * clientHeight);
        const { data } = annotation;
        const { rotationPoints } = data.handles;
        const { slabThicknessPoints } = data.handles;
        const viewportIdArray = [];
        for (let i = 0; i < rotationPoints.length - 1; ++i) {
            const otherViewport = rotationPoints[i][1];
            const viewportControllable = this._getReferenceLineControllable(otherViewport.id);
            const viewportDraggableRotatable = this._getReferenceLineDraggableRotatable(otherViewport.id);
            if (!viewportControllable ||
                (!viewportDraggableRotatable && !minimalCrosshairConfig.enabled)) {
                continue;
            }
            const lineSegment1 = {
                start: {
                    x: rotationPoints[i][2][0],
                    y: rotationPoints[i][2][1],
                },
                end: {
                    x: rotationPoints[i][3][0],
                    y: rotationPoints[i][3][1],
                },
            };
            const distanceToPoint1 = lineSegment.distanceToPoint([lineSegment1.start.x, lineSegment1.start.y], [lineSegment1.end.x, lineSegment1.end.y], [canvasCoords[0], canvasCoords[1]]);
            const lineSegment2 = {
                start: {
                    x: rotationPoints[i + 1][2][0],
                    y: rotationPoints[i + 1][2][1],
                },
                end: {
                    x: rotationPoints[i + 1][3][0],
                    y: rotationPoints[i + 1][3][1],
                },
            };
            const distanceToPoint2 = lineSegment.distanceToPoint([lineSegment2.start.x, lineSegment2.start.y], [lineSegment2.end.x, lineSegment2.end.y], [canvasCoords[0], canvasCoords[1]]);
            if (distanceToPoint1 <= proximity || distanceToPoint2 <= proximity) {
                viewportIdArray.push(otherViewport.id);
                data.handles.activeOperation = OPERATION.DRAG;
            }
            i++;
        }
        for (let i = 0; i < slabThicknessPoints.length - 1; ++i) {
            const otherViewport = slabThicknessPoints[i][1];
            if (viewportIdArray.find((id) => id === otherViewport.id)) {
                continue;
            }
            const viewportControllable = this._getReferenceLineControllable(otherViewport.id);
            const viewportSlabThicknessControlsOn = !minimalCrosshairConfig.enabled &&
                this._getReferenceLineSlabThicknessControlsOn(otherViewport.id);
            if (!viewportControllable || !viewportSlabThicknessControlsOn) {
                continue;
            }
            const stPointLineCanvas1 = slabThicknessPoints[i][2];
            const stPointLineCanvas2 = slabThicknessPoints[i][3];
            const centerCanvas = vec2.create();
            vec2.add(centerCanvas, stPointLineCanvas1, stPointLineCanvas2);
            vec2.scale(centerCanvas, centerCanvas, 0.5);
            const canvasUnitVectorFromCenter = vec2.create();
            vec2.subtract(canvasUnitVectorFromCenter, stPointLineCanvas1, centerCanvas);
            vec2.normalize(canvasUnitVectorFromCenter, canvasUnitVectorFromCenter);
            const canvasVectorFromCenterStart = vec2.create();
            vec2.scale(canvasVectorFromCenterStart, canvasUnitVectorFromCenter, canvasDiagonalLength * 0.05);
            const stPointLineCanvas1Start = vec2.create();
            const stPointLineCanvas2Start = vec2.create();
            vec2.add(stPointLineCanvas1Start, centerCanvas, canvasVectorFromCenterStart);
            vec2.subtract(stPointLineCanvas2Start, centerCanvas, canvasVectorFromCenterStart);
            const lineSegment1 = {
                start: {
                    x: stPointLineCanvas1Start[0],
                    y: stPointLineCanvas1Start[1],
                },
                end: {
                    x: stPointLineCanvas1[0],
                    y: stPointLineCanvas1[1],
                },
            };
            const distanceToPoint1 = lineSegment.distanceToPoint([lineSegment1.start.x, lineSegment1.start.y], [lineSegment1.end.x, lineSegment1.end.y], [canvasCoords[0], canvasCoords[1]]);
            const lineSegment2 = {
                start: {
                    x: stPointLineCanvas2Start[0],
                    y: stPointLineCanvas2Start[1],
                },
                end: {
                    x: stPointLineCanvas2[0],
                    y: stPointLineCanvas2[1],
                },
            };
            const distanceToPoint2 = lineSegment.distanceToPoint([lineSegment2.start.x, lineSegment2.start.y], [lineSegment2.end.x, lineSegment2.end.y], [canvasCoords[0], canvasCoords[1]]);
            if (distanceToPoint1 <= proximity || distanceToPoint2 <= proximity) {
                viewportIdArray.push(otherViewport.id);
                data.handles.activeOperation = null;
            }
            i++;
        }
        data.activeViewportIds = [...viewportIdArray];
        this.editData = {
            annotation,
        };
        return data.handles.activeOperation === OPERATION.DRAG ? true : false;
    }
}
CrosshairsTool.minimalModeExamples = new Map([
    ['Default', { enabled: false, lineLengthInPx: 40 }],
    ['Minimal 40px', { enabled: true, lineLengthInPx: 40 }],
    ['Minimal 80px', { enabled: true, lineLengthInPx: 80 }],
]);
function getMinimalCrosshairConfig(configuration) {
    const minimal = configuration?.minimal;
    if (!minimal?.enabled) {
        return {
            enabled: false,
            lineLengthInPx: 0,
        };
    }
    const lineLengthInPx = typeof minimal.lineLengthInPx === 'number' &&
        Number.isFinite(minimal.lineLengthInPx)
        ? minimal.lineLengthInPx
        : 40;
    return {
        enabled: true,
        lineLengthInPx: Math.max(0, lineLengthInPx),
    };
}
function getSegmentMidpoint(start, end) {
    const midpoint = vec2.create();
    vec2.add(midpoint, start, end);
    vec2.scale(midpoint, midpoint, 0.5);
    return midpoint;
}
CrosshairsTool.toolName = 'Crosshairs';
export default CrosshairsTool;
