// @cornerstonejs/tools/dist/esm/tools/VolumeCroppingTool.js
import vtkPolyData from '@kitware/vtk.js/Common/DataModel/PolyData';
import { mat3, mat4, vec3 } from 'gl-matrix';
import vtkMath from '@kitware/vtk.js/Common/Core/Math';
import vtkActor from '@kitware/vtk.js/Rendering/Core/Actor';
import vtkSphereSource from '@kitware/vtk.js/Filters/Sources/SphereSource';
import vtkMapper from '@kitware/vtk.js/Rendering/Core/Mapper';
import vtkPlane from '@kitware/vtk.js/Common/DataModel/Plane';
import { BaseTool } from './base';
import { getRenderingEngine, getEnabledElementByIds, getEnabledElement, Enums, triggerEvent, eventTarget, } from '@cornerstonejs/core';
import { getToolGroup } from '../store/ToolGroupManager';
import getViewportICamera from '../utilities/getViewportICamera';
import setViewportCamera, { resetViewportCamera, } from '../utilities/setViewportCamera';
import { Events } from '../enums';
import { PLANEINDEX, SPHEREINDEX, NUM_CLIPPING_PLANES, extractVolumeDirectionVectors, parseCornerKey, copyClippingPlanes, } from '../utilities/volumeCropping';
import { addLine3DBetweenPoints, calculateAdaptiveSphereRadius, } from '../utilities/draw3D';
import { isDragOwnedBy } from '../utilities/interactionDragCoordinator';
import { applyViewportPresentation, getViewportPresentation, } from '../utilities/viewportPresentation';
class VolumeCroppingTool extends BaseTool {
    constructor(toolProps = {}, defaultToolProps = {
        configuration: {
            showCornerSpheres: true,
            showHandles: true,
            showClippingPlanes: true,
            mobile: {
                enabled: false,
                opacity: 0.8,
            },
            initialCropFactor: 0.08,
            sphereColors: {
                SAGITTAL: [1.0, 1.0, 0.0],
                CORONAL: [0.0, 1.0, 0.0],
                AXIAL: [1.0, 0.0, 0.0],
                CORNERS: [0.0, 0.0, 1.0],
            },
            sphereRadius: 8,
            grabSpherePixelDistance: 20,
            rotateIncrementDegrees: 2,
            rotateSampleDistanceFactor: 2,
            rotateClippingPlanesIncrementDegrees: 5,
        },
    }) {
        super(toolProps, defaultToolProps);
        this._resizeObservers = new Map();
        this._hasResolutionChanged = false;
        this.originalClippingPlanes = [];
        this.draggingSphereIndex = null;
        this.rotatePlanesOnDrag = false;
        this.suppressPlaneRotationForCurrentDrag = false;
        this.cornerDragOffset = null;
        this.faceDragOffset = null;
        this.volumeDirectionVectors = null;
        this.sphereStates = [];
        this.edgeLines = {};
        this.onSetToolConfiguration = () => {
        };
        this.onSetToolEnabled = () => {
        };
        this.onCameraModified = (evt) => {
            const { element } = evt.currentTarget
                ? { element: evt.currentTarget }
                : evt.detail;
            const enabledElement = getEnabledElement(element);
            this._updateClippingPlanes(enabledElement.viewport);
            enabledElement.viewport.render();
        };
        this.preMouseDownCallback = (evt) => {
            const eventDetail = evt.detail;
            const { element } = eventDetail;
            const enabledElement = getEnabledElement(element);
            const { viewport } = enabledElement;
            this.suppressPlaneRotationForCurrentDrag = isDragOwnedBy(viewport.id, 'orientation-controller');
            const actor = this._getVolumeActor(viewport);
            if (!actor) {
                return false;
            }
            const mapper = actor.getMapper();
            const mouseCanvas = [
                evt.detail.currentPoints.canvas[0],
                evt.detail.currentPoints.canvas[1],
            ];
            this.draggingSphereIndex = null;
            this.cornerDragOffset = null;
            this.faceDragOffset = null;
            for (let i = 0; i < this.sphereStates.length; ++i) {
                const sphereCanvas = viewport.worldToCanvas(this.sphereStates[i].point);
                const dist = Math.sqrt(Math.pow(mouseCanvas[0] - sphereCanvas[0], 2) +
                    Math.pow(mouseCanvas[1] - sphereCanvas[1], 2));
                if (dist < this.configuration.grabSpherePixelDistance) {
                    this.draggingSphereIndex = i;
                    element.style.cursor = 'grabbing';
                    const sphereState = this.sphereStates[i];
                    const mouseWorld = viewport.canvasToWorld(mouseCanvas);
                    if (sphereState.isCorner) {
                        this.cornerDragOffset = [
                            sphereState.point[0] - mouseWorld[0],
                            sphereState.point[1] - mouseWorld[1],
                            sphereState.point[2] - mouseWorld[2],
                        ];
                        this.faceDragOffset = null;
                    }
                    else {
                        const directionVector = this._getDirectionVectorForAxis(sphereState.axis);
                        const delta = [
                            sphereState.point[0] - mouseWorld[0],
                            sphereState.point[1] - mouseWorld[1],
                            sphereState.point[2] - mouseWorld[2],
                        ];
                        this.faceDragOffset =
                            delta[0] * directionVector[0] +
                                delta[1] * directionVector[1] +
                                delta[2] * directionVector[2];
                        this.cornerDragOffset = null;
                    }
                    return true;
                }
            }
            const hasSampleDistance = 'getSampleDistance' in mapper || 'getCurrentSampleDistance' in mapper;
            if (!hasSampleDistance) {
                return true;
            }
            const originalSampleDistance = mapper.getSampleDistance();
            if (!this._hasResolutionChanged) {
                const { rotateSampleDistanceFactor } = this.configuration;
                mapper.setSampleDistance(originalSampleDistance * rotateSampleDistanceFactor);
                this._hasResolutionChanged = true;
                if (this.cleanUp !== null) {
                    document.removeEventListener('mouseup', this.cleanUp);
                }
                this.cleanUp = () => {
                    mapper.setSampleDistance(originalSampleDistance);
                    evt.target.style.cursor = '';
                    if (this.draggingSphereIndex !== null) {
                        const sphereState = this.sphereStates[this.draggingSphereIndex];
                        const [viewport3D] = this._getViewportsInfo();
                        const renderingEngine = getRenderingEngine(viewport3D.renderingEngineId);
                        const viewport = renderingEngine.getViewport(viewport3D.viewportId);
                        if (sphereState.isCorner) {
                            this._updateCornerSpheres();
                            this._updateFaceSpheresFromCorners();
                            this._updateClippingPlanesFromFaceSpheres(viewport);
                        }
                    }
                    this.draggingSphereIndex = null;
                    this.cornerDragOffset = null;
                    this.faceDragOffset = null;
                    this.suppressPlaneRotationForCurrentDrag = false;
                    viewport.render();
                    this._hasResolutionChanged = false;
                };
                document.addEventListener('mouseup', this.cleanUp, { once: true });
            }
            return true;
        };
        this._onMouseMoveSphere = (evt) => {
            if (this.draggingSphereIndex === null) {
                return false;
            }
            const sphereState = this.sphereStates[this.draggingSphereIndex];
            if (!sphereState) {
                return false;
            }
            const { viewport, world } = this._getViewportAndWorldCoords(evt);
            if (!viewport || !world) {
                return false;
            }
            if (sphereState.isCorner) {
                const newCorner = this.cornerDragOffset
                    ? vec3.add([0, 0, 0], world, this.cornerDragOffset)
                    : world;
                const oldCorner = sphereState.point;
                const axisFlags = parseCornerKey(sphereState.uid);
                const { xDir, yDir, zDir } = this._getDirectionVectors();
                if (!xDir || !yDir || !zDir)
                    return false;
                const delta = [
                    newCorner[0] - oldCorner[0],
                    newCorner[1] - oldCorner[1],
                    newCorner[2] - oldCorner[2],
                ];
                const deltaX = delta[0] * xDir[0] + delta[1] * xDir[1] + delta[2] * xDir[2];
                const deltaY = delta[0] * yDir[0] + delta[1] * yDir[1] + delta[2] * yDir[2];
                const deltaZ = delta[0] * zDir[0] + delta[1] * zDir[1] + delta[2] * zDir[2];
                if (axisFlags.isXMin) {
                    const faceXMin = this.sphereStates[SPHEREINDEX.XMIN];
                    const newPoint = [
                        faceXMin.point[0] + deltaX * xDir[0],
                        faceXMin.point[1] + deltaX * xDir[1],
                        faceXMin.point[2] + deltaX * xDir[2],
                    ];
                    this._updateSpherePosition(SPHEREINDEX.XMIN, newPoint);
                }
                else if (axisFlags.isXMax) {
                    const faceXMax = this.sphereStates[SPHEREINDEX.XMAX];
                    const newPoint = [
                        faceXMax.point[0] + deltaX * xDir[0],
                        faceXMax.point[1] + deltaX * xDir[1],
                        faceXMax.point[2] + deltaX * xDir[2],
                    ];
                    this._updateSpherePosition(SPHEREINDEX.XMAX, newPoint);
                }
                if (axisFlags.isYMin) {
                    const faceYMin = this.sphereStates[SPHEREINDEX.YMIN];
                    const newPoint = [
                        faceYMin.point[0] + deltaY * yDir[0],
                        faceYMin.point[1] + deltaY * yDir[1],
                        faceYMin.point[2] + deltaY * yDir[2],
                    ];
                    this._updateSpherePosition(SPHEREINDEX.YMIN, newPoint);
                }
                else if (axisFlags.isYMax) {
                    const faceYMax = this.sphereStates[SPHEREINDEX.YMAX];
                    const newPoint = [
                        faceYMax.point[0] + deltaY * yDir[0],
                        faceYMax.point[1] + deltaY * yDir[1],
                        faceYMax.point[2] + deltaY * yDir[2],
                    ];
                    this._updateSpherePosition(SPHEREINDEX.YMAX, newPoint);
                }
                if (axisFlags.isZMin) {
                    const faceZMin = this.sphereStates[SPHEREINDEX.ZMIN];
                    const newPoint = [
                        faceZMin.point[0] + deltaZ * zDir[0],
                        faceZMin.point[1] + deltaZ * zDir[1],
                        faceZMin.point[2] + deltaZ * zDir[2],
                    ];
                    this._updateSpherePosition(SPHEREINDEX.ZMIN, newPoint);
                }
                else if (axisFlags.isZMax) {
                    const faceZMax = this.sphereStates[SPHEREINDEX.ZMAX];
                    const newPoint = [
                        faceZMax.point[0] + deltaZ * zDir[0],
                        faceZMax.point[1] + deltaZ * zDir[1],
                        faceZMax.point[2] + deltaZ * zDir[2],
                    ];
                    this._updateSpherePosition(SPHEREINDEX.ZMAX, newPoint);
                }
                this._updateCornerSpheres();
                this._updateFaceSpheresFromCorners();
            }
            else {
                const directionVector = this._getDirectionVectorForAxis(sphereState.axis);
                const delta = [
                    world[0] - sphereState.point[0],
                    world[1] - sphereState.point[1],
                    world[2] - sphereState.point[2],
                ];
                const distanceAlongAxis = delta[0] * directionVector[0] +
                    delta[1] * directionVector[1] +
                    delta[2] * directionVector[2];
                const adjustedDistance = this.faceDragOffset !== null
                    ? distanceAlongAxis + this.faceDragOffset
                    : distanceAlongAxis;
                const newPoint = [
                    sphereState.point[0] + adjustedDistance * directionVector[0],
                    sphereState.point[1] + adjustedDistance * directionVector[1],
                    sphereState.point[2] + adjustedDistance * directionVector[2],
                ];
                this._updateSpherePosition(this.draggingSphereIndex, newPoint);
                this._updateCornerSpheresFromFaces();
                this._updateFaceSpheresFromCorners();
                this._updateCornerSpheres();
            }
            this._updateClippingPlanesFromFaceSpheres(viewport);
            viewport.render();
            this._notifyClippingPlanesChanged();
            return true;
        };
        this._onControlToolChange = (evt) => {
            const viewport = this._getViewport();
            if (evt.detail.seriesInstanceUID !== this.seriesInstanceUID) {
                return;
            }
            if (evt.detail.clippingPlanes &&
                evt.detail.clippingPlanes.length >= NUM_CLIPPING_PLANES) {
                this.originalClippingPlanes = copyClippingPlanes(evt.detail.clippingPlanes);
                this._updateFaceSpheresFromClippingPlanes();
                this._updateCornerSpheresFromFaces();
                this._updateFaceSpheresFromCorners();
                this._updateCornerSpheres();
                const mapper = this._getVolumeMapper(viewport);
                if (mapper) {
                    this._applyClippingPlanesToMapper(mapper);
                }
                viewport.render();
                this._notifyClippingPlanesChanged(viewport);
            }
            else {
                this._notifyClippingPlanesChanged(viewport);
            }
        };
        this._getViewportsInfo = () => {
            const toolGroup = getToolGroup(this.toolGroupId);
            return toolGroup?.viewportsInfo || [];
        };
        this._initialize3DViewports = (viewportsInfo) => {
            if (!viewportsInfo?.length || !viewportsInfo[0]) {
                console.warn('VolumeCroppingTool: No viewportsInfo available for initialization of volumecroppingtool.');
                return;
            }
            const viewport = this._getViewport();
            if (!viewport) {
                return;
            }
            const volumeActor = this._getVolumeActor(viewport);
            if (!volumeActor) {
                console.warn('VolumeCroppingTool: No volume actors found in the viewport.');
                return;
            }
            const imageData = volumeActor.getMapper().getInputData();
            if (!imageData) {
                console.warn('VolumeCroppingTool: No image data found for volume actor.');
                return;
            }
            this.seriesInstanceUID = imageData.seriesInstanceUID || 'unknown';
            this.volumeDirectionVectors = extractVolumeDirectionVectors(imageData);
            const { xDir, yDir, zDir } = this.volumeDirectionVectors;
            const dimensions = imageData.getDimensions();
            const cropFactor = this.configuration.initialCropFactor || 0.1;
            const xMin = cropFactor * dimensions[0];
            const xMax = (1 - cropFactor) * dimensions[0];
            const yMin = cropFactor * dimensions[1];
            const yMax = (1 - cropFactor) * dimensions[1];
            const zMin = cropFactor * dimensions[2];
            const zMax = (1 - cropFactor) * dimensions[2];
            const xCenter = (xMin + xMax) / 2;
            const yCenter = (yMin + yMax) / 2;
            const zCenter = (zMin + zMax) / 2;
            const faceXMin = imageData.indexToWorld([xMin, yCenter, zCenter]);
            const faceXMax = imageData.indexToWorld([xMax, yCenter, zCenter]);
            const faceYMin = imageData.indexToWorld([xCenter, yMin, zCenter]);
            const faceYMax = imageData.indexToWorld([xCenter, yMax, zCenter]);
            const faceZMin = imageData.indexToWorld([xCenter, yCenter, zMin]);
            const faceZMax = imageData.indexToWorld([xCenter, yCenter, zMax]);
            const planeXMin = vtkPlane.newInstance({
                origin: faceXMin,
                normal: xDir,
            });
            const planeXMax = vtkPlane.newInstance({
                origin: faceXMax,
                normal: [-xDir[0], -xDir[1], -xDir[2]],
            });
            const planeYMin = vtkPlane.newInstance({
                origin: faceYMin,
                normal: yDir,
            });
            const planeYMax = vtkPlane.newInstance({
                origin: faceYMax,
                normal: [-yDir[0], -yDir[1], -yDir[2]],
            });
            const planeZMin = vtkPlane.newInstance({
                origin: faceZMin,
                normal: zDir,
            });
            const planeZMax = vtkPlane.newInstance({
                origin: faceZMax,
                normal: [-zDir[0], -zDir[1], -zDir[2]],
            });
            const planes = [
                planeXMin,
                planeXMax,
                planeYMin,
                planeYMax,
                planeZMin,
                planeZMax,
            ];
            const originalPlanes = planes.map((plane) => ({
                origin: [...plane.getOrigin()],
                normal: [...plane.getNormal()],
            }));
            this.originalClippingPlanes = originalPlanes;
            const diag0 = imageData.indexToWorld([0, 0, 0]);
            const diag1 = imageData.indexToWorld([
                dimensions[0],
                dimensions[1],
                dimensions[2],
            ]);
            const diagonal = vec3.distance(diag0, diag1);
            const adaptiveRadius = calculateAdaptiveSphereRadius(diagonal, this.configuration);
            this._addSphere(viewport, faceXMin, 'x', 'min', null, adaptiveRadius);
            this._addSphere(viewport, faceXMax, 'x', 'max', null, adaptiveRadius);
            this._addSphere(viewport, faceYMin, 'y', 'min', null, adaptiveRadius);
            this._addSphere(viewport, faceYMax, 'y', 'max', null, adaptiveRadius);
            this._addSphere(viewport, faceZMin, 'z', 'min', null, adaptiveRadius);
            this._addSphere(viewport, faceZMax, 'z', 'max', null, adaptiveRadius);
            const corners = [
                imageData.indexToWorld([xMin, yMin, zMin]),
                imageData.indexToWorld([xMin, yMin, zMax]),
                imageData.indexToWorld([xMin, yMax, zMin]),
                imageData.indexToWorld([xMin, yMax, zMax]),
                imageData.indexToWorld([xMax, yMin, zMin]),
                imageData.indexToWorld([xMax, yMin, zMax]),
                imageData.indexToWorld([xMax, yMax, zMin]),
                imageData.indexToWorld([xMax, yMax, zMax]),
            ];
            const cornerKeys = [
                'XMIN_YMIN_ZMIN',
                'XMIN_YMIN_ZMAX',
                'XMIN_YMAX_ZMIN',
                'XMIN_YMAX_ZMAX',
                'XMAX_YMIN_ZMIN',
                'XMAX_YMIN_ZMAX',
                'XMAX_YMAX_ZMIN',
                'XMAX_YMAX_ZMAX',
            ];
            for (let i = 0; i < corners.length; i++) {
                this._addSphere(viewport, corners[i], 'corner', null, cornerKeys[i], adaptiveRadius);
            }
            const edgeCornerPairs = [
                ['XMIN_YMIN_ZMIN', 'XMAX_YMIN_ZMIN'],
                ['XMIN_YMIN_ZMAX', 'XMAX_YMIN_ZMAX'],
                ['XMIN_YMAX_ZMIN', 'XMAX_YMAX_ZMIN'],
                ['XMIN_YMAX_ZMAX', 'XMAX_YMAX_ZMAX'],
                ['XMIN_YMIN_ZMIN', 'XMIN_YMAX_ZMIN'],
                ['XMIN_YMIN_ZMAX', 'XMIN_YMAX_ZMAX'],
                ['XMAX_YMIN_ZMIN', 'XMAX_YMAX_ZMIN'],
                ['XMAX_YMIN_ZMAX', 'XMAX_YMAX_ZMAX'],
                ['XMIN_YMIN_ZMIN', 'XMIN_YMIN_ZMAX'],
                ['XMIN_YMAX_ZMIN', 'XMIN_YMAX_ZMAX'],
                ['XMAX_YMIN_ZMIN', 'XMAX_YMIN_ZMAX'],
                ['XMAX_YMAX_ZMIN', 'XMAX_YMAX_ZMAX'],
            ];
            edgeCornerPairs.forEach(([key1, key2]) => {
                const state1 = this.sphereStates.find((s) => s.uid === `corner_${key1}`);
                const state2 = this.sphereStates.find((s) => s.uid === `corner_${key2}`);
                if (state1 && state2) {
                    const uid = `edge_${key1}_${key2}`;
                    const { actor, source } = addLine3DBetweenPoints(viewport, state1.point, state2.point, [0.7, 0.7, 0.7], uid, this.configuration.showHandles);
                    this.edgeLines[uid] = { actor, source, key1, key2 };
                }
            });
            const mapper = volumeActor.getMapper();
            mapper.addClippingPlane(planeXMin);
            mapper.addClippingPlane(planeXMax);
            mapper.addClippingPlane(planeYMin);
            mapper.addClippingPlane(planeYMax);
            mapper.addClippingPlane(planeZMin);
            mapper.addClippingPlane(planeZMax);
            eventTarget.addEventListener(Events.VOLUMECROPPINGCONTROL_TOOL_CHANGED, (evt) => {
                this._onControlToolChange(evt);
            });
            viewport.render();
        };
        this._getViewportAndWorldCoords = (evt) => {
            const viewport = this._getViewport();
            const x = evt.detail.currentPoints.canvas[0];
            const y = evt.detail.currentPoints.canvas[1];
            const world = viewport.canvasToWorld([x, y]);
            return { viewport, world };
        };
        this._getViewport = () => {
            const viewportsInfo = this._getViewportsInfo();
            if (!viewportsInfo || viewportsInfo.length === 0) {
                return null;
            }
            const [viewport3D] = viewportsInfo;
            const renderingEngine = getRenderingEngine(viewport3D.renderingEngineId);
            const viewport = renderingEngine?.getViewport(viewport3D.viewportId);
            return viewport || null;
        };
        this._onNewVolume = () => {
            const viewportsInfo = this._getViewportsInfo();
            this.originalClippingPlanes = [];
            this.sphereStates = [];
            this.edgeLines = {};
            this._initialize3DViewports(viewportsInfo);
        };
        this._rotateClippingPlanes = (evt) => {
            const { element, currentPoints, lastPoints } = evt.detail;
            const currentPointsCanvas = currentPoints.canvas;
            const lastPointsCanvas = lastPoints.canvas;
            const rotateIncrementDegrees = this.configuration.rotateClippingPlanesIncrementDegrees ??
                this.configuration.rotateIncrementDegrees ??
                5;
            const enabledElement = getEnabledElement(element);
            const { viewport } = enabledElement;
            const camera = getViewportICamera(viewport);
            const width = element.clientWidth;
            const height = element.clientHeight;
            const normalizedPosition = [
                currentPointsCanvas[0] / width,
                currentPointsCanvas[1] / height,
            ];
            const normalizedPreviousPosition = [
                lastPointsCanvas[0] / width,
                lastPointsCanvas[1] / height,
            ];
            const normalizedCenter = [0.5, 0.5];
            const radsq = (1.0 + Math.abs(normalizedCenter[0])) ** 2.0;
            const op = [normalizedPreviousPosition[0], 0, 0];
            const oe = [normalizedPosition[0], 0, 0];
            const opsq = op[0] ** 2;
            const oesq = oe[0] ** 2;
            const lop = opsq > radsq ? 0 : Math.sqrt(radsq - opsq);
            const loe = oesq > radsq ? 0 : Math.sqrt(radsq - oesq);
            const nop = [op[0], 0, lop];
            vtkMath.normalize(nop);
            const noe = [oe[0], 0, loe];
            vtkMath.normalize(noe);
            const dot = vtkMath.dot(nop, noe);
            if (Math.abs(dot) > 0.0001) {
                const angleX = 20 *
                    Math.acos(vtkMath.clampValue(dot, -1.0, 1.0)) *
                    Math.sign(normalizedPosition[0] - normalizedPreviousPosition[0]) *
                    rotateIncrementDegrees;
                const upVec = camera.viewUp;
                const atV = camera.viewPlaneNormal;
                const rightV = [0, 0, 0];
                const forwardV = [0, 0, 0];
                vtkMath.cross(upVec, atV, rightV);
                vtkMath.normalize(rightV);
                vtkMath.cross(atV, rightV, forwardV);
                vtkMath.normalize(forwardV);
                const angleY = 20 *
                    (normalizedPosition[1] - normalizedPreviousPosition[1]) *
                    rotateIncrementDegrees;
                let rotationCenter;
                if (this.sphereStates.length >= NUM_CLIPPING_PLANES) {
                    const faces = [
                        this.sphereStates[SPHEREINDEX.XMIN],
                        this.sphereStates[SPHEREINDEX.XMAX],
                        this.sphereStates[SPHEREINDEX.YMIN],
                        this.sphereStates[SPHEREINDEX.YMAX],
                        this.sphereStates[SPHEREINDEX.ZMIN],
                        this.sphereStates[SPHEREINDEX.ZMAX],
                    ];
                    rotationCenter = [
                        (faces[0].point[0] +
                            faces[1].point[0] +
                            faces[2].point[0] +
                            faces[3].point[0] +
                            faces[4].point[0] +
                            faces[5].point[0]) /
                            NUM_CLIPPING_PLANES,
                        (faces[0].point[1] +
                            faces[1].point[1] +
                            faces[2].point[1] +
                            faces[3].point[1] +
                            faces[4].point[1] +
                            faces[5].point[1]) /
                            NUM_CLIPPING_PLANES,
                        (faces[0].point[2] +
                            faces[1].point[2] +
                            faces[2].point[2] +
                            faces[3].point[2] +
                            faces[4].point[2] +
                            faces[5].point[2]) /
                            NUM_CLIPPING_PLANES,
                    ];
                }
                else {
                    rotationCenter = [0, 0, 0];
                }
                const transformX = mat4.identity(new Float32Array(16));
                mat4.translate(transformX, transformX, rotationCenter);
                mat4.rotate(transformX, transformX, (angleX * Math.PI) / 180, forwardV);
                mat4.translate(transformX, transformX, [
                    -rotationCenter[0],
                    -rotationCenter[1],
                    -rotationCenter[2],
                ]);
                const transformY = mat4.identity(new Float32Array(16));
                mat4.translate(transformY, transformY, rotationCenter);
                mat4.rotate(transformY, transformY, (angleY * Math.PI) / 180, rightV);
                mat4.translate(transformY, transformY, [
                    -rotationCenter[0],
                    -rotationCenter[1],
                    -rotationCenter[2],
                ]);
                const transform = mat4.create();
                mat4.multiply(transform, transformY, transformX);
                const normalTransformX4 = mat4.identity(new Float32Array(16));
                mat4.rotate(normalTransformX4, normalTransformX4, (angleX * Math.PI) / 180, forwardV);
                const normalTransformX = mat3.create();
                mat3.fromMat4(normalTransformX, normalTransformX4);
                const normalTransformY4 = mat4.identity(new Float32Array(16));
                mat4.rotate(normalTransformY4, normalTransformY4, (angleY * Math.PI) / 180, rightV);
                const normalTransformY = mat3.create();
                mat3.fromMat4(normalTransformY, normalTransformY4);
                const normalTransform = mat3.create();
                mat3.multiply(normalTransform, normalTransformY, normalTransformX);
                for (let i = 0; i < this.originalClippingPlanes.length; ++i) {
                    const plane = this.originalClippingPlanes[i];
                    const originVec = vec3.fromValues(plane.origin[0], plane.origin[1], plane.origin[2]);
                    vec3.transformMat4(originVec, originVec, transform);
                    plane.origin = [originVec[0], originVec[1], originVec[2]];
                    const normalVec = vec3.fromValues(plane.normal[0], plane.normal[1], plane.normal[2]);
                    vec3.transformMat3(normalVec, normalVec, normalTransform);
                    vec3.normalize(normalVec, normalVec);
                    plane.normal = [normalVec[0], normalVec[1], normalVec[2]];
                }
                if (this.sphereStates.length >= NUM_CLIPPING_PLANES) {
                    this.sphereStates[SPHEREINDEX.XMIN].point = [
                        ...this.originalClippingPlanes[PLANEINDEX.XMIN].origin,
                    ];
                    this.sphereStates[SPHEREINDEX.XMAX].point = [
                        ...this.originalClippingPlanes[PLANEINDEX.XMAX].origin,
                    ];
                    this.sphereStates[SPHEREINDEX.YMIN].point = [
                        ...this.originalClippingPlanes[PLANEINDEX.YMIN].origin,
                    ];
                    this.sphereStates[SPHEREINDEX.YMAX].point = [
                        ...this.originalClippingPlanes[PLANEINDEX.YMAX].origin,
                    ];
                    this.sphereStates[SPHEREINDEX.ZMIN].point = [
                        ...this.originalClippingPlanes[PLANEINDEX.ZMIN].origin,
                    ];
                    this.sphereStates[SPHEREINDEX.ZMAX].point = [
                        ...this.originalClippingPlanes[PLANEINDEX.ZMAX].origin,
                    ];
                    [
                        SPHEREINDEX.XMIN,
                        SPHEREINDEX.XMAX,
                        SPHEREINDEX.YMIN,
                        SPHEREINDEX.YMAX,
                        SPHEREINDEX.ZMIN,
                        SPHEREINDEX.ZMAX,
                    ].forEach((idx) => {
                        const s = this.sphereStates[idx];
                        s.sphereSource.setCenter(...s.point);
                        s.sphereSource.modified();
                    });
                    const cornerIndices = [
                        SPHEREINDEX.XMIN_YMIN_ZMIN,
                        SPHEREINDEX.XMIN_YMIN_ZMAX,
                        SPHEREINDEX.XMIN_YMAX_ZMIN,
                        SPHEREINDEX.XMIN_YMAX_ZMAX,
                        SPHEREINDEX.XMAX_YMIN_ZMIN,
                        SPHEREINDEX.XMAX_YMIN_ZMAX,
                        SPHEREINDEX.XMAX_YMAX_ZMIN,
                        SPHEREINDEX.XMAX_YMAX_ZMAX,
                    ];
                    cornerIndices.forEach((idx) => {
                        const cornerState = this.sphereStates[idx];
                        if (cornerState) {
                            const cornerVec = vec3.fromValues(cornerState.point[0], cornerState.point[1], cornerState.point[2]);
                            vec3.transformMat4(cornerVec, cornerVec, transform);
                            cornerState.point = [
                                cornerVec[0],
                                cornerVec[1],
                                cornerVec[2],
                            ];
                            cornerState.sphereSource.setCenter(...cornerState.point);
                            cornerState.sphereSource.modified();
                        }
                    });
                    Object.values(this.edgeLines).forEach(({ source, key1, key2 }) => {
                        const state1 = this.sphereStates.find((s) => s.uid === `corner_${key1}`);
                        const state2 = this.sphereStates.find((s) => s.uid === `corner_${key2}`);
                        if (state1 && state2) {
                            const points = source.getPoints();
                            points.setPoint(0, state1.point[0], state1.point[1], state1.point[2]);
                            points.setPoint(1, state2.point[0], state2.point[1], state2.point[2]);
                            points.modified();
                            source.modified();
                        }
                    });
                }
                this._updateClippingPlanes(viewport);
                viewport.render();
                triggerEvent(eventTarget, Events.VOLUMECROPPING_TOOL_CHANGED, {
                    originalClippingPlanes: this.originalClippingPlanes,
                    viewportId: viewport.id,
                    renderingEngineId: viewport.renderingEngineId,
                    seriesInstanceUID: this.seriesInstanceUID,
                });
            }
        };
        this._rotateCamera = (viewport, centerWorld, axis, angle) => {
            const vtkCamera = viewport.getVtkActiveCamera();
            const viewUp = vtkCamera.getViewUp();
            const focalPoint = vtkCamera.getFocalPoint();
            const position = vtkCamera.getPosition();
            const newPosition = [0, 0, 0];
            const newFocalPoint = [0, 0, 0];
            const newViewUp = [0, 0, 0];
            const transform = mat4.identity(new Float32Array(16));
            mat4.translate(transform, transform, centerWorld);
            mat4.rotate(transform, transform, angle, axis);
            mat4.translate(transform, transform, [
                -centerWorld[0],
                -centerWorld[1],
                -centerWorld[2],
            ]);
            vec3.transformMat4(newPosition, position, transform);
            vec3.transformMat4(newFocalPoint, focalPoint, transform);
            mat4.identity(transform);
            mat4.rotate(transform, transform, angle, axis);
            vec3.transformMat4(newViewUp, viewUp, transform);
            setViewportCamera(viewport, {
                position: newPosition,
                viewUp: newViewUp,
                focalPoint: newFocalPoint,
            });
        };
        this.touchDragCallback = this._dragCallback.bind(this);
        this.mouseDragCallback = this._dragCallback.bind(this);
    }
    onSetToolActive() {
        const viewportsInfo = this._getViewportsInfo();
        const subscribeToElementResize = () => {
            viewportsInfo.forEach(({ viewportId, renderingEngineId }) => {
                if (!this._resizeObservers.has(viewportId)) {
                    const { viewport } = getEnabledElementByIds(viewportId, renderingEngineId) || { viewport: null };
                    if (!viewport) {
                        return;
                    }
                    const { element } = viewport;
                    const resizeObserver = new ResizeObserver(() => {
                        const element = getEnabledElementByIds(viewportId, renderingEngineId);
                        if (!element) {
                            return;
                        }
                        const { viewport } = element;
                        const viewPresentation = getViewportPresentation(viewport);
                        resetViewportCamera(viewport);
                        applyViewportPresentation(viewport, viewPresentation);
                        viewport.render();
                    });
                    resizeObserver.observe(element);
                    this._resizeObservers.set(viewportId, resizeObserver);
                }
            });
        };
        subscribeToElementResize();
        this._viewportAddedListener = (evt) => {
            if (evt.detail.toolGroupId === this.toolGroupId) {
                subscribeToElementResize();
            }
        };
        eventTarget.addEventListener(Events.TOOLGROUP_VIEWPORT_ADDED, this._viewportAddedListener);
        this._unsubscribeToViewportNewVolumeSet(viewportsInfo);
        this._subscribeToViewportNewVolumeSet(viewportsInfo);
        if (this.sphereStates && this.sphereStates.length === 0) {
            this.originalClippingPlanes = [];
            this._initialize3DViewports(viewportsInfo);
        }
        this.configuration.showClippingPlanes = false;
        this.configuration.showHandles = false;
        const viewport = this._getViewport();
        if (viewport &&
            this.originalClippingPlanes &&
            this.originalClippingPlanes.length > 0) {
            this._updateClippingPlanes(viewport);
            if (this.sphereStates && this.sphereStates.length > 0) {
                this._updateHandlesVisibility();
            }
            viewport.render();
        }
    }
    onSetToolDisabled() {
        this._resizeObservers.forEach((resizeObserver, viewportId) => {
            resizeObserver.disconnect();
            this._resizeObservers.delete(viewportId);
        });
        if (this._viewportAddedListener) {
            eventTarget.removeEventListener(Events.TOOLGROUP_VIEWPORT_ADDED, this._viewportAddedListener);
            this._viewportAddedListener = null;
        }
        const viewportsInfo = this._getViewportsInfo();
        this._unsubscribeToViewportNewVolumeSet(viewportsInfo);
    }
    setHandlesVisible(visible) {
        this.configuration.showHandles = visible;
        if (visible) {
            this._updateFaceSpheresFromClippingPlanes();
            this._updateCornerSpheres();
        }
        this._updateHandlesVisibility();
        const viewportsInfo = this._getViewportsInfo();
        const [viewport3D] = viewportsInfo;
        const renderingEngine = getRenderingEngine(viewport3D.renderingEngineId);
        const viewport = renderingEngine.getViewport(viewport3D.viewportId);
        viewport.render();
    }
    getHandlesVisible() {
        return this.configuration.showHandles;
    }
    setHandleRadius(radius) {
        this.configuration.sphereRadius = radius;
        this.sphereStates.forEach((state) => {
            if (state?.sphereSource?.setRadius) {
                state.sphereSource.setRadius(radius);
                state.sphereSource.modified();
            }
        });
        const viewportsInfo = this._getViewportsInfo();
        const [viewport3D] = viewportsInfo;
        if (!viewport3D) {
            return;
        }
        const renderingEngine = getRenderingEngine(viewport3D.renderingEngineId);
        const viewport = renderingEngine?.getViewport(viewport3D.viewportId);
        viewport?.render();
    }
    getClippingPlanesVisible() {
        return this.configuration.showClippingPlanes;
    }
    setClippingPlanesVisible(visible) {
        this.configuration.showClippingPlanes = visible;
        const viewport = this._getViewport();
        this._updateClippingPlanes(viewport);
        if (this.sphereStates && this.sphereStates.length > 0) {
            this.configuration.showHandles = visible;
            this._updateHandlesVisibility();
        }
        if (visible &&
            viewport &&
            this.originalClippingPlanes?.length >= NUM_CLIPPING_PLANES) {
            this._notifyClippingPlanesChanged(viewport);
        }
        viewport.render();
    }
    getRotatePlanesOnDrag() {
        return this.rotatePlanesOnDrag;
    }
    setRotatePlanesOnDrag(enable) {
        this.rotatePlanesOnDrag = enable;
        const viewport = this._getViewport();
        if (viewport) {
            viewport.render();
        }
    }
    _dragCallback(evt) {
        const { element, currentPoints, lastPoints } = evt.detail;
        if (this.draggingSphereIndex !== null) {
            this._onMouseMoveSphere(evt);
        }
        else {
            const shiftKey = evt.detail.event?.shiftKey ?? false;
            if ((this.rotatePlanesOnDrag === true || shiftKey) &&
                !this.suppressPlaneRotationForCurrentDrag) {
                this._rotateClippingPlanes(evt);
                return;
            }
            const currentPointsCanvas = currentPoints.canvas;
            const lastPointsCanvas = lastPoints.canvas;
            const { rotateIncrementDegrees } = this.configuration;
            const enabledElement = getEnabledElement(element);
            const { viewport } = enabledElement;
            const camera = getViewportICamera(viewport);
            const width = element.clientWidth;
            const height = element.clientHeight;
            const normalizedPosition = [
                currentPointsCanvas[0] / width,
                currentPointsCanvas[1] / height,
            ];
            const normalizedPreviousPosition = [
                lastPointsCanvas[0] / width,
                lastPointsCanvas[1] / height,
            ];
            const center = [width * 0.5, height * 0.5];
            const centerWorld = viewport.canvasToWorld(center);
            const normalizedCenter = [0.5, 0.5];
            const radsq = (1.0 + Math.abs(normalizedCenter[0])) ** 2.0;
            const op = [normalizedPreviousPosition[0], 0, 0];
            const oe = [normalizedPosition[0], 0, 0];
            const opsq = op[0] ** 2;
            const oesq = oe[0] ** 2;
            const lop = opsq > radsq ? 0 : Math.sqrt(radsq - opsq);
            const loe = oesq > radsq ? 0 : Math.sqrt(radsq - oesq);
            const nop = [op[0], 0, lop];
            vtkMath.normalize(nop);
            const noe = [oe[0], 0, loe];
            vtkMath.normalize(noe);
            const dot = vtkMath.dot(nop, noe);
            if (Math.abs(dot) > 0.0001) {
                const angleX = -2 *
                    Math.acos(vtkMath.clampValue(dot, -1.0, 1.0)) *
                    Math.sign(normalizedPosition[0] - normalizedPreviousPosition[0]) *
                    rotateIncrementDegrees;
                const upVec = camera.viewUp;
                const atV = camera.viewPlaneNormal;
                const rightV = [0, 0, 0];
                const forwardV = [0, 0, 0];
                vtkMath.cross(upVec, atV, rightV);
                vtkMath.normalize(rightV);
                vtkMath.cross(atV, rightV, forwardV);
                vtkMath.normalize(forwardV);
                vtkMath.normalize(upVec);
                this._rotateCamera(viewport, centerWorld, forwardV, angleX);
                const angleY = (normalizedPreviousPosition[1] - normalizedPosition[1]) *
                    rotateIncrementDegrees;
                this._rotateCamera(viewport, centerWorld, rightV, angleY);
            }
            viewport.render();
        }
    }
    _updateClippingPlanes(viewport) {
        const actor = this._getVolumeActor(viewport);
        const mapper = this._getVolumeMapper(viewport);
        if (!actor || !mapper) {
            return;
        }
        const matrix = actor.getMatrix();
        if (!this.configuration.showClippingPlanes) {
            mapper.removeAllClippingPlanes();
            return;
        }
        const rot = mat3.create();
        mat3.fromMat4(rot, matrix);
        const normalMatrix = mat3.create();
        mat3.invert(normalMatrix, rot);
        mat3.transpose(normalMatrix, normalMatrix);
        const originalPlanes = this.originalClippingPlanes;
        if (!originalPlanes || !originalPlanes.length) {
            return;
        }
        mapper.removeAllClippingPlanes();
        const transformedOrigins = [];
        const transformedNormals = [];
        for (let i = 0; i < originalPlanes.length; ++i) {
            const plane = originalPlanes[i];
            const oVec = vec3.create();
            vec3.transformMat4(oVec, new Float32Array(plane.origin), matrix);
            const o = [oVec[0], oVec[1], oVec[2]];
            const nVec = vec3.create();
            vec3.transformMat3(nVec, new Float32Array(plane.normal), normalMatrix);
            vec3.normalize(nVec, nVec);
            const n = [nVec[0], nVec[1], nVec[2]];
            transformedOrigins.push(o);
            transformedNormals.push(n);
        }
        for (let i = 0; i < transformedOrigins.length; ++i) {
            const planeInstance = vtkPlane.newInstance({
                origin: transformedOrigins[i],
                normal: transformedNormals[i],
            });
            mapper.addClippingPlane(planeInstance);
        }
    }
    _updateHandlesVisibility() {
        this.sphereStates.forEach((state) => {
            if (state.sphereActor) {
                state.sphereActor.setVisibility(this.configuration.showHandles);
            }
        });
        Object.values(this.edgeLines).forEach(({ actor }) => {
            if (actor) {
                actor.setVisibility(this.configuration.showHandles);
            }
        });
    }
    _addSphere(viewport, point, axis, position, cornerKey = null, adaptiveRadius) {
        const uid = cornerKey ? `corner_${cornerKey}` : `${axis}_${position}`;
        const sphereState = this.sphereStates.find((s) => s.uid === uid);
        if (sphereState) {
            return;
        }
        const sphereSource = vtkSphereSource.newInstance();
        sphereSource.setCenter(point);
        sphereSource.setRadius(adaptiveRadius);
        const sphereMapper = vtkMapper.newInstance();
        sphereMapper.setInputConnection(sphereSource.getOutputPort());
        const sphereActor = vtkActor.newInstance();
        sphereActor.setMapper(sphereMapper);
        let color = [0.0, 1.0, 0.0];
        const sphereColors = this.configuration.sphereColors || {};
        if (cornerKey) {
            color = sphereColors.CORNERS || [0.0, 0.0, 1.0];
        }
        else if (axis === 'z') {
            color = sphereColors.AXIAL || [1.0, 0.0, 0.0];
        }
        else if (axis === 'x') {
            color = sphereColors.SAGITTAL || [1.0, 1.0, 0.0];
        }
        else if (axis === 'y') {
            color = sphereColors.CORONAL || [0.0, 1.0, 0.0];
        }
        const idx = this.sphereStates.findIndex((s) => s.uid === uid);
        const pointCopy = [point[0], point[1], point[2]];
        if (idx === -1) {
            this.sphereStates.push({
                point: pointCopy,
                axis,
                uid,
                sphereSource,
                sphereActor,
                isCorner: !!cornerKey,
                color,
            });
        }
        else {
            this.sphereStates[idx].point = pointCopy;
            this.sphereStates[idx].sphereSource = sphereSource;
        }
        sphereActor.getProperty().setColor(color);
        sphereActor.setVisibility(this.configuration.showHandles);
        viewport.addActor({ actor: sphereActor, uid: uid });
    }
    _getDirectionVectorForAxis(axis) {
        if (this.originalClippingPlanes &&
            this.originalClippingPlanes.length >= NUM_CLIPPING_PLANES) {
            switch (axis) {
                case 'x':
                    return this.originalClippingPlanes[PLANEINDEX.XMIN]
                        .normal;
                case 'y':
                    return this.originalClippingPlanes[PLANEINDEX.YMIN]
                        .normal;
                case 'z':
                    return this.originalClippingPlanes[PLANEINDEX.ZMIN]
                        .normal;
                default:
                    return [1, 0, 0];
            }
        }
        if (!this.volumeDirectionVectors) {
            if (axis === 'x')
                return [1, 0, 0];
            if (axis === 'y')
                return [0, 1, 0];
            if (axis === 'z')
                return [0, 0, 1];
            return [1, 0, 0];
        }
        switch (axis) {
            case 'x':
                return this.volumeDirectionVectors.xDir;
            case 'y':
                return this.volumeDirectionVectors.yDir;
            case 'z':
                return this.volumeDirectionVectors.zDir;
            default:
                return [1, 0, 0];
        }
    }
    _getVolumeActor(viewport) {
        const vp = viewport || this._getViewport();
        return vp
            ?.getActors?.()
            ?.find((entry) => entry.actor?.getClassName?.() === 'vtkVolume')
            ?.actor;
    }
    _getVolumeMapper(viewport) {
        const actor = this._getVolumeActor(viewport);
        return actor?.getMapper();
    }
    _applyClippingPlanesToMapper(mapper) {
        mapper.removeAllClippingPlanes();
        for (let i = 0; i < NUM_CLIPPING_PLANES; ++i) {
            const plane = vtkPlane.newInstance({
                origin: this.originalClippingPlanes[i].origin,
                normal: this.originalClippingPlanes[i].normal,
            });
            mapper.addClippingPlane(plane);
        }
    }
    _updateClippingPlanesFromFaceSpheres(viewport) {
        const mapper = this._getVolumeMapper(viewport);
        if (!mapper) {
            return;
        }
        this.originalClippingPlanes[PLANEINDEX.XMIN].origin = [
            ...this.sphereStates[SPHEREINDEX.XMIN].point,
        ];
        this.originalClippingPlanes[PLANEINDEX.XMAX].origin = [
            ...this.sphereStates[SPHEREINDEX.XMAX].point,
        ];
        this.originalClippingPlanes[PLANEINDEX.YMIN].origin = [
            ...this.sphereStates[SPHEREINDEX.YMIN].point,
        ];
        this.originalClippingPlanes[PLANEINDEX.YMAX].origin = [
            ...this.sphereStates[SPHEREINDEX.YMAX].point,
        ];
        this.originalClippingPlanes[PLANEINDEX.ZMIN].origin = [
            ...this.sphereStates[SPHEREINDEX.ZMIN].point,
        ];
        this.originalClippingPlanes[PLANEINDEX.ZMAX].origin = [
            ...this.sphereStates[SPHEREINDEX.ZMAX].point,
        ];
        this._applyClippingPlanesToMapper(mapper);
    }
    _updateCornerSpheresFromFaces() {
        const { xDir, yDir, zDir } = this._getDirectionVectors();
        if (!xDir || !yDir || !zDir)
            return;
        const faceXMin = this.sphereStates[SPHEREINDEX.XMIN].point;
        const faceXMax = this.sphereStates[SPHEREINDEX.XMAX].point;
        const faceYMin = this.sphereStates[SPHEREINDEX.YMIN].point;
        const faceYMax = this.sphereStates[SPHEREINDEX.YMAX].point;
        const faceZMin = this.sphereStates[SPHEREINDEX.ZMIN].point;
        const faceZMax = this.sphereStates[SPHEREINDEX.ZMAX].point;
        const corners = [
            {
                key: 'XMIN_YMIN_ZMIN',
                pos: this._calculateCornerFromFaces(faceXMin, faceYMin, faceZMin, xDir, yDir, zDir),
            },
            {
                key: 'XMIN_YMIN_ZMAX',
                pos: this._calculateCornerFromFaces(faceXMin, faceYMin, faceZMax, xDir, yDir, zDir),
            },
            {
                key: 'XMIN_YMAX_ZMIN',
                pos: this._calculateCornerFromFaces(faceXMin, faceYMax, faceZMin, xDir, yDir, zDir),
            },
            {
                key: 'XMIN_YMAX_ZMAX',
                pos: this._calculateCornerFromFaces(faceXMin, faceYMax, faceZMax, xDir, yDir, zDir),
            },
            {
                key: 'XMAX_YMIN_ZMIN',
                pos: this._calculateCornerFromFaces(faceXMax, faceYMin, faceZMin, xDir, yDir, zDir),
            },
            {
                key: 'XMAX_YMIN_ZMAX',
                pos: this._calculateCornerFromFaces(faceXMax, faceYMin, faceZMax, xDir, yDir, zDir),
            },
            {
                key: 'XMAX_YMAX_ZMIN',
                pos: this._calculateCornerFromFaces(faceXMax, faceYMax, faceZMin, xDir, yDir, zDir),
            },
            {
                key: 'XMAX_YMAX_ZMAX',
                pos: this._calculateCornerFromFaces(faceXMax, faceYMax, faceZMax, xDir, yDir, zDir),
            },
        ];
        for (const corner of corners) {
            const stateIndex = this.sphereStates.findIndex((s) => s.uid === `corner_${corner.key}`);
            if (stateIndex !== -1) {
                this._updateSpherePosition(stateIndex, corner.pos);
            }
        }
    }
    _updateFaceSpheresFromCorners() {
        if (!this.volumeDirectionVectors)
            return;
        const corners = [
            this.sphereStates[SPHEREINDEX.XMIN_YMIN_ZMIN].point,
            this.sphereStates[SPHEREINDEX.XMIN_YMIN_ZMAX].point,
            this.sphereStates[SPHEREINDEX.XMIN_YMAX_ZMIN].point,
            this.sphereStates[SPHEREINDEX.XMIN_YMAX_ZMAX].point,
            this.sphereStates[SPHEREINDEX.XMAX_YMIN_ZMIN].point,
            this.sphereStates[SPHEREINDEX.XMAX_YMIN_ZMAX].point,
            this.sphereStates[SPHEREINDEX.XMAX_YMAX_ZMIN].point,
            this.sphereStates[SPHEREINDEX.XMAX_YMAX_ZMAX].point,
        ];
        const faceXMin = this._averagePoints([
            corners[0],
            corners[1],
            corners[2],
            corners[3],
        ]);
        const faceXMax = this._averagePoints([
            corners[4],
            corners[5],
            corners[6],
            corners[7],
        ]);
        const faceYMin = this._averagePoints([
            corners[0],
            corners[1],
            corners[4],
            corners[5],
        ]);
        const faceYMax = this._averagePoints([
            corners[2],
            corners[3],
            corners[6],
            corners[7],
        ]);
        const faceZMin = this._averagePoints([
            corners[0],
            corners[2],
            corners[4],
            corners[6],
        ]);
        const faceZMax = this._averagePoints([
            corners[1],
            corners[3],
            corners[5],
            corners[7],
        ]);
        this._updateSpherePosition(SPHEREINDEX.XMIN, faceXMin);
        this._updateSpherePosition(SPHEREINDEX.XMAX, faceXMax);
        this._updateSpherePosition(SPHEREINDEX.YMIN, faceYMin);
        this._updateSpherePosition(SPHEREINDEX.YMAX, faceYMax);
        this._updateSpherePosition(SPHEREINDEX.ZMIN, faceZMin);
        this._updateSpherePosition(SPHEREINDEX.ZMAX, faceZMax);
    }
    _updateCornerSpheres() {
        const { xDir, yDir, zDir } = this._getDirectionVectors();
        if (!xDir || !yDir || !zDir)
            return;
        const faceXMin = this.sphereStates[SPHEREINDEX.XMIN].point;
        const faceXMax = this.sphereStates[SPHEREINDEX.XMAX].point;
        const faceYMin = this.sphereStates[SPHEREINDEX.YMIN].point;
        const faceYMax = this.sphereStates[SPHEREINDEX.YMAX].point;
        const faceZMin = this.sphereStates[SPHEREINDEX.ZMIN].point;
        const faceZMax = this.sphereStates[SPHEREINDEX.ZMAX].point;
        const corners = [
            {
                key: 'XMIN_YMIN_ZMIN',
                pos: this._calculateCornerFromProjection(faceXMin, faceYMin, faceZMin, xDir, yDir, zDir),
            },
            {
                key: 'XMIN_YMIN_ZMAX',
                pos: this._calculateCornerFromProjection(faceXMin, faceYMin, faceZMax, xDir, yDir, zDir),
            },
            {
                key: 'XMIN_YMAX_ZMIN',
                pos: this._calculateCornerFromProjection(faceXMin, faceYMax, faceZMin, xDir, yDir, zDir),
            },
            {
                key: 'XMIN_YMAX_ZMAX',
                pos: this._calculateCornerFromProjection(faceXMin, faceYMax, faceZMax, xDir, yDir, zDir),
            },
            {
                key: 'XMAX_YMIN_ZMIN',
                pos: this._calculateCornerFromProjection(faceXMax, faceYMin, faceZMin, xDir, yDir, zDir),
            },
            {
                key: 'XMAX_YMIN_ZMAX',
                pos: this._calculateCornerFromProjection(faceXMax, faceYMin, faceZMax, xDir, yDir, zDir),
            },
            {
                key: 'XMAX_YMAX_ZMIN',
                pos: this._calculateCornerFromProjection(faceXMax, faceYMax, faceZMin, xDir, yDir, zDir),
            },
            {
                key: 'XMAX_YMAX_ZMAX',
                pos: this._calculateCornerFromProjection(faceXMax, faceYMax, faceZMax, xDir, yDir, zDir),
            },
        ];
        for (const corner of corners) {
            const stateIndex = this.sphereStates.findIndex((s) => s.uid === `corner_${corner.key}`);
            if (stateIndex !== -1) {
                this._updateSpherePosition(stateIndex, corner.pos);
            }
        }
        this._updateEdgeLines();
    }
    _unsubscribeToViewportNewVolumeSet(viewportsInfo) {
        viewportsInfo.forEach(({ viewportId, renderingEngineId }) => {
            const { viewport } = getEnabledElementByIds(viewportId, renderingEngineId);
            const { element } = viewport;
            element.removeEventListener(Enums.Events.VOLUME_VIEWPORT_NEW_VOLUME, this._onNewVolume);
        });
    }
    _subscribeToViewportNewVolumeSet(viewports) {
        viewports.forEach(({ viewportId, renderingEngineId }) => {
            const { viewport } = getEnabledElementByIds(viewportId, renderingEngineId);
            const { element } = viewport;
            element.addEventListener(Enums.Events.VOLUME_VIEWPORT_NEW_VOLUME, this._onNewVolume);
        });
    }
    _getDirectionVectors() {
        const hasPlanes = this.originalClippingPlanes?.length >= NUM_CLIPPING_PLANES;
        return {
            xDir: hasPlanes
                ? this.originalClippingPlanes[PLANEINDEX.XMIN].normal
                : this.volumeDirectionVectors?.xDir || [1, 0, 0],
            yDir: hasPlanes
                ? this.originalClippingPlanes[PLANEINDEX.YMIN].normal
                : this.volumeDirectionVectors?.yDir || [0, 1, 0],
            zDir: hasPlanes
                ? this.originalClippingPlanes[PLANEINDEX.ZMIN].normal
                : this.volumeDirectionVectors?.zDir || [0, 0, 1],
        };
    }
    _updateSpherePosition(sphereIndex, newPoint) {
        const state = this.sphereStates[sphereIndex];
        if (state) {
            state.point = newPoint;
            state.sphereSource.setCenter(...newPoint);
            state.sphereSource.modified();
        }
    }
    _updateFaceSpheresFromClippingPlanes() {
        const faceMappings = [
            { sphereIdx: SPHEREINDEX.XMIN, planeIdx: PLANEINDEX.XMIN },
            { sphereIdx: SPHEREINDEX.XMAX, planeIdx: PLANEINDEX.XMAX },
            { sphereIdx: SPHEREINDEX.YMIN, planeIdx: PLANEINDEX.YMIN },
            { sphereIdx: SPHEREINDEX.YMAX, planeIdx: PLANEINDEX.YMAX },
            { sphereIdx: SPHEREINDEX.ZMIN, planeIdx: PLANEINDEX.ZMIN },
            { sphereIdx: SPHEREINDEX.ZMAX, planeIdx: PLANEINDEX.ZMAX },
        ];
        faceMappings.forEach(({ sphereIdx, planeIdx }) => {
            const newPoint = [
                ...this.originalClippingPlanes[planeIdx].origin,
            ];
            this._updateSpherePosition(sphereIdx, newPoint);
        });
    }
    _averagePoints(points) {
        const sum = points.reduce((acc, p) => [acc[0] + p[0], acc[1] + p[1], acc[2] + p[2]], [0, 0, 0]);
        return [
            sum[0] / points.length,
            sum[1] / points.length,
            sum[2] / points.length,
        ];
    }
    _notifyClippingPlanesChanged(viewport) {
        const eventData = {
            originalClippingPlanes: this.originalClippingPlanes,
            seriesInstanceUID: this.seriesInstanceUID,
        };
        if (viewport) {
            eventData.viewportId = viewport.id;
            eventData.renderingEngineId = viewport.renderingEngineId;
        }
        triggerEvent(eventTarget, Events.VOLUMECROPPING_TOOL_CHANGED, eventData);
    }
    _calculateCornerFromFaces(faceX, faceY, faceZ, xDir, yDir, zDir) {
        const deltaXY = [
            faceY[0] - faceX[0],
            faceY[1] - faceX[1],
            faceY[2] - faceX[2],
        ];
        const distY = deltaXY[0] * yDir[0] + deltaXY[1] * yDir[1] + deltaXY[2] * yDir[2];
        const deltaXZ = [
            faceZ[0] - faceX[0],
            faceZ[1] - faceX[1],
            faceZ[2] - faceX[2],
        ];
        const distZ = deltaXZ[0] * zDir[0] + deltaXZ[1] * zDir[1] + deltaXZ[2] * zDir[2];
        return [
            faceX[0] + distY * yDir[0] + distZ * zDir[0],
            faceX[1] + distY * yDir[1] + distZ * zDir[1],
            faceX[2] + distY * yDir[2] + distZ * zDir[2],
        ];
    }
    _calculateCornerFromProjection(faceX, faceY, faceZ, xDir, yDir, zDir) {
        const dX = faceX[0] * xDir[0] + faceX[1] * xDir[1] + faceX[2] * xDir[2];
        const dY = faceY[0] * yDir[0] + faceY[1] * yDir[1] + faceY[2] * yDir[2];
        const dZ = faceZ[0] * zDir[0] + faceZ[1] * zDir[1] + faceZ[2] * zDir[2];
        return [
            dX * xDir[0] + dY * yDir[0] + dZ * zDir[0],
            dX * xDir[1] + dY * yDir[1] + dZ * zDir[1],
            dX * xDir[2] + dY * yDir[2] + dZ * zDir[2],
        ];
    }
    _updateEdgeLines() {
        Object.values(this.edgeLines).forEach(({ source, key1, key2 }) => {
            const state1 = this.sphereStates.find((s) => s.uid === `corner_${key1}`);
            const state2 = this.sphereStates.find((s) => s.uid === `corner_${key2}`);
            if (state1 && state2) {
                const points = source.getPoints();
                points.setPoint(0, state1.point[0], state1.point[1], state1.point[2]);
                points.setPoint(1, state2.point[0], state2.point[1], state2.point[2]);
                points.modified();
                source.modified();
            }
        });
    }
}
VolumeCroppingTool.toolName = 'VolumeCropping';
export default VolumeCroppingTool;
