// @cornerstonejs/tools/dist/esm/synchronizers/callbacks/zoomPanSyncCallback.js
import { getRenderingEngine, viewportHasPan, viewportHasZoom, viewportProjection, } from '@cornerstonejs/core';
const zoomPanProjectionSelector = {
    pan: true,
    zoom: true,
};
function viewportHasViewStateSetter(viewport) {
    return (Boolean(viewport) &&
        typeof viewport === 'object' &&
        'setViewState' in viewport &&
        typeof viewport.setViewState === 'function');
}
function syncProjectionZoomPan(sourceViewport, targetViewport, options) {
    if (!viewportHasViewStateSetter(targetViewport)) {
        return false;
    }
    const sourcePresentation = viewportProjection.getPresentation(sourceViewport, {
        selector: zoomPanProjectionSelector,
    });
    if (!sourcePresentation) {
        return false;
    }
    const presentationPatch = {};
    if (options?.syncZoom !== false &&
        typeof sourcePresentation.zoom === 'number') {
        presentationPatch.zoom = sourcePresentation.zoom;
    }
    if (options?.syncPan !== false && sourcePresentation.pan) {
        presentationPatch.pan = sourcePresentation.pan;
    }
    if (presentationPatch.zoom === undefined &&
        presentationPatch.pan === undefined) {
        return false;
    }
    const nextViewState = viewportProjection.withPresentation(targetViewport, presentationPatch);
    if (!nextViewState) {
        return false;
    }
    targetViewport.setViewState(nextViewState);
    return true;
}
export default function zoomPanSyncCallback(synchronizerInstance, sourceViewport, targetViewport) {
    const renderingEngine = getRenderingEngine(targetViewport.renderingEngineId);
    if (!renderingEngine) {
        throw new Error(`No RenderingEngine for Id: ${targetViewport.renderingEngineId}`);
    }
    const options = synchronizerInstance.getOptions(targetViewport.viewportId);
    const tViewport = renderingEngine.getViewport(targetViewport.viewportId);
    const sViewport = renderingEngine.getViewport(sourceViewport.viewportId);
    const syncedByProjection = syncProjectionZoomPan(sViewport, tViewport, options);
    if (syncedByProjection) {
        tViewport.render();
        return;
    }
    if (options?.syncZoom !== false &&
        viewportHasZoom(sViewport) &&
        viewportHasZoom(tViewport)) {
        const srcZoom = sViewport.getZoom();
        tViewport.setZoom(srcZoom);
    }
    if (options?.syncPan !== false &&
        viewportHasPan(sViewport) &&
        viewportHasPan(tViewport)) {
        const srcPan = sViewport.getPan();
        tViewport.setPan(srcPan);
    }
    tViewport.render();
}
