// @cornerstonejs/tools/dist/esm/tools/displayTools/Labelmap/labelmapRenderPlan/stackImagePlan.js
import { getCurrentLabelmapImageIdsForViewport } from '../../../../stateManagement/segmentation/getCurrentLabelmapImageIdForViewport';
import { triggerSegmentationDataModified } from '../../../../stateManagement/segmentation/triggerSegmentationEvents';
import { syncStackLabelmapActors } from '../syncStackLabelmapActors';
import { createLabelmapRepresentationUID } from '../labelmapRepresentationUID';
import { createLabelmapRenderPlan } from './createLabelmapRenderPlan';
function createStackImageLabelmapPlan({ isVolumeImageMapper, renderMode, segmentationId, useSliceRendering, viewport, }) {
    return createLabelmapRenderPlan({
        isVolumeImageMapper,
        kind: 'legacy-stack-image',
        renderMode,
        segmentationId,
        updateAfterMount: false,
        useSliceRendering,
        viewport,
        canRenderCurrentViewport: () => hasCurrentStackLabelmapImageIds(viewport, segmentationId),
        getExpectedRepresentationUIDs: () => getExpectedStackLabelmapRepresentationUIDs(viewport, segmentationId),
        mount: () => mountStackLabelmapActors(viewport, segmentationId),
        update: () => syncStackLabelmapActors(viewport, segmentationId),
    });
}
function hasCurrentStackLabelmapImageIds(viewport, segmentationId) {
    return !!getCurrentLabelmapImageIdsForViewport(viewport.id, segmentationId)
        ?.length;
}
function getExpectedStackLabelmapRepresentationUIDs(viewport, segmentationId) {
    return (getCurrentLabelmapImageIdsForViewport(viewport.id, segmentationId)?.map((imageId) => createLabelmapRepresentationUID({
        segmentationId,
        referencedId: imageId,
    })) ?? []);
}
async function mountStackLabelmapActors(viewport, segmentationId) {
    syncStackLabelmapActors(viewport, segmentationId);
    triggerSegmentationDataModified(segmentationId);
}
export { createStackImageLabelmapPlan };
