// @cornerstonejs/tools/dist/esm/stateManagement/segmentation/segmentIndex.js
import { getToolGroupForViewport } from '../../store/ToolGroupManager';
import { invalidateBrushCursor } from '../../utilities/segmentation/invalidateBrushCursor';
import { getSegmentation } from './getSegmentation';
import { getViewportIdsWithSegmentation } from './getViewportIdsWithSegmentation';
import { triggerSegmentationModified } from './triggerSegmentationEvents';
import { getActiveSegmentIndex } from './getActiveSegmentIndex';
import { getSegmentationRepresentations } from './getSegmentationRepresentation';
import { ensureLabelmapState, getSegmentBinding, getLabelmaps, setSegmentBinding, syncLegacyLabelmapData, } from './helpers/labelmapSegmentationState';
function setActiveSegmentIndex(segmentationId, segmentIndex) {
    const segmentation = getSegmentation(segmentationId);
    if (typeof segmentIndex === 'string') {
        console.warn('segmentIndex is a string, converting to number');
        segmentIndex = Number(segmentIndex);
    }
    Object.values(segmentation.segments).forEach((segment) => {
        segment.active = false;
    });
    if (!segmentation.segments[segmentIndex]) {
        segmentation.segments[segmentIndex] = {
            segmentIndex,
            label: '',
            locked: false,
            cachedStats: {},
            active: false,
        };
        if (segmentation.representationData?.Labelmap) {
            ensureLabelmapState(segmentation);
            const primaryLayer = getLabelmaps(segmentation)[0];
            if (primaryLayer) {
                setSegmentBinding(segmentation, segmentIndex, {
                    labelmapId: primaryLayer.labelmapId,
                    labelValue: segmentIndex,
                });
                syncLegacyLabelmapData(segmentation);
            }
        }
    }
    if (segmentation.representationData?.Labelmap &&
        !getSegmentBinding(segmentation, segmentIndex)) {
        ensureLabelmapState(segmentation);
        const primaryLayer = getLabelmaps(segmentation)[0];
        if (primaryLayer) {
            setSegmentBinding(segmentation, segmentIndex, {
                labelmapId: primaryLayer.labelmapId,
                labelValue: segmentIndex,
            });
            syncLegacyLabelmapData(segmentation);
        }
    }
    if (segmentation.segments[segmentIndex].active !== true) {
        segmentation.segments[segmentIndex].active = true;
        triggerSegmentationModified(segmentationId);
    }
    const viewportIds = getViewportIdsWithSegmentation(segmentationId);
    viewportIds.forEach((viewportId) => {
        const representations = getSegmentationRepresentations(viewportId, {
            segmentationId,
        });
        representations.forEach((representation) => {
            if (!representation.segments[segmentIndex]) {
                representation.segments[segmentIndex] = {
                    visible: true,
                };
            }
        });
    });
    viewportIds.forEach((viewportId) => {
        const toolGroup = getToolGroupForViewport(viewportId);
        if (!toolGroup) {
            return;
        }
        invalidateBrushCursor(toolGroup.id);
    });
}
export { setActiveSegmentIndex, getActiveSegmentIndex };
