// @cornerstonejs/tools/dist/esm/stateManagement/segmentation/labelmapModel/labelmapSegmentBindings.js
import { ensureLabelmapState } from './normalizeLabelmapSegmentationData';
import { getLabelmap, getLabelmaps, removeLabelmap, } from './labelmapLayerStore';
import { syncLegacyLabelmapData } from './labelmapLegacyAdapter';
function getSegmentBinding(segmentation, segmentIndex) {
    return ensureLabelmapState(segmentation)?.segmentBindings?.[segmentIndex];
}
function setSegmentBinding(segmentation, segmentIndex, binding) {
    const labelmapState = ensureLabelmapState(segmentation);
    if (!labelmapState) {
        return;
    }
    const previousBinding = labelmapState.segmentBindings[segmentIndex];
    if (previousBinding) {
        const previousLayer = labelmapState.labelmaps[previousBinding.labelmapId];
        if (previousLayer?.labelToSegmentIndex) {
            delete previousLayer.labelToSegmentIndex[previousBinding.labelValue];
        }
    }
    labelmapState.segmentBindings[segmentIndex] = binding;
    const layer = labelmapState.labelmaps[binding.labelmapId];
    if (layer) {
        layer.labelToSegmentIndex ||= {};
        layer.labelToSegmentIndex[binding.labelValue] = segmentIndex;
    }
}
function getSegmentsOnLabelmap(segmentation, labelmapId) {
    const labelmapState = ensureLabelmapState(segmentation);
    if (!labelmapState) {
        return [];
    }
    return Object.entries(labelmapState.segmentBindings)
        .filter(([, binding]) => binding.labelmapId === labelmapId)
        .map(([segmentIndex]) => Number(segmentIndex))
        .sort((a, b) => a - b);
}
function getSegmentIndexForLabelValue(segmentation, labelmapId, labelValue) {
    const layer = getLabelmap(segmentation, labelmapId);
    if (!layer || labelValue == null) {
        return;
    }
    return layer.labelToSegmentIndex?.[labelValue] ?? labelValue;
}
function getLabelValueForSegment(segmentation, segmentIndex) {
    return (getSegmentBinding(segmentation, segmentIndex)?.labelValue ?? segmentIndex);
}
function getLabelmapForSegment(segmentation, segmentIndex) {
    const binding = getSegmentBinding(segmentation, segmentIndex);
    if (!binding) {
        return;
    }
    return getLabelmap(segmentation, binding.labelmapId);
}
function resolveLabelmapForSegment(segmentation, segmentIndex) {
    return (getLabelmapForSegment(segmentation, segmentIndex) ??
        getLabelmaps(segmentation)[0]);
}
function removeSegmentBinding(segmentation, segmentIndex) {
    const labelmapState = ensureLabelmapState(segmentation);
    const binding = labelmapState?.segmentBindings?.[segmentIndex];
    if (!labelmapState || !binding) {
        return;
    }
    const layer = labelmapState.labelmaps[binding.labelmapId];
    if (layer?.labelToSegmentIndex) {
        delete layer.labelToSegmentIndex[binding.labelValue];
    }
    delete labelmapState.segmentBindings[segmentIndex];
    if (getSegmentsOnLabelmap(segmentation, binding.labelmapId).length === 0) {
        removeLabelmap(segmentation, binding.labelmapId);
    }
    syncLegacyLabelmapData(segmentation);
}
export { getLabelmapForSegment, getLabelValueForSegment, getSegmentBinding, getSegmentIndexForLabelValue, getSegmentsOnLabelmap, removeSegmentBinding, resolveLabelmapForSegment, setSegmentBinding, };
