// @cornerstonejs/tools/dist/esm/stateManagement/segmentation/labelmapModel/normalizeLabelmapSegmentationData.js
const SOURCE_REPRESENTATION_NAME = 'binaryLabelmap';
function getSegmentOrder(segmentation) {
    if (segmentation.segmentOrder?.length) {
        return [...segmentation.segmentOrder];
    }
    return Object.keys(segmentation.segments)
        .map(Number)
        .sort((a, b) => a - b);
}
function getPrimaryLabelmapId(segmentationId) {
    return `${segmentationId}-storage-0`;
}
function getPrimaryLabelmapType(labelmapData) {
    return labelmapData.volumeId ? 'volume' : 'stack';
}
function createPrimaryLabelmapLayer(segmentation, labelmapData, labelmapId = getPrimaryLabelmapId(segmentation.segmentationId)) {
    const layer = {
        labelmapId,
        storageKind: getPrimaryLabelmapType(labelmapData),
        labelToSegmentIndex: {},
    };
    if (labelmapData.volumeId != null) {
        layer.volumeId = labelmapData.volumeId;
    }
    if (labelmapData.referencedVolumeId != null) {
        layer.referencedVolumeId = labelmapData.referencedVolumeId;
    }
    if (labelmapData.referencedImageIds != null) {
        layer.referencedImageIds = labelmapData.referencedImageIds;
    }
    if (labelmapData.imageIds != null) {
        layer.imageIds = labelmapData.imageIds;
    }
    return layer;
}
function resolvePrimaryLabelmapId(segmentation, labelmapData) {
    const storedLabelmapId = labelmapData.primaryLabelmapId;
    if (storedLabelmapId && labelmapData.labelmaps?.[storedLabelmapId]) {
        return storedLabelmapId;
    }
    const fallbackLabelmapId = Object.keys(labelmapData.labelmaps ?? {})[0] ??
        getPrimaryLabelmapId(segmentation.segmentationId);
    labelmapData.primaryLabelmapId = fallbackLabelmapId;
    return fallbackLabelmapId;
}
function ensureLabelmapState(segmentation) {
    const labelmapData = segmentation.representationData.Labelmap;
    if (!labelmapData) {
        return;
    }
    labelmapData.labelmaps ||= {};
    const primaryLabelmapId = resolvePrimaryLabelmapId(segmentation, labelmapData);
    labelmapData.labelmaps[primaryLabelmapId] ||= createPrimaryLabelmapLayer(segmentation, labelmapData, primaryLabelmapId);
    labelmapData.segmentBindings ||= {};
    labelmapData.sourceRepresentationName ||= SOURCE_REPRESENTATION_NAME;
    getSegmentOrder(segmentation).forEach((segmentIndex) => {
        labelmapData.segmentBindings[segmentIndex] ||= {
            labelmapId: primaryLabelmapId,
            labelValue: segmentIndex,
        };
    });
    Object.values(labelmapData.labelmaps).forEach((layer) => {
        layer.labelToSegmentIndex = {};
    });
    Object.entries(labelmapData.segmentBindings).forEach(([segmentIndex, binding]) => {
        const layer = labelmapData.labelmaps[binding.labelmapId];
        if (!layer) {
            return;
        }
        layer.labelToSegmentIndex ||= {};
        layer.labelToSegmentIndex[binding.labelValue] = Number(segmentIndex);
    });
    return labelmapData;
}
export { SOURCE_REPRESENTATION_NAME, ensureLabelmapState, getPrimaryLabelmapId, getSegmentOrder, };
