// @dental/web/src/lib/aiOrchestrator.ts
import { smartBookingParser } from './smartBookingParser';
import { parseVisitDictationLocal } from './smartVisitParser';
import { parsePatientDictationLocal } from './smartPatientParser';
import { Prompts } from './aiPromptRouter';

import { parsePriceDictationLocal } from './smartPriceParser';

export function parsePaymentDictationLocal(input: string): { amount: string, method: string | null, taxDeductionCode: string | null } {
  const lower = input.toLowerCase();
  
  // Extract amount
  let amount = "";
  const thousandsMatch = lower.match(/(\d+)[\s]*(褌褘褋|褌褘褋褟褔|褌\.褉\.|褌褉)/i);
  if (thousandsMatch && thousandsMatch[1]) {
    amount = String(parseInt(thousandsMatch[1], 10) * 1000);
  } else {
    const match = input.replace(/\s+/g, '').match(/\d{2,7}/);
    if (match) amount = match[0];
  }
  
  // Extract method
  let method: string | null = null;
  if (/(泻邪褉褌|褌械褉屑懈薪邪谢|斜械蟹薪邪谢)/i.test(lower)) method = "card";
  else if (/(薪邪谢懈褔|薪邪谢)/i.test(lower)) method = "cash";
  else if (/(褋斜锌|锌械褉械胁芯写|斜邪薪泻|qr)/i.test(lower)) method = "bank_transfer";
  else if (/(芯薪谢邪泄薪|褋褋褘谢泻)/i.test(lower)) method = "online";
  
  // Extract tax deduction
  let taxDeductionCode: string | null = null;
  if (/(胁褘褔械褌|泻芯写 1|薪邪谢芯谐)/i.test(lower)) taxDeductionCode = "1";
  else if (/(泻芯写 2)/i.test(lower)) taxDeductionCode = "2";
  
  return { amount, method, taxDeductionCode };
}
export type AiIntent = 
  | "schedule_appointment"
  | "fill_emk"
  | "parse_patient_document"
  | "manage_prices"
  | "clinical_audit"
  | "imaging_analysis"
  | "patient_communication"
  | "unknown";

export interface AiRouterResult<T> {
  source: "local_algorithm" | "llm_required";
  data?: T;
  suggestedPrompt?: string;
  systemPrompt?: string;
}

/**
 * AI Orchestrator: Smartly routes tasks between fast local parsing and heavy LLM usage.
 * Protects API limits (Local First pattern) and provides rich prompts when LLM is needed.
 */
export class AiOrchestrator {
  
  /**
   * Determine the intent of the text (NLP Routing)
   */
  static detectIntent(input: string): AiIntent {
    const lower = input.toLowerCase();
    
    // 0. Check for Price management
    if (/(写芯斜邪胁褜.*胁 锌褉邪泄褋|褍褋谢褍谐褍|褑械薪邪|褋褌芯懈屑芯褋褌褜.*褉褍斜|锌褉邪泄褋)/i.test(lower) && /\d/.test(lower)) {
      return "manage_prices";
    }

    // 1. Check for strong temporal markers and booking actions (Schedule Priority)
    const scheduleMatch = /(蟹邪锌懈褕|锌褉懈械屑|褉邪褋锌懈褋邪薪|蟹邪锌懈褋褜|锌械褉械薪械褋|锌械褉械蟹邪锌懈褕|芯褌屑械薪|褍写邪谢懈|褍斜械褉懈 蟹邪锌懈褋褜)/i.test(lower);
    const timeMatch = /(薪邪 蟹邪胁褌褉邪|薪邪 褋械谐芯写薪褟|胁 \d{1,2}:\d{2}|褋 \d{1,2}|胁 \d{1,2} 褔邪褋|薪邪 \d{1,2} 褔邪褋|褔械褉械蟹 薪械写械谢褞|锌芯褋谢械蟹邪胁褌褉邪)/i.test(lower);
    
    // If it has both a patient indicator ("薪芯胁褘泄 锌邪褑懈械薪褌", name) + time, or strong booking verb -> It's Schedule
    const isSchedule = scheduleMatch || (timeMatch && /(锌邪褑懈械薪褌|泻 胁褉邪褔褍|泻 褏懈褉褍褉谐褍|泻 褌械褉邪锌械胁褌褍|薪邪)/i.test(lower));
    if (isSchedule) return "schedule_appointment";
    
    // 2. Check for clinical / medical record keywords (EMK)
    const isEmk = /(锌械褉胁懈褔薪褘泄 芯褋屑芯褌褉|褋芯蟹写邪泄 泻邪褉褌褍|卸邪谢芯斜|写懈邪谐薪芯蟹|芯斜褗械泻褌懈胁薪芯|谢械褔械薪懈械|斜芯谢褜|锌褍谢褜锌懈褌|泻邪褉懈械褋|锌械褉懈芯写芯薪褌懈褌|蟹褍斜|褝泻褋褌懈褉锌邪褑|锌谢芯屑斜|褝薪写芯|泻邪薪邪谢|胁褋泻褉褘谢|褉械薪褌谐械薪)/i.test(lower);
    // Note: If they mention imaging but it's clearly an EMK dictation, EMK wins unless they explicitly ask for imaging analysis
    const isImagingOnly = /(泻谢泻褌|褉械薪褌谐械薪|锌邪薪芯褉邪屑|褋薪懈屑芯泻|泻懈褋褌|泻\/褌|褉械褌械薪褑|屑褉褌)/i.test(lower) && !/(卸邪谢芯斜|写懈邪谐薪芯蟹|芯斜褗械泻褌懈胁薪芯|谢械褔械薪懈械|斜芯谢褜|锌褍谢褜锌懈褌|泻邪褉懈械褋|锌械褉懈芯写芯薪褌懈褌|褝泻褋褌懈褉锌邪褑|锌谢芯屑斜)/i.test(lower);
    
    if (isImagingOnly) return "imaging_analysis";
    if (isEmk) return "fill_emk";
    
    // 3. Clinical Audit 
    if (/(锌褉芯胁械褉褜 泻邪褉褌褍|邪褍写懈褌|褞褉懈写懈褔械褋泻|泻芯褋褟泻|写芯泻褍屑械薪褌|褋芯谐谢邪褋懈|芯褕懈斜泻|锌褉芯褕谢褘泄 屑械褋褟褑)/i.test(lower)) return "clinical_audit";
    
    // 4. Patient Communication
    if (/(薪邪锌懈褕懈|薪邪锌芯屑薪懈|芯褌蟹褘胁|褌械谢械谐褉邪屑|褋芯芯斜褖械薪|褋屑褋|whatsapp|锌芯蟹胁芯薪懈)/i.test(lower)) return "patient_communication";
    
    // 5. Patient Document Extraction
    if (/(锌邪褋锌芯褉褌|邪薪泻械褌|胁褘写邪薪|褌械谢械褎芯薪|写邪褌邪 褉芯卸写械薪|褋薪懈谢褋|懈薪薪|褎懈芯|锌芯谢懈褋)/i.test(lower)) return "parse_patient_document";
    
    return "unknown";
  }

  /**
   * Route EMK Dictation
   * Local algorithm first. If it finds tooth or any field, it succeeds. Otherwise fallback to LLM.
   */
  static processEmkDictation(input: string): AiRouterResult<ReturnType<typeof parseVisitDictationLocal>> {
    const localResult = parseVisitDictationLocal(input);
    const hasTooth = localResult.toothUpdates.length > 0;
    const hasEmkFields = !!(localResult.emkUpdates.complaint || localResult.emkUpdates.objectiveStatus || localResult.emkUpdates.treatmentPlan || localResult.emkUpdates.diagnosis);
    
    if (hasTooth || hasEmkFields) {
      return { source: "local_algorithm", data: localResult };
    }

    return { 
      source: "llm_required", 
      systemPrompt: `${Prompts.System.Base}\n${Prompts.System.StrictJSON}`,
      suggestedPrompt: Prompts.Medical.StructureEmk(input)
    };
  }

  /**
   * Route Schedule Booking
   */
  static processScheduleBooking(input: string, dashboard: any): AiRouterResult<ReturnType<typeof smartBookingParser>> {
    const localResult = smartBookingParser(input, dashboard);
    
    // Confident match requires finding an existing patient, a new patient name, OR at least time/date + reason
    if (localResult.patientId || localResult.patientName || (localResult.startsAt && localResult.reason)) {
      return { source: "local_algorithm", data: localResult };
    }

    return {
      source: "llm_required",
      systemPrompt: `${Prompts.System.Base}\n${Prompts.System.StrictJSON}`,
      suggestedPrompt: Prompts.Schedule.AnalyzeNote(input)
    };
  }

  /**
   * Route Patient Info Extraction
   */
  static processPatientInfo(input: string): AiRouterResult<ReturnType<typeof parsePatientDictationLocal>> {
    const localResult = parsePatientDictationLocal(input);
    
    if (localResult.fullName || localResult.phone || localResult.birthDate) {
      return { source: "local_algorithm", data: localResult };
    }

    return {
      source: "llm_required",
      systemPrompt: `${Prompts.System.Base}\n${Prompts.System.StrictJSON}`,
      suggestedPrompt: Prompts.Patient.ExtractDetails(input)
    };
  }

  /**
   * Route Price Management Dictation
   */
  static processPriceDictation(input: string): AiRouterResult<ReturnType<typeof parsePriceDictationLocal>> {
    const localResult = parsePriceDictationLocal(input);
    
    // Confident match requires finding a service name and a price
    if (localResult.serviceName && localResult.price) {
      return { source: "local_algorithm", data: localResult };
    }

    return {
      source: "llm_required",
      systemPrompt: `${Prompts.System.Base}\n${Prompts.System.StrictJSON}`,
      suggestedPrompt: "Extract serviceName, price (number), and category from: " + input
    };
  }

  /**
   * Route Clinical Audit (Pure AI Task)
   */
  static processClinicalAudit(input: string): AiRouterResult<null> {
    return {
      source: "llm_required",
      systemPrompt: `${Prompts.System.Base}\n${Prompts.System.StrictJSON}`,
      suggestedPrompt: Prompts.Medical.ClinicalAudit(input)
    };
  }

  /**
   * Route Imaging Report Analysis (Pure AI Task)
   */
  static processImagingAnalysis(input: string): AiRouterResult<null> {
    return {
      source: "llm_required",
      systemPrompt: `${Prompts.System.Base}\n${Prompts.System.StrictJSON}`,
      suggestedPrompt: Prompts.Imaging.AnalyzeCTReport(input)
    };
  }

  /**
   * Route Patient Communication (Pure AI Task)
   */
  static processPatientCommunication(patientName: string, recentProcedure: string): AiRouterResult<null> {
    return {
      source: "llm_required",
      systemPrompt: `${Prompts.System.Base}\n${Prompts.System.StrictJSON}`,
      suggestedPrompt: Prompts.Marketing.PatientFollowUp(patientName, recentProcedure)
    };
  }


  /**
   * Route Marketing Review Generation (Pure AI Task)
   */
  static processMarketingReview(reviewText: string, tone: string, clinicName: string, seoKeys: string[]): AiRouterResult<null> {
    return {
      source: "llm_required",
      systemPrompt: `${Prompts.System.Base}\n${Prompts.System.StrictJSON}`,
      suggestedPrompt: Prompts.Marketing.GenerateReviewReply(reviewText, tone, clinicName, seoKeys)
    };
  }


  /**
   * Route Payment / Finance Dictation
   */
  static processPaymentDictation(input: string): AiRouterResult<ReturnType<typeof parsePaymentDictationLocal>> {
    const localResult = parsePaymentDictationLocal(input);
    return {
      source: "local_algorithm",
      data: localResult
    };
  }

  /**
   * Parses global voice navigation commands (e.g. view switching, searching, date filtering).
   */
  static parseGlobalNavigation(input: string): {
    view?: string;
    query?: string;
    date?: string;
    feedbackText: string;
  } {
    const lower = input.toLowerCase().trim().replace(/[.,!?]/g, "");
    
    // 1. Check for search query (e.g. "薪邪泄写懈 锌邪褑懈械薪褌邪 袠胁邪薪芯胁", "芯褌泻褉芯泄 泻邪褉褌褍 袩械褌褉芯胁邪")
    const searchMatch = lower.match(/^(?:薪邪泄写懈|芯褌泻褉芯泄|懈褖懈|锌芯懈褋泻|薪邪泄写懈 锌邪褑懈械薪褌邪|芯褌泻褉芯泄 泻邪褉褌褍|芯褌泻褉芯泄 泻邪褉褌芯褔泻褍)\s+(.+)$/i);
    if (searchMatch && searchMatch[1]) {
      const queryName = searchMatch[1].trim();
      // Capitalize first letter for visual query cleanliness if possible
      const capitalizedQuery = queryName.charAt(0).toUpperCase() + queryName.slice(1);
      return {
        view: "patients",
        query: capitalizedQuery,
        feedbackText: `袠褖褍 锌邪褑懈械薪褌邪 ${capitalizedQuery}.`
      };
    }

    // 2. Check view mapping synonyms
    let view: string | undefined = undefined;
    let feedbackText = "";

    if (/(?:锌械褉械泄写懈|芯褌泻褉芯泄|锌械褉械泻谢褞褔懈|锌芯泻邪卸懈)\s+(?:胁|薪邪)?\s*(?:褋屑械薪|褉邪斜芯褌褍|锌械褉胁)/i.test(lower) || lower === "褋屑械薪邪") {
      view = "shift";
      feedbackText = "袨褌泻褉褘胁邪褞 褋屑械薪褍.";
    } else if (/(?:锌械褉械泄写懈|芯褌泻褉芯泄|锌械褉械泻谢褞褔懈|锌芯泻邪卸懈)\s+(?:胁|薪邪)?\s*(?:蟹邪锌懈褋|褉邪褋锌懈褋邪薪|泻邪谢械薪写邪褉|芯褔械褉械写)/i.test(lower) || /(蟹邪锌懈褋懈|褉邪褋锌懈褋邪薪懈械|泻邪谢械薪写邪褉褜)/i.test(lower)) {
      view = "schedule";
      feedbackText = "袨褌泻褉褘胁邪褞 褉邪褋锌懈褋邪薪懈械.";
    } else if (/(?:锌械褉械泄写懈|芯褌泻褉芯泄|锌械褉械泻谢褞褔懈|锌芯泻邪卸懈)\s+(?:胁|薪邪)?\s*(?:锌邪褑懈械薪褌|泻邪褉褌芯褔)/i.test(lower) || lower === "锌邪褑懈械薪褌褘") {
      view = "patients";
      feedbackText = "袨褌泻褉褘胁邪褞 褋锌懈褋芯泻 锌邪褑懈械薪褌芯胁.";
    } else if (/(?:锌械褉械泄写懈|芯褌泻褉芯泄|锌械褉械泻谢褞褔懈|锌芯泻邪卸懈)\s+(?:胁|薪邪)?\s*(?:褋薪懈屑|褉械薪褌谐械薪|胁懈蟹懈芯谐褉|泻谢泻褌|泻褌)/i.test(lower) || /(褋薪懈屑泻懈|褉械薪褌谐械薪|泻谢泻褌|胁懈蟹懈芯谐褉邪褎)/i.test(lower)) {
      view = "imaging";
      feedbackText = "袨褌泻褉褘胁邪褞 褋薪懈屑泻懈.";
    } else if (/(?:锌械褉械泄写懈|芯褌泻褉芯泄|锌械褉械泻谢褞褔懈|锌芯泻邪卸懈)\s+(?:胁|薪邪)?\s*(?:锌褉懈械屑|锌褉懈褢屑|褌械泻褍褖)/i.test(lower) || lower === "锌褉懈械屑" || lower === "锌褉懈褢屑") {
      view = "visit";
      feedbackText = "袨褌泻褉褘胁邪褞 褌械泻褍褖懈泄 锌褉懈械屑.";
    } else if (/(?:锌械褉械泄写懈|芯褌泻褉芯泄|锌械褉械泻谢褞褔懈|锌芯泻邪卸懈)\s+(?:胁|薪邪)?\s*(?:写芯泻褍屑械薪褌|写芯谐芯胁芯褉|褋芯谐谢邪褋懈|褋锌褉邪胁泻)/i.test(lower) || lower === "写芯泻褍屑械薪褌褘") {
      view = "documents";
      feedbackText = "袨褌泻褉褘胁邪褞 写芯泻褍屑械薪褌褘.";
    } else if (/(?:锌械褉械泄写懈|芯褌泻褉芯泄|锌械褉械泻谢褞褔懈|锌芯泻邪卸懈)\s+(?:胁|薪邪)?\s*(?:芯锌谢邪褌|褎懈薪邪薪褋|泻邪褋褋|写芯谢谐)/i.test(lower) || /(芯锌谢邪褌褘|褎懈薪邪薪褋褘|泻邪褋褋邪)/i.test(lower)) {
      view = "finance";
      feedbackText = "袨褌泻褉褘胁邪褞 褎懈薪邪薪褋褘.";
    } else if (/(?:锌械褉械泄写懈|芯褌泻褉芯泄|锌械褉械泻谢褞褔懈|锌芯泻邪卸懈)\s+(?:胁|薪邪)?\s*(?:褋胁褟蟹|褋芯芯斜褖械薪|蟹邪写邪褔|褔邪褌|褌械谢械谐褉邪屑)/i.test(lower) || lower === "褋胁褟蟹褜" || lower === "褋芯芯斜褖械薪懈褟") {
      view = "communications";
      feedbackText = "袨褌泻褉褘胁邪褞 褋胁褟蟹褜.";
    } else if (/(?:锌械褉械泄写懈|芯褌泻褉芯泄|锌械褉械泻谢褞褔懈|锌芯泻邪卸懈)\s+(?:胁|薪邪)?\s*(?:薪邪褋褌褉芯泄泻|锌褉芯褎懈谢|泻谢懈薪懈泻)/i.test(lower) || lower === "薪邪褋褌褉芯泄泻懈") {
      view = "settings";
      feedbackText = "袨褌泻褉褘胁邪褞 薪邪褋褌褉芯泄泻懈.";
    } else if (/(?:锌械褉械泄写懈|芯褌泻褉芯泄|锌械褉械泻谢褞褔懈|锌芯泻邪卸懈)\s+(?:胁|薪邪)?\s*(?:屑邪褉泻械褌懈薪谐|seo|褋械芯|芯褌蟹褘胁)/i.test(lower) || lower === "屑邪褉泻械褌懈薪谐") {
      view = "marketing";
      feedbackText = "袨褌泻褉褘胁邪褞 屑邪褉泻械褌懈薪谐.";
    }

    // 3. Date filtering for Schedule (if view is schedule or active, e.g. "蟹邪锌懈褋懈 薪邪 蟹邪胁褌褉邪", "褉邪褋锌懈褋邪薪懈械 褋械谐芯写薪褟")
    if (view === "schedule" || (!view && /(?:蟹邪锌懈褋|褉邪褋锌懈褋邪薪|泻邪谢械薪写邪褉|锌褉懈械屑|锌褉懈褢屑|蟹邪胁褌褉邪|褋械谐芯写薪褟|胁褔械褉邪|锌芯褋谢械蟹邪胁褌褉邪)/i.test(lower))) {
      const now = new Date();
      const offset = now.getTimezoneOffset();
      const getLocalDateString = (timeMs: number) => {
        return new Date(timeMs - (offset * 60 * 1000)).toISOString().split('T')[0];
      };

      let dateStr: string | undefined = undefined;
      let dateWord = "";

      if (lower.includes("褋械谐芯写薪褟")) {
        dateStr = getLocalDateString(now.getTime());
        dateWord = "褋械谐芯写薪褟";
      } else if (lower.includes("蟹邪胁褌褉邪")) {
        dateStr = getLocalDateString(now.getTime() + 24 * 60 * 60 * 1000);
        dateWord = "蟹邪胁褌褉邪";
      } else if (lower.includes("胁褔械褉邪")) {
        dateStr = getLocalDateString(now.getTime() - 24 * 60 * 60 * 1000);
        dateWord = "胁褔械褉邪";
      } else if (lower.includes("锌芯褋谢械蟹邪胁褌褉邪")) {
        dateStr = getLocalDateString(now.getTime() + 2 * 24 * 60 * 60 * 1000);
        dateWord = "锌芯褋谢械蟹邪胁褌褉邪";
      }

      if (dateStr) {
        return {
          view: view || "schedule",
          date: dateStr,
          feedbackText: view ? `袨褌泻褉褘胁邪褞 褉邪褋锌懈褋邪薪懈械 薪邪 ${dateWord}.` : `袩芯泻邪蟹褘胁邪褞 蟹邪锌懈褋懈 薪邪 ${dateWord}.`
        };
      }
    }

    if (view) {
      return { view, feedbackText };
    }

    return { feedbackText: "" };
  }
}
