// @dental/web/src/lib/semanticParser.ts
export interface ParsedDictation {
  complaints: string[];
  anamnesis: string[];
  objective: string[];
  diagnosis: string[];
  treatmentPlan: string[];
  toothNumbers: string[];
  hasMedicalTerms: boolean;
}

const terms = {
  complaints: ["斜芯谢懈褌", "斜芯谢褜", "卸邪谢芯斜", "薪芯械褌", "褔褍胁褋褌胁懈褌械谢褜薪芯褋褌褜", "褉械邪泻褑懈褟", "芯褌泻芯谢芯谢邪褋褜", "胁褘锌邪谢邪"],
  anamnesis: ["写邪胁薪芯", "屑械褋褟褑", "薪械写械谢褞", "胁褔械褉邪", "褉邪薪械械 谢械褔械薪", "薪械 芯斜褉邪褖邪谢褋褟"],
  objective: ["泻邪褉懈芯蟹薪邪褟 锌芯谢芯褋褌褜", "褉邪蟹褉褍褕械薪", "谐懈锌械褉械屑懈褟", "芯褌械泻", "锌械褉泻褍褋褋懈褟", "蟹芯薪写懈褉芯胁邪薪懈械", "薪邪谢械褌"],
  diagnosis: ["泻邪褉懈械褋", "锌褍谢褜锌懈褌", "锌械褉懈芯写芯薪褌懈褌", "谐懈薪谐懈胁懈褌", "褍写邪谢械薪懈械"],
  treatmentPlan: ["邪薪械褋褌械蟹懈褟", "锌褉械锌邪褉懈褉芯胁邪薪懈械", "锌谢芯屑斜邪", "薪械泻褉褝泻褌芯屑懈褟", "褝薪写芯写芯薪褌懈褟", "泻褌", "褋薪懈屑芯泻", "锌褉芯褎谐懈谐懈械薪邪"]
};

const toothRegex = /\b([1-4][1-8]|[5-8][1-5])\b(?!\s*[:.\-]\s*\d+)(?!\s*(?:褔邪褋芯胁|褔邪褋邪|褔|褍褌褉邪|写薪褟|胁械褔械褉邪|屑懈薪|屑懈薪褍褌|褟薪胁邪褉褟|褎械胁褉邪谢褟|屑邪褉褌邪|邪锌褉械谢褟|屑邪褟|懈褞薪褟|懈褞谢褟|邪胁谐褍褋褌邪|褋械薪褌褟斜褉褟|芯泻褌褟斜褉褟|薪芯褟斜褉褟|写械泻邪斜褉褟|褉褍斜|褉褍斜谢械泄|褌褘褋))/gi;

export function parseDictation(text: string): ParsedDictation {
  const result: ParsedDictation = {
    complaints: [],
    anamnesis: [],
    objective: [],
    diagnosis: [],
    treatmentPlan: [],
    toothNumbers: [],
    hasMedicalTerms: false
  };

  const lowerText = text.toLowerCase();
  
  // Find teeth
  const teeth = text.match(toothRegex);
  if (teeth) {
    result.toothNumbers = Array.from(new Set(teeth));
  }

  // Very naive sentence extraction for MVP
  const sentences = text.split(/[.?!]/).map(s => s.trim()).filter(Boolean);

  for (const sentence of sentences) {
    const lower = sentence.toLowerCase();
    let categorized = false;

    if (terms.complaints.some(t => lower.includes(t))) {
      result.complaints.push(sentence);
      categorized = true;
    }
    if (terms.anamnesis.some(t => lower.includes(t))) {
      result.anamnesis.push(sentence);
      categorized = true;
    }
    if (terms.objective.some(t => lower.includes(t))) {
      result.objective.push(sentence);
      categorized = true;
    }
    if (terms.diagnosis.some(t => lower.includes(t))) {
      result.diagnosis.push(sentence);
      categorized = true;
    }
    if (terms.treatmentPlan.some(t => lower.includes(t))) {
      result.treatmentPlan.push(sentence);
      categorized = true;
    }
    
    // If we matched any medical term
    if (categorized) {
      result.hasMedicalTerms = true;
    }
  }

  return result;
}
