// @cornerstonejs/tools/dist/esm/tools/VolumeCroppingControlTool.js
import { vec2 } from 'gl-matrix';
import vtkMath from '@kitware/vtk.js/Common/Core/Math';
import { AnnotationTool } from './base';
import { getRenderingEngine, getEnabledElementByIds, getEnabledElement, Enums, CONSTANTS, triggerEvent, eventTarget, convertColorArrayToRgbString, } from '@cornerstonejs/core';
import { getToolGroup } from '../store/ToolGroupManager';
import getViewportICamera from '../utilities/getViewportICamera';
import { resetViewportCamera } from '../utilities/setViewportCamera';
import { addAnnotation, getAnnotations, removeAnnotation, } from '../stateManagement/annotation/annotationState';
import { drawCircle as drawCircleSvg, drawLine as drawLineSvg, } from '../drawingSvg';
import { state } from '../store/state';
import { Events } from '../enums';
import { getViewportIdsWithToolToRender } from '../utilities/viewportFilters';
import { resetElementCursor, hideElementCursor, } from '../cursors/elementCursor';
import * as lineSegment from '../utilities/math/line';
import { isAnnotationLocked } from '../stateManagement/annotation/annotationLocking';
import triggerAnnotationRenderForViewportIds from '../utilities/triggerAnnotationRenderForViewportIds';
import { NUM_CLIPPING_PLANES, LINE_INTERSECTION_TOLERANCE, POINT_PROXIMITY_THRESHOLD_PIXELS, copyClippingPlanes, getColorKeyForPlaneIndex, getOrientationFromNormal, computePlanePlaneIntersection, findLineBoundsIntersection, } from '../utilities/volumeCropping';
const OPERATION = {
    DRAG: 1,
    ROTATE: 2,
    SLAB: 3,
};
class VolumeCroppingControlTool extends AnnotationTool {
    constructor(toolProps = {}, defaultToolProps = {
        supportedInteractionTypes: ['Mouse'],
        configuration: {
            viewportIndicators: false,
            viewportIndicatorsConfig: {
                radius: 5,
                x: null,
                y: null,
            },
            extendReferenceLines: true,
            initialCropFactor: 0.2,
            mobile: {
                enabled: false,
                opacity: 0.8,
            },
            lineColors: {
                AXIAL: [1.0, 0.0, 0.0],
                CORONAL: [0.0, 1.0, 0.0],
                SAGITTAL: [1.0, 1.0, 0.0],
                UNKNOWN: [0.0, 0.0, 1.0],
            },
            lineWidth: 1.5,
            lineWidthActive: 2.5,
        },
    }) {
        super(toolProps, defaultToolProps);
        this.clippingPlanes = [];
        this.initializeViewport = ({ renderingEngineId, viewportId, }) => {
            if (!renderingEngineId || !viewportId) {
                console.warn('VolumeCroppingControlTool: Missing renderingEngineId or viewportId');
                return;
            }
            const enabledElement = getEnabledElementByIds(viewportId, renderingEngineId);
            if (!enabledElement) {
                return;
            }
            const { viewport } = enabledElement;
            const volumeActors = viewport.getActors();
            if (volumeActors && volumeActors.length > 0) {
                const imageData = volumeActors[0].actor.getMapper().getInputData();
                if (imageData) {
                    this.seriesInstanceUID = imageData.seriesInstanceUID || 'unknown';
                }
            }
            const { element } = viewport;
            const { position, focalPoint } = getViewportICamera(viewport);
            let annotations = this._getAnnotations(enabledElement);
            annotations = this.filterInteractableAnnotationsForElement(element, annotations);
            if (annotations?.length) {
                removeAnnotation(annotations[0].annotationUID);
            }
            const orientation = getOrientationFromNormal(getViewportICamera(viewport).viewPlaneNormal);
            const annotation = {
                highlighted: false,
                metadata: {
                    cameraPosition: [...position],
                    cameraFocalPoint: [...focalPoint],
                    toolName: this.getToolName(),
                },
                data: {
                    handles: {
                        activeOperation: null,
                        clippingPlanes: this.clippingPlanes.length > 0
                            ? copyClippingPlanes(this.clippingPlanes)
                            : [],
                    },
                    activeViewportIds: [],
                    viewportId,
                    referenceLines: [],
                    orientation,
                },
            };
            addAnnotation(annotation, element);
        };
        this._getViewportsInfo = () => {
            const viewports = getToolGroup(this.toolGroupId).viewportsInfo;
            return viewports;
        };
        this.resetCroppingSpheres = () => {
            const viewportsInfo = this._getViewportsInfo();
            for (const viewportInfo of viewportsInfo) {
                const { viewportId, renderingEngineId } = viewportInfo;
                const enabledElement = getEnabledElementByIds(viewportId, renderingEngineId);
                const viewport = enabledElement.viewport;
                const resetPan = true;
                const resetZoom = true;
                const resetToCenter = true;
                const resetRotation = true;
                const suppressEvents = true;
                resetViewportCamera(viewport, {
                    resetPan,
                    resetZoom,
                    resetToCenter,
                    resetRotation,
                    suppressEvents,
                });
                viewport.resetSlabThickness();
                const { element } = viewport;
                let annotations = this._getAnnotations(enabledElement);
                annotations = this.filterInteractableAnnotationsForElement(element, annotations);
                if (annotations.length) {
                    removeAnnotation(annotations[0].annotationUID);
                }
                viewport.render();
            }
            this._initializeViewports(viewportsInfo);
        };
        this._initializeViewports = (viewportsInfo) => {
            if (!viewportsInfo?.length || !viewportsInfo[0]) {
                console.warn('VolumeCroppingControlTool: No valid viewportsInfo for initialization.');
                return;
            }
            viewportsInfo.forEach((vpInfo) => {
                this.initializeViewport(vpInfo);
            });
            triggerAnnotationRenderForViewportIds(viewportsInfo.map(({ viewportId }) => viewportId));
        };
        this.cancel = () => {
        };
        this.isPointNearTool = (element, annotation, canvasCoords, proximity) => {
            if (this._pointNearTool(element, annotation, canvasCoords, POINT_PROXIMITY_THRESHOLD_PIXELS)) {
                return true;
            }
            return false;
        };
        this.toolSelectedCallback = (evt, annotation, interactionType) => {
            const eventDetail = evt.detail;
            const { element } = eventDetail;
            annotation.highlighted = true;
            this._activateModify(element);
            hideElementCursor(element);
            evt.preventDefault();
        };
        this.onResetCamera = (evt) => {
            this.resetCroppingSpheres();
        };
        this.mouseMoveCallback = (evt, filteredToolAnnotations) => {
            if (!filteredToolAnnotations) {
                return;
            }
            const { element, currentPoints } = evt.detail;
            const canvasCoords = currentPoints.canvas;
            let imageNeedsUpdate = false;
            for (let i = 0; i < filteredToolAnnotations.length; i++) {
                const annotation = filteredToolAnnotations[i];
                if (isAnnotationLocked(annotation.annotationUID)) {
                    continue;
                }
                const { data, highlighted } = annotation;
                if (!data.handles) {
                    continue;
                }
                data.activeViewportIds = [];
                const near = this._pointNearTool(element, annotation, canvasCoords, POINT_PROXIMITY_THRESHOLD_PIXELS);
                const nearToolAndNotMarkedActive = near && !highlighted;
                const notNearToolAndMarkedActive = !near && highlighted;
                if (nearToolAndNotMarkedActive || notNearToolAndMarkedActive) {
                    annotation.highlighted = !highlighted;
                    imageNeedsUpdate = true;
                }
            }
            return imageNeedsUpdate;
        };
        this.filterInteractableAnnotationsForElement = (element, annotations) => {
            if (!annotations || !annotations.length) {
                return [];
            }
            const enabledElement = getEnabledElement(element);
            let orientation = null;
            if (enabledElement.viewport) {
                const { viewPlaneNormal } = getViewportICamera(enabledElement.viewport);
                if (viewPlaneNormal) {
                    orientation = getOrientationFromNormal(viewPlaneNormal);
                }
            }
            const filtered = annotations.filter((annotation) => {
                if (annotation.data.orientation &&
                    orientation &&
                    annotation.data.orientation === orientation) {
                    return true;
                }
                return false;
            });
            return filtered;
        };
        this.renderAnnotation = (enabledElement, svgDrawingHelper) => {
            function lineIntersection2D(p1, p2, q1, q2) {
                const s1_x = p2[0] - p1[0];
                const s1_y = p2[1] - p1[1];
                const s2_x = q2[0] - q1[0];
                const s2_y = q2[1] - q1[1];
                const denom = -s2_x * s1_y + s1_x * s2_y;
                if (Math.abs(denom) < LINE_INTERSECTION_TOLERANCE) {
                    return null;
                }
                const s = (-s1_y * (p1[0] - q1[0]) + s1_x * (p1[1] - q1[1])) / denom;
                const t = (s2_x * (p1[1] - q1[1]) - s2_y * (p1[0] - q1[0])) / denom;
                if (s >= 0 && s <= 1 && t >= 0 && t <= 1) {
                    return [p1[0] + t * s1_x, p1[1] + t * s1_y];
                }
                return null;
            }
            const viewportsInfo = this._getViewportsInfo();
            if (!viewportsInfo || viewportsInfo.length === 0) {
                return false;
            }
            let renderStatus = false;
            const { viewport, renderingEngine } = enabledElement;
            const { element } = viewport;
            const annotations = this._getAnnotations(enabledElement);
            const camera = getViewportICamera(viewport);
            const filteredToolAnnotations = this.filterInteractableAnnotationsForElement(element, annotations);
            const viewportAnnotation = filteredToolAnnotations[0];
            if (!viewportAnnotation || !viewportAnnotation.data) {
                return renderStatus;
            }
            const annotationUID = viewportAnnotation.annotationUID;
            const { clientWidth, clientHeight } = viewport.canvas;
            const canvasDiagonalLength = Math.sqrt(clientWidth * clientWidth + clientHeight * clientHeight);
            const data = viewportAnnotation.data;
            let clippingPlanes = viewportAnnotation.data.handles.clippingPlanes;
            if (!clippingPlanes || clippingPlanes.length < NUM_CLIPPING_PLANES) {
                if (this.clippingPlanes &&
                    this.clippingPlanes.length >= NUM_CLIPPING_PLANES) {
                    clippingPlanes = this.clippingPlanes;
                    data.handles.clippingPlanes = copyClippingPlanes(this.clippingPlanes);
                }
                else {
                    return false;
                }
            }
            const { viewPlaneNormal, focalPoint } = camera;
            const referenceLines = [];
            const planeTypes = [
                'min',
                'max',
                'min',
                'max',
                'min',
                'max',
            ];
            for (let planeIndex = 0; planeIndex < NUM_CLIPPING_PLANES; planeIndex++) {
                const clippingPlane = clippingPlanes[planeIndex];
                const intersection = computePlanePlaneIntersection(clippingPlane, viewPlaneNormal, focalPoint);
                if (!intersection) {
                    continue;
                }
                const lineBounds = findLineBoundsIntersection(intersection.point, intersection.direction, viewport);
                if (!lineBounds) {
                    continue;
                }
                referenceLines.push([
                    {
                        id: viewport.id,
                        canvas: viewport.canvas,
                    },
                    lineBounds.start,
                    lineBounds.end,
                    planeTypes[planeIndex],
                    planeIndex,
                ]);
            }
            data.referenceLines = referenceLines;
            const viewportColor = this._getReferenceLineColor(viewport.id);
            const defaultColor = viewportColor !== undefined ? viewportColor : 'rgb(200, 200, 200)';
            referenceLines.forEach((line, lineIndex) => {
                const [otherViewport, startPoint, endPoint, type, planeIndex] = line;
                if (planeIndex === undefined ||
                    planeIndex < 0 ||
                    planeIndex >= NUM_CLIPPING_PLANES) {
                    return;
                }
                const intersections = [];
                for (let j = 0; j < referenceLines.length; ++j) {
                    if (j === lineIndex) {
                        continue;
                    }
                    const otherLine = referenceLines[j];
                    const intersection = lineIntersection2D(startPoint, endPoint, otherLine[1], otherLine[2]);
                    if (intersection) {
                        intersections.push({
                            with: otherLine[3],
                            point: intersection,
                        });
                    }
                }
                const colorKey = getColorKeyForPlaneIndex(planeIndex);
                const lineColors = this.configuration.lineColors || {};
                const colorArr = colorKey
                    ? lineColors[colorKey] || lineColors.UNKNOWN || [1.0, 0.0, 0.0]
                    : [1.0, 0.0, 0.0];
                const color = convertColorArrayToRgbString(colorArr);
                const viewportControllable = this._getReferenceLineControllable(otherViewport.id);
                const selectedViewportId = data.activeViewportIds.find((id) => id === otherViewport.id);
                let lineWidth = this.configuration.lineWidth ?? 1.5;
                const lineActive = data.handles.activeOperation !== null &&
                    data.handles.activeOperation === OPERATION.DRAG &&
                    selectedViewportId;
                if (lineActive) {
                    lineWidth = this.configuration.activeLineWidth ?? 2.5;
                }
                const lineUID = `plane_${planeIndex}`;
                if (intersections.length === 2) {
                    drawLineSvg(svgDrawingHelper, annotationUID, lineUID, intersections[0].point, intersections[1].point, {
                        color,
                        lineWidth,
                    });
                }
                else {
                    drawLineSvg(svgDrawingHelper, annotationUID, lineUID, startPoint, endPoint, {
                        color,
                        lineWidth,
                    });
                }
                if (this.configuration.extendReferenceLines &&
                    intersections.length === 2) {
                    const sortedIntersections = intersections
                        .map((intersection) => ({
                        ...intersection,
                        distance: vec2.distance(startPoint, intersection.point),
                    }))
                        .sort((a, b) => a.distance - b.distance);
                    drawLineSvg(svgDrawingHelper, annotationUID, lineUID + '_dashed_before', startPoint, sortedIntersections[0].point, { color, lineWidth, lineDash: [4, 4] });
                    drawLineSvg(svgDrawingHelper, annotationUID, lineUID + '_dashed_after', sortedIntersections[1].point, endPoint, { color, lineWidth, lineDash: [4, 4] });
                }
            });
            renderStatus = true;
            if (this.configuration.viewportIndicators) {
                const { viewportIndicatorsConfig } = this.configuration;
                const xOffset = viewportIndicatorsConfig?.xOffset || 0.95;
                const yOffset = viewportIndicatorsConfig?.yOffset || 0.05;
                const referenceColorCoordinates = [
                    clientWidth * xOffset,
                    clientHeight * yOffset,
                ];
                const circleRadius = viewportIndicatorsConfig?.circleRadius || canvasDiagonalLength * 0.01;
                const circleUID = '0';
                drawCircleSvg(svgDrawingHelper, annotationUID, circleUID, referenceColorCoordinates, circleRadius, { color: defaultColor, fill: defaultColor });
            }
            return renderStatus;
        };
        this._getAnnotations = (enabledElement) => {
            const { viewport } = enabledElement;
            const annotations = getAnnotations(this.getToolName(), viewport.element) || [];
            const viewportIds = this._getViewportsInfo().map(({ viewportId }) => viewportId);
            const toolGroupAnnotations = annotations.filter((annotation) => {
                const { data } = annotation;
                return viewportIds.includes(data.viewportId);
            });
            return toolGroupAnnotations;
        };
        this._onSphereMoved = (evt) => {
            if (evt.detail.originalClippingPlanes) {
                this._syncWithVolumeCroppingTool(evt.detail.originalClippingPlanes);
            }
            else {
                if (evt.detail.seriesInstanceUID !== this.seriesInstanceUID) {
                    return;
                }
                return;
            }
        };
        this._onNewVolume = () => {
            const viewportsInfo = this._getViewportsInfo();
            if (viewportsInfo && viewportsInfo.length > 0) {
                const { viewportId, renderingEngineId } = viewportsInfo[0];
                const renderingEngine = getRenderingEngine(renderingEngineId);
                if (!renderingEngine) {
                    return;
                }
                const viewport = renderingEngine.getViewport(viewportId);
                if (!viewport) {
                    return;
                }
                const volumeActors = viewport.getActors();
                if (volumeActors.length > 0) {
                    const imageData = volumeActors[0].actor.getMapper().getInputData();
                    if (imageData) {
                        this.seriesInstanceUID = imageData.seriesInstanceUID;
                    }
                }
            }
            this._initializeViewports(viewportsInfo);
            triggerEvent(eventTarget, Events.VOLUMECROPPINGCONTROL_TOOL_CHANGED, {
                toolGroupId: this.toolGroupId,
                viewportsInfo: viewportsInfo,
                seriesInstanceUID: this.seriesInstanceUID,
            });
        };
        this._activateModify = (element) => {
            state.isInteractingWithTool = !this.configuration.mobile?.enabled;
            element.addEventListener(Events.MOUSE_UP, this._endCallback);
            element.addEventListener(Events.MOUSE_DRAG, this._dragCallback);
            element.addEventListener(Events.MOUSE_CLICK, this._endCallback);
            element.addEventListener(Events.TOUCH_END, this._endCallback);
            element.addEventListener(Events.TOUCH_DRAG, this._dragCallback);
            element.addEventListener(Events.TOUCH_TAP, this._endCallback);
        };
        this._deactivateModify = (element) => {
            state.isInteractingWithTool = false;
            element.removeEventListener(Events.MOUSE_UP, this._endCallback);
            element.removeEventListener(Events.MOUSE_DRAG, this._dragCallback);
            element.removeEventListener(Events.MOUSE_CLICK, this._endCallback);
            element.removeEventListener(Events.TOUCH_END, this._endCallback);
            element.removeEventListener(Events.TOUCH_DRAG, this._dragCallback);
            element.removeEventListener(Events.TOUCH_TAP, this._endCallback);
        };
        this._endCallback = (evt) => {
            const eventDetail = evt.detail;
            const { element } = eventDetail;
            this.editData.annotation.data.handles.activeOperation = null;
            this.editData.annotation.data.activeViewportIds = [];
            this._deactivateModify(element);
            resetElementCursor(element);
            this.editData = null;
            const requireSameOrientation = false;
            const viewportIdsToRender = getViewportIdsWithToolToRender(element, this.getToolName(), requireSameOrientation);
            triggerAnnotationRenderForViewportIds(viewportIdsToRender);
        };
        this._dragCallback = (evt) => {
            const eventDetail = evt.detail;
            const delta = eventDetail.deltaPoints.world;
            if (Math.abs(delta[0]) < 1e-3 &&
                Math.abs(delta[1]) < 1e-3 &&
                Math.abs(delta[2]) < 1e-3) {
                return;
            }
            const { element } = eventDetail;
            const enabledElement = getEnabledElement(element);
            const { viewport } = enabledElement;
            if (viewport.type === Enums.ViewportType.VOLUME_3D) {
                return;
            }
            const annotations = this._getAnnotations(enabledElement);
            const filteredToolAnnotations = this.filterInteractableAnnotationsForElement(element, annotations);
            const viewportAnnotation = filteredToolAnnotations[0];
            if (!viewportAnnotation) {
                return;
            }
            const { handles } = viewportAnnotation.data;
            const clippingPlanes = handles.clippingPlanes;
            if (handles.activeOperation === OPERATION.DRAG &&
                clippingPlanes &&
                clippingPlanes.length >= NUM_CLIPPING_PLANES) {
                if (handles.activePlaneIndex !== undefined &&
                    handles.activePlaneIndex >= 0 &&
                    handles.activePlaneIndex < NUM_CLIPPING_PLANES) {
                    const planeIndex = handles.activePlaneIndex;
                    const plane = clippingPlanes[planeIndex];
                    const normal = plane.normal;
                    const dotProd = vtkMath.dot(delta, normal);
                    const moveDistance = dotProd;
                    plane.origin[0] += normal[0] * moveDistance;
                    plane.origin[1] += normal[1] * moveDistance;
                    plane.origin[2] += normal[2] * moveDistance;
                }
                else if (handles.activeType === 'min') {
                    clippingPlanes[0].origin[0] += delta[0];
                    clippingPlanes[2].origin[1] += delta[1];
                    clippingPlanes[4].origin[2] += delta[2];
                }
                else if (handles.activeType === 'max') {
                    clippingPlanes[1].origin[0] += delta[0];
                    clippingPlanes[3].origin[1] += delta[1];
                    clippingPlanes[5].origin[2] += delta[2];
                }
                this.clippingPlanes = copyClippingPlanes(clippingPlanes);
                const viewportsInfo = this._getViewportsInfo();
                viewportsInfo.forEach(({ viewportId, renderingEngineId }) => {
                    const enabledElement = getEnabledElementByIds(viewportId, renderingEngineId);
                    if (enabledElement) {
                        const annotations = this._getAnnotations(enabledElement);
                        annotations.forEach((annotation) => {
                            if (annotation.data?.handles) {
                                annotation.data.handles.clippingPlanes = copyClippingPlanes(this.clippingPlanes);
                            }
                        });
                    }
                });
                triggerAnnotationRenderForViewportIds(viewportsInfo.map(({ viewportId }) => viewportId));
                triggerEvent(eventTarget, Events.VOLUMECROPPINGCONTROL_TOOL_CHANGED, {
                    toolGroupId: this.toolGroupId,
                    clippingPlanes: this.clippingPlanes,
                    handleType: handles.activeType,
                    seriesInstanceUID: this.seriesInstanceUID,
                });
            }
        };
        this._getReferenceLineColor =
            toolProps.configuration?.getReferenceLineColor ||
                (() => 'rgb(0, 200, 0)');
        this._getReferenceLineControllable =
            toolProps.configuration?.getReferenceLineControllable || (() => true);
        const viewportsInfo = getToolGroup(this.toolGroupId)?.viewportsInfo;
        eventTarget.addEventListener(Events.VOLUMECROPPING_TOOL_CHANGED, this._onSphereMoved);
        if (viewportsInfo && viewportsInfo.length > 0) {
            const { viewportId, renderingEngineId } = viewportsInfo[0];
            const renderingEngine = getRenderingEngine(renderingEngineId);
            if (!renderingEngine) {
                return;
            }
            const viewport = renderingEngine.getViewport(viewportId);
            if (!viewport) {
                return;
            }
            const volumeActors = viewport.getActors();
            if (!volumeActors || !volumeActors.length) {
                console.warn(`VolumeCroppingControlTool: No volume actors found in viewport ${viewportId}.`);
                return;
            }
            const imageData = volumeActors[0].actor.getMapper().getInputData();
            if (imageData) {
                this.seriesInstanceUID = imageData.seriesInstanceUID || 'unknown';
            }
        }
    }
    onSetToolInactive() {
    }
    onSetToolActive() {
        const viewportsInfo = this._getViewportsInfo();
        let anyAnnotationExists = false;
        for (const vpInfo of viewportsInfo) {
            const enabledElement = getEnabledElementByIds(vpInfo.viewportId, vpInfo.renderingEngineId);
            const annotations = this._getAnnotations(enabledElement);
            if (annotations && annotations.length > 0) {
                anyAnnotationExists = true;
                break;
            }
        }
        if (!anyAnnotationExists) {
            this._unsubscribeToViewportNewVolumeSet(viewportsInfo);
            this._subscribeToViewportNewVolumeSet(viewportsInfo);
            this._initializeViewports(viewportsInfo);
            triggerEvent(eventTarget, Events.VOLUMECROPPINGCONTROL_TOOL_CHANGED, {
                toolGroupId: this.toolGroupId,
                viewportsInfo: viewportsInfo,
                seriesInstanceUID: this.seriesInstanceUID,
            });
        }
        else {
            for (const vpInfo of viewportsInfo) {
                const enabledElement = getEnabledElementByIds(vpInfo.viewportId, vpInfo.renderingEngineId);
                if (!enabledElement) {
                    continue;
                }
                const annotations = this._getAnnotations(enabledElement);
                if (annotations && annotations.length > 0) {
                    annotations.forEach((annotation) => {
                        removeAnnotation(annotation.annotationUID);
                    });
                }
                enabledElement.viewport.render();
            }
        }
    }
    onSetToolEnabled() {
        eventTarget.addEventListener(Events.VOLUMECROPPING_TOOL_CHANGED, this._onSphereMoved);
    }
    onSetToolDisabled() {
        const viewportsInfo = this._getViewportsInfo();
        this._unsubscribeToViewportNewVolumeSet(viewportsInfo);
        eventTarget.removeEventListener(Events.VOLUMECROPPING_TOOL_CHANGED, this._onSphereMoved);
        viewportsInfo.forEach(({ renderingEngineId, viewportId }) => {
            const enabledElement = getEnabledElementByIds(viewportId, renderingEngineId);
            if (!enabledElement) {
                return;
            }
            const annotations = this._getAnnotations(enabledElement);
            if (annotations?.length) {
                annotations.forEach((annotation) => {
                    removeAnnotation(annotation.annotationUID);
                });
            }
        });
    }
    _syncWithVolumeCroppingTool(originalClippingPlanes) {
        if (!originalClippingPlanes ||
            originalClippingPlanes.length < NUM_CLIPPING_PLANES) {
            return;
        }
        this.clippingPlanes = copyClippingPlanes(originalClippingPlanes);
        const viewportsInfo = this._getViewportsInfo();
        viewportsInfo.forEach(({ viewportId, renderingEngineId }) => {
            const enabledElement = getEnabledElementByIds(viewportId, renderingEngineId);
            if (enabledElement) {
                const annotations = this._getAnnotations(enabledElement);
                annotations.forEach((annotation) => {
                    if (annotation.data?.handles) {
                        annotation.data.handles.clippingPlanes = copyClippingPlanes(this.clippingPlanes);
                    }
                });
            }
        });
        triggerAnnotationRenderForViewportIds(viewportsInfo.map(({ viewportId }) => viewportId));
    }
    addNewAnnotation(evt) {
        const eventDetail = evt.detail;
        const { element } = eventDetail;
        const enabledElement = getEnabledElement(element);
        const { viewport } = enabledElement;
        const annotations = this._getAnnotations(enabledElement);
        const filteredAnnotations = this.filterInteractableAnnotationsForElement(viewport.element, annotations);
        if (!filteredAnnotations ||
            filteredAnnotations.length === 0 ||
            !filteredAnnotations[0]) {
            return null;
        }
        const { data } = filteredAnnotations[0];
        const viewportIdArray = [];
        const referenceLines = data.referenceLines || [];
        for (let i = 0; i < referenceLines.length; ++i) {
            const otherViewport = referenceLines[i][0];
            const viewportControllable = this._getReferenceLineControllable(otherViewport.id);
            if (!viewportControllable) {
                continue;
            }
            viewportIdArray.push(otherViewport.id);
        }
        data.activeViewportIds = [...viewportIdArray];
        data.handles.activeOperation = OPERATION.DRAG;
        evt.preventDefault();
        hideElementCursor(element);
        this._activateModify(element);
        return filteredAnnotations[0];
    }
    handleSelectedCallback(evt, annotation, handle, interactionType) {
        this.toolSelectedCallback(evt, annotation, interactionType);
    }
    _unsubscribeToViewportNewVolumeSet(viewportsInfo) {
        viewportsInfo.forEach(({ viewportId, renderingEngineId }) => {
            const { viewport } = getEnabledElementByIds(viewportId, renderingEngineId);
            const { element } = viewport;
            element.removeEventListener(Enums.Events.VOLUME_VIEWPORT_NEW_VOLUME, this._onNewVolume);
        });
    }
    _subscribeToViewportNewVolumeSet(viewports) {
        viewports.forEach(({ viewportId, renderingEngineId }) => {
            const { viewport } = getEnabledElementByIds(viewportId, renderingEngineId);
            const { element } = viewport;
            element.addEventListener(Enums.Events.VOLUME_VIEWPORT_NEW_VOLUME, this._onNewVolume);
        });
    }
    _pointNearTool(element, annotation, canvasCoords, proximity) {
        const { data } = annotation;
        const referenceLines = data.referenceLines;
        const viewportIdArray = [];
        if (referenceLines) {
            for (let i = 0; i < referenceLines.length; ++i) {
                const [otherViewport, startPoint, endPoint, type, planeIndex] = referenceLines[i];
                const distance = lineSegment.distanceToPoint(startPoint, endPoint, [
                    canvasCoords[0],
                    canvasCoords[1],
                ]);
                if (distance <= proximity) {
                    viewportIdArray.push(otherViewport.id);
                    data.handles.activeOperation = OPERATION.DRAG;
                    data.handles.activeType = type;
                    data.handles.activePlaneIndex = planeIndex;
                }
            }
        }
        data.activeViewportIds = [...viewportIdArray];
        this.editData = {
            annotation,
        };
        return data.handles.activeOperation === OPERATION.DRAG ? true : false;
    }
}
VolumeCroppingControlTool.toolName = 'VolumeCroppingControl';
export default VolumeCroppingControlTool;
